diff --git a/src/main/scala/xiangshan/backend/fu/NewCSR/NewCSR.scala b/src/main/scala/xiangshan/backend/fu/NewCSR/NewCSR.scala
index 217ede9f02e..6266846eb89 100644
--- a/src/main/scala/xiangshan/backend/fu/NewCSR/NewCSR.scala
+++ b/src/main/scala/xiangshan/backend/fu/NewCSR/NewCSR.scala
@@ -394,7 +394,7 @@ class NewCSR(implicit val p: Parameters) extends Module
 
   val intrVec = RegEnable(intrMod.io.out.interruptVec.bits, 0.U, intrMod.io.out.interruptVec.valid)
   val debug = RegEnable(intrMod.io.out.debug, false.B, intrMod.io.out.interruptVec.valid)
-  val nmi = RegEnable(intrMod.io.out.nmi, false.B, intrMod.io.out.interruptVec.valid)
+  val nmi = intrMod.io.out.nmi
   val virtualInterruptIsHvictlInject = RegEnable(intrMod.io.out.virtualInterruptIsHvictlInject, false.B, intrMod.io.out.interruptVec.valid)
   val irToHS = RegEnable(intrMod.io.out.irToHS, false.B, intrMod.io.out.interruptVec.valid)
   val irToVS = RegEnable(intrMod.io.out.irToVS, false.B, intrMod.io.out.interruptVec.valid)
@@ -408,7 +408,6 @@ class NewCSR(implicit val p: Parameters) extends Module
 
   trapHandleMod.io.in.trapInfo.valid := hasTrap
   trapHandleMod.io.in.trapInfo.bits.trapVec := trapVec.asUInt
-  trapHandleMod.io.in.trapInfo.bits.nmi := nmi
   trapHandleMod.io.in.trapInfo.bits.intrVec := intrVec
   trapHandleMod.io.in.trapInfo.bits.isInterrupt := trapIsInterrupt
   trapHandleMod.io.in.trapInfo.bits.irToHS := irToHS
diff --git a/src/main/scala/xiangshan/backend/fu/NewCSR/TrapHandleModule.scala b/src/main/scala/xiangshan/backend/fu/NewCSR/TrapHandleModule.scala
index e4b13e51500..82f86d64092 100644
--- a/src/main/scala/xiangshan/backend/fu/NewCSR/TrapHandleModule.scala
+++ b/src/main/scala/xiangshan/backend/fu/NewCSR/TrapHandleModule.scala
@@ -24,7 +24,6 @@ class TrapHandleModule extends Module {
   private val hvien = io.in.hvien.asUInt
 
   private val hasTrap = trapInfo.valid
-  private val hasNMI = hasTrap && trapInfo.bits.nmi
   private val hasIR = hasTrap && trapInfo.bits.isInterrupt
   private val hasEX = hasTrap && !trapInfo.bits.isInterrupt
 
@@ -57,8 +56,8 @@ class TrapHandleModule extends Module {
 
   // nmi handle in MMode only and default handler is mtvec
   private val  mHasIR = hasIR
-  private val hsHasIR = hasIR && irToHS & !hasNMI
-  private val vsHasIR = hasIR && irToVS & !hasNMI
+  private val hsHasIR = hasIR && irToHS
+  private val vsHasIR = hasIR && irToVS
 
   private val  mHasEX =  mEXVec.orR
   private val hsHasEX = hsEXVec.orR
@@ -118,7 +117,6 @@ class TrapHandleIO extends Bundle {
   val in = Input(new Bundle {
     val trapInfo = ValidIO(new Bundle {
       val trapVec = UInt(64.W)
-      val nmi = Bool()
       val intrVec = UInt(8.W)
       val isInterrupt = Bool()
       val singleStep = Bool()
