diff --git a/src/main/scala/xiangshan/frontend/bpu/WriteBuffer.scala b/src/main/scala/xiangshan/frontend/bpu/WriteBuffer.scala
index 81c92486a50..381376eb8d2 100644
--- a/src/main/scala/xiangshan/frontend/bpu/WriteBuffer.scala
+++ b/src/main/scala/xiangshan/frontend/bpu/WriteBuffer.scala
@@ -36,7 +36,7 @@ import xiangshan.XSModule
  * @param numEntries The number of entries in the write buffer
  * @param numPorts The number of write ports
  * @param hasCnt Whether the write request bundle has a counter field, used to update the entry's useful counter
- * @param hasFlush Whether the write buffer has a flush signal, used to reset the write bufferq
+ * @param hasFlush Whether the write buffer has a flush signal, used to reset the write buffer
 */
 class WriteBuffer[T <: WriteReqBundle](
     gen:            T,
@@ -134,7 +134,7 @@ class WriteBuffer[T <: WriteReqBundle](
       val notUseful    = notUsefulVec.reduce(_ || _)
       val notUsefulIdx = PriorityEncoder(notUsefulVec)
       val victim       = Mux(notUseful, notUsefulIdx, replacerWay(portIdx))
-      // if this wirte port !hit need to write a new entry
+      // if this write port !hit need to write a new entry
       when(!hit) {
         entries(portIdx)(victim)     := io.write(portIdx).bits
         valids(portIdx)(victim)      := true.B
@@ -160,7 +160,7 @@ class WriteBuffer[T <: WriteReqBundle](
         }
       }
 
-      // If the write request has conuter need update counter
+      // If the write request has counter need update counter
       val temporarily = WireInit(io.write(portIdx).bits)
       if (hasCnt) {
         when(hit) {
diff --git a/src/main/scala/xiangshan/frontend/bpu/phr/Phr.scala b/src/main/scala/xiangshan/frontend/bpu/phr/Phr.scala
index ae2bff97f18..53eaca821ea 100644
--- a/src/main/scala/xiangshan/frontend/bpu/phr/Phr.scala
+++ b/src/main/scala/xiangshan/frontend/bpu/phr/Phr.scala
@@ -39,12 +39,15 @@ class Phr(implicit p: Parameters) extends PhrModule with HasPhrParameters with H
   }
   val io: PhrIO = IO(new PhrIO)
 
-  private val phr = RegInit(0.U.asTypeOf(Vec(PhrHistoryLength, Bool())))
+  private val phr    = RegInit(0.U.asTypeOf(Vec(PhrHistoryLength, Bool())))
+  private val phrPtr = RegInit(0.U.asTypeOf(new PhrPtr))
+
+  private def getPhr(ptr: PhrPtr): UInt =
+    (Cat(phr.asUInt, phr.asUInt) >> (ptr.value + 1.U))(PhrHistoryLength - 1, 0)
 
   /*
    * PHR train from redirct/s2_prediction/s3_prediction
    */
-  private val phrPtr = RegInit(0.U.asTypeOf(new PhrPtr))
 
   private val s0_stall = io.train.s0_stall
   private val s1_valid = io.train.s1_valid
@@ -70,21 +73,22 @@ class Phr(implicit p: Parameters) extends PhrModule with HasPhrParameters with H
   private val s2_phrPtr    = RegEnable(s1_phrPtr, 0.U.asTypeOf(new PhrPtr), s1_fire)
   private val s3_phrPtr    = RegEnable(s2_phrPtr, 0.U.asTypeOf(new PhrPtr), s2_fire)
 
+  private val s0_phrValue = getPhr(s0_phrPtr) // debug use it
+  private val phrValue    = getPhr(phrPtr)
+
   private val redirectData    = WireInit(0.U.asTypeOf(new PhrUpdateData))
   private val s1_overrideData = WireInit(0.U.asTypeOf(new PhrUpdateData))
   private val s3_override     = WireInit(false.B)
   private val s3_overrideData = WireInit(0.U.asTypeOf(new PhrUpdateData))
 
-  private val updateData     = WireInit(0.U.asTypeOf(new PhrUpdateData))
-  private val updatePc       = WireInit(0.U.asTypeOf(PrunedAddr(VAddrBits)))
-  private val updateOverride = WireInit(false.B)
-  private val redirctPhr     = WireInit(0.U(PhrHistoryLength.W))
+  private val updateData  = WireInit(0.U.asTypeOf(new PhrUpdateData))
+  private val updatePc    = WireInit(0.U.asTypeOf(PrunedAddr(VAddrBits)))
+  private val redirectPhr = WireInit(0.U(PhrHistoryLength.W))
 
   redirectData.valid  := io.train.redirect.valid
   redirectData.taken  := io.train.redirect.bits.taken
   redirectData.pc     := io.train.redirect.bits.startVAddr
   redirectData.phrPtr := io.train.redirect.bits.speculationMeta.phrHistPtr
-  // redirectData.phrPtr := io.train.redirect.bits.speculationMeta.phrHistPtr
 
   s3_override               := io.train.s3_override
   s3_overrideData.valid     := s3_override
@@ -99,9 +103,6 @@ class Phr(implicit p: Parameters) extends PhrModule with HasPhrParameters with H
   s1_overrideData.phrPtr    := s1_phrPtr
   s1_overrideData.foldedPhr := s1_foldedPhrReg
 
-  updatePc       := updateData.pc
-  updateOverride := redirectData.valid || s3_override
-
   updateData := MuxCase(
     0.U.asTypeOf(new PhrUpdateData),
     Seq(
@@ -110,12 +111,12 @@ class Phr(implicit p: Parameters) extends PhrModule with HasPhrParameters with H
       s1_valid           -> s1_overrideData
     )
   )
+
+  updatePc := updateData.pc
+
   private val shiftBits =
     (((updatePc >> 1) ^ (updatePc >> 3)) ^ ((updatePc >> 5) ^ (updatePc >> 7)))(Shamt - 1, 0)
 
-  private def getPhr(ptr: PhrPtr): UInt =
-    (Cat(phr.asUInt, phr.asUInt) >> (ptr.value + 1.U))(PhrHistoryLength - 1, 0)
-
   when(updateData.valid) {
     phrPtr    := updateData.phrPtr
     s0_phrPtr := updateData.phrPtr
@@ -134,18 +135,18 @@ class Phr(implicit p: Parameters) extends PhrModule with HasPhrParameters with H
    */
   AllFoldedHistoryInfo.foreach { info =>
     s0_foldedPhr.getHistWithInfo(info).foldedHist :=
-      computeFoldedHist(phr.asUInt, info.FoldedLength)(info.HistoryLength)
+      computeFoldedHist(phrValue, info.FoldedLength)(info.HistoryLength)
   }
 
   when(redirectData.valid) {
-    redirctPhr := getPhr(redirectData.phrPtr)
+    redirectPhr := getPhr(redirectData.phrPtr)
     AllFoldedHistoryInfo.foreach { info =>
       redirectData.foldedPhr.getHistWithInfo(info).foldedHist :=
-        computeFoldedHist(redirctPhr, info.FoldedLength)(info.HistoryLength)
+        computeFoldedHist(redirectPhr, info.FoldedLength)(info.HistoryLength)
     }
     s0_foldedPhr := redirectData.foldedPhr
     when(redirectData.taken) {
-      s0_foldedPhr := redirectData.foldedPhr.update(VecInit(redirctPhr.asBools), redirectData.phrPtr, Shamt, shiftBits)
+      s0_foldedPhr := redirectData.foldedPhr.update(VecInit(redirectPhr.asBools), redirectData.phrPtr, Shamt, shiftBits)
     }
   }.elsewhen(s3_override) {
     s0_foldedPhr := s3_foldedPhrReg
@@ -161,8 +162,6 @@ class Phr(implicit p: Parameters) extends PhrModule with HasPhrParameters with H
     s0_foldedPhr := s0_foldedPhrReg
   }
 
-  private val phrValue = getPhr(phrPtr)
-
   AllFoldedHistoryInfo.foreach { info =>
     histFoldedPhr.getHistWithInfo(info).foldedHist :=
       computeFoldedHist(phrValue, info.FoldedLength)(info.HistoryLength)
@@ -180,7 +179,7 @@ class Phr(implicit p: Parameters) extends PhrModule with HasPhrParameters with H
       computeFoldedHist(predictHist, info.FoldedLength)(info.HistoryLength)
   }
 
-  io.phrPtr         := phrPtr
+  io.phrPtr         := s1_phrPtr
   io.phrs           := phr
   io.s0_foldedPhr   := s0_foldedPhr
   io.s1_foldedPhr   := s1_foldedPhrReg
@@ -249,9 +248,10 @@ class Phr(implicit p: Parameters) extends PhrModule with HasPhrParameters with H
   dontTouch(s0_foldedPhr)
   dontTouch(s1_foldedPhrReg)
   dontTouch(s2_foldedPhrReg)
+  dontTouch(s0_phrValue)
   dontTouch(phrValue)
   dontTouch(histFoldedPhr)
-  dontTouch(redirctPhr)
+  dontTouch(redirectPhr)
   dontTouch(diffFoldedPhr)
   dontTouch(s0_phrPtr)
   dontTouch(s1_phrPtr)
