diff --git a/scripts/parser.py b/scripts/parser.py
index 4aadb7cd029..a28828e94d0 100644
--- a/scripts/parser.py
+++ b/scripts/parser.py
@@ -58,9 +58,7 @@ def __init__(self, name):
 
     def add_line(self, line):
         debug_dontCare = False
-        if "RegFile" in self.name and "@(posedge clock)" in line:
-            line = line.replace("posedge", "negedge")
-        elif "RenameTable" in self.name:
+        if "RenameTable" in self.name:
             if line.strip().startswith("assign io_debug_rdata_"):
                 debug_dontCare = True
         elif "SynRegfileSlice" in self.name:
diff --git a/src/main/scala/xiangshan/backend/Backend.scala b/src/main/scala/xiangshan/backend/Backend.scala
index a4325bacce5..34a16728c32 100644
--- a/src/main/scala/xiangshan/backend/Backend.scala
+++ b/src/main/scala/xiangshan/backend/Backend.scala
@@ -235,6 +235,7 @@ class BackendImp(override val wrapper: Backend)(implicit p: Parameters) extends
   ctrlBlock.io.debugEnqLsq.resp := io.mem.lsqEnqIO.resp
   ctrlBlock.io.debugEnqLsq.req := memScheduler.io.memIO.get.lsqEnqIO.req
   ctrlBlock.io.debugEnqLsq.needAlloc := memScheduler.io.memIO.get.lsqEnqIO.needAlloc
+  ctrlBlock.io.debugEnqLsq.iqAccept := memScheduler.io.memIO.get.lsqEnqIO.iqAccept
 
   intScheduler.io.fromTop.hartId := io.fromTop.hartId
   intScheduler.io.fromCtrlBlock.flush := ctrlBlock.io.toIssueBlock.flush
diff --git a/src/main/scala/xiangshan/backend/issue/Dispatch2Iq.scala b/src/main/scala/xiangshan/backend/issue/Dispatch2Iq.scala
index c7a430bfa65..1c5df3da3ea 100644
--- a/src/main/scala/xiangshan/backend/issue/Dispatch2Iq.scala
+++ b/src/main/scala/xiangshan/backend/issue/Dispatch2Iq.scala
@@ -926,6 +926,7 @@ class Dispatch2IqMemImp(override val wrapper: Dispatch2Iq)(implicit p: Parameter
     }
     enqLsqIO.req(i).valid := io.in(i).fire && !isAMOVec(i) && !isSegment(i)
     enqLsqIO.req(i).bits := io.in(i).bits
+    enqLsqIO.iqAccept(i) := io.in(i).ready
 
     // This is to make it easier to calculate in LSQ.
     // Both scalar instructions and vector instructions with FLOW equal to 1 have a NUM value of 1.”
diff --git a/src/main/scala/xiangshan/mem/lsqueue/LSQWrapper.scala b/src/main/scala/xiangshan/mem/lsqueue/LSQWrapper.scala
index 20e56b2ccc0..c43c2ffe7ff 100644
--- a/src/main/scala/xiangshan/mem/lsqueue/LSQWrapper.scala
+++ b/src/main/scala/xiangshan/mem/lsqueue/LSQWrapper.scala
@@ -30,6 +30,7 @@ import xiangshan.mem._
 import xiangshan.backend._
 import xiangshan.backend.rob.RobLsqIO
 import coupledL2.{CMOReq, CMOResp}
+import xiangshan.backend.fu.FuType
 
 class ExceptionAddrIO(implicit p: Parameters) extends XSBundle {
   val isStore = Input(Bool())
@@ -55,6 +56,7 @@ class LsqEnqIO(implicit p: Parameters) extends MemBlockBundle {
   val canAccept = Output(Bool())
   val needAlloc = Vec(LSQEnqWidth, Input(UInt(2.W)))
   val req       = Vec(LSQEnqWidth, Flipped(ValidIO(new DynInst)))
+  val iqAccept  = Input(Vec(LSQEnqWidth, Bool()))
   val resp      = Vec(LSQEnqWidth, Output(new LSIdx))
 }
 
@@ -311,23 +313,20 @@ class LsqEnqCtrl(implicit p: Parameters) extends XSModule
   val sqCounter = RegInit(StoreQueueSize.U(log2Up(StoreQueueSize + 1).W))
   val canAccept = RegInit(false.B)
 
-  val loadEnqVec = io.enq.req.zip(io.enq.needAlloc).map(x => x._1.valid && x._2(0))
-  val storeEnqVec = io.enq.req.zip(io.enq.needAlloc).map(x => x._1.valid && x._2(1))
-  val isLastUopVec = io.enq.req.map(_.bits.lastUop)
-  val vLoadFlow = io.enq.req.map(_.bits.numLsElem)
-  val vStoreFlow = io.enq.req.map(_.bits.numLsElem)
-  val validVLoadFlow = vLoadFlow.zipWithIndex.map{case (vLoadFlowNumItem, index) => Mux(loadEnqVec(index), vLoadFlowNumItem, 0.U)}
-  val validVStoreFlow = vStoreFlow.zipWithIndex.map{case (vStoreFlowNumItem, index) => Mux(storeEnqVec(index), vStoreFlowNumItem, 0.U)}
-  val enqVLoadOffsetNumber = validVLoadFlow.reduce(_ + _)
-  val enqVStoreOffsetNumber = validVStoreFlow.reduce(_ + _)
-  val validVLoadOffset = 0.U +: vLoadFlow.zip(io.enq.needAlloc)
-                                .map{case (flow, needAllocItem) => Mux(needAllocItem(0).asBool, flow, 0.U)}
-                                .slice(0, validVLoadFlow.length - 1)
-  val validVStoreOffset = 0.U +: vStoreFlow.zip(io.enq.needAlloc)
-                                .map{case (flow, needAllocItem) => Mux(needAllocItem(1).asBool, flow, 0.U)}
-                                .slice(0, validVStoreFlow.length - 1)
-  val lqAllocNumber = enqVLoadOffsetNumber
-  val sqAllocNumber = enqVStoreOffsetNumber
+  val blockVec = io.enq.iqAccept.map(!_) :+ true.B
+  val numLsElem = io.enq.req.map(_.bits.numLsElem)
+  val needEnqLoadQueue = VecInit(io.enq.req.map(x => FuType.isLoad(x.bits.fuType) || FuType.isVNonsegLoad(x.bits.fuType)))
+  val needEnqStoreQueue = VecInit(io.enq.req.map(x => FuType.isStore(x.bits.fuType) || FuType.isVNonsegStore(x.bits.fuType)))
+  val loadQueueElem = needEnqLoadQueue.zip(numLsElem).map(x => Mux(x._1, x._2, 0.U))
+  val storeQueueElem = needEnqStoreQueue.zip(numLsElem).map(x => Mux(x._1, x._2, 0.U))
+  val loadFlowPopCount = 0.U +: loadQueueElem.zipWithIndex.map{ case (l, i) =>
+    loadQueueElem.take(i + 1).reduce(_ + _)
+  }
+  val storeFlowPopCount = 0.U +: storeQueueElem.zipWithIndex.map { case (s, i) =>
+    storeQueueElem.take(i + 1).reduce(_ + _)
+  }
+  val lqAllocNumber = PriorityMux(blockVec.zip(loadFlowPopCount))
+  val sqAllocNumber = PriorityMux(blockVec.zip(storeFlowPopCount))
 
   io.lqFreeCount  := lqCounter
   io.sqFreeCount  := sqCounter
@@ -370,13 +369,14 @@ class LsqEnqCtrl(implicit p: Parameters) extends XSModule
   val lqOffset = Wire(Vec(io.enq.resp.length, UInt(lqPtr.value.getWidth.W)))
   val sqOffset = Wire(Vec(io.enq.resp.length, UInt(sqPtr.value.getWidth.W)))
   for ((resp, i) <- io.enq.resp.zipWithIndex) {
-    lqOffset(i) := validVLoadOffset.take(i + 1).reduce(_ + _)
+    lqOffset(i) := loadFlowPopCount(i)
     resp.lqIdx := lqPtr + lqOffset(i)
-    sqOffset(i) := validVStoreOffset.take(i + 1).reduce(_ + _)
+    sqOffset(i) := storeFlowPopCount(i)
     resp.sqIdx := sqPtr + sqOffset(i)
   }
 
   io.enqLsq.needAlloc := RegNext(io.enq.needAlloc)
+  io.enqLsq.iqAccept := RegNext(io.enq.iqAccept)
   io.enqLsq.req.zip(io.enq.req).zip(io.enq.resp).foreach{ case ((toLsq, enq), resp) =>
     val do_enq = enq.valid && !io.redirect.valid && io.enq.canAccept
     toLsq.valid := RegNext(do_enq)
