diff --git a/src/main/scala/top/Configs.scala b/src/main/scala/top/Configs.scala
index 0bb999d8bf6..a1cc2e0445a 100644
--- a/src/main/scala/top/Configs.scala
+++ b/src/main/scala/top/Configs.scala
@@ -212,7 +212,7 @@ class MinimalConfig(n: Int = 1) extends Config(
           )),
         )),
         L2NBanks = 2,
-        prefetcher = None // if L2 pf_recv_node does not exist, disable SMS prefetcher
+        prefetcher = Nil // if L2 pf_recv_node does not exist, disable SMS prefetcher
       )
     )
     case SoCParamsKey =>
diff --git a/src/main/scala/xiangshan/Parameters.scala b/src/main/scala/xiangshan/Parameters.scala
index 60782f17a83..d0f259f18aa 100644
--- a/src/main/scala/xiangshan/Parameters.scala
+++ b/src/main/scala/xiangshan/Parameters.scala
@@ -26,7 +26,7 @@ import xiangshan.backend.datapath.RdConfig._
 import xiangshan.backend.datapath.WbConfig._
 import xiangshan.backend.exu.ExeUnitParams
 import xiangshan.backend.fu.FuConfig._
-import xiangshan.backend.issue.{IntScheduler, IssueBlockParams, MemScheduler, SchdBlockParams, SchedulerType, VfScheduler, FpScheduler}
+import xiangshan.backend.issue.{FpScheduler, IntScheduler, IssueBlockParams, MemScheduler, SchdBlockParams, SchedulerType, VfScheduler}
 import xiangshan.backend.regfile._
 import xiangshan.backend.BackendParams
 import xiangshan.backend.trace._
@@ -44,7 +44,7 @@ import xiangshan.cache.wpu.WPUParameters
 import coupledL2._
 import coupledL2.tl2chi._
 import xiangshan.backend.datapath.WakeUpConfig
-import xiangshan.mem.prefetch.{PrefetcherParams, SMSParams}
+import xiangshan.mem.prefetch.{PrefetcherParams, SMSParams, StreamStrideParams, TLBPlace}
 
 import scala.math.{max, min, pow}
 
@@ -188,7 +188,7 @@ case class XSCoreParameters
   MemRegCacheSize: Int = 12,
   intSchdVlWbPort: Int = 0,
   vfSchdVlWbPort: Int = 1,
-  prefetcher: Option[PrefetcherParams] = Some(SMSParams()),
+  prefetcher: Seq[PrefetcherParams] = Seq(StreamStrideParams(), SMSParams()),
   IfuRedirectNum: Int = 1,
   LoadPipelineWidth: Int = 3,
   StorePipelineWidth: Int = 2,
@@ -577,6 +577,13 @@ trait HasXSParameter {
 
   def HasBitmapCheck = coreParams.HasBitmapCheck
   def HasBitmapCheckDefault = coreParams.HasBitmapCheckDefault
+  
+  /** prefetch config */
+  def prefetcherSeq = coreParams.prefetcher
+  def prefetcherNum = max(prefetcherSeq.size, 1) //TODO lyq: 1 for simpler code generation, but it's also ugly
+  def PfNumInDtlbLD = prefetcherSeq.count(_.tlbPlace == TLBPlace.dtlb_ld)
+  def PfNumInDtlbPF = prefetcherSeq.count(_.tlbPlace == TLBPlace.dtlb_pf) + 1 // 1 for l2 prefetch
+
   def HasMExtension = coreParams.HasMExtension
   def HasCExtension = coreParams.HasCExtension
   def HasHExtension = coreParams.HasHExtension
diff --git a/src/main/scala/xiangshan/cache/mmu/Repeater.scala b/src/main/scala/xiangshan/cache/mmu/Repeater.scala
index e687c8b0c9b..4d2e86fd93f 100644
--- a/src/main/scala/xiangshan/cache/mmu/Repeater.scala
+++ b/src/main/scala/xiangshan/cache/mmu/Repeater.scala
@@ -206,34 +206,15 @@ class PTWFilterEntry(Width: Int, Size: Int, hasHint: Boolean = false)(implicit p
   io.memidx := 0.U.asTypeOf(new MemBlockidxBundle)
   io.getGpa := 0.U
 
-  // ugly code, should be optimized later
-  require(Width <= 4, s"DTLB Filter Width ($Width) must equal or less than 4")
-  if (Width == 1) {
-    require(Size == 8, s"prefetch filter Size ($Size) should be 8")
-    canenq(0) := !(Cat(v).andR)
-    enqidx(0) := firstValidIndex(v, false.B)
-  } else if (Width == 2) {
-    require(Size == 8, s"store filter Size ($Size) should be 8")
-    canenq(0) := !(Cat(v.take(Size/2)).andR)
-    enqidx(0) := firstValidIndex(v.take(Size/2), false.B)
-    canenq(1) := !(Cat(v.drop(Size/2)).andR)
-    enqidx(1) := firstValidIndex(v.drop(Size/2), false.B) + (Size/2).U
-  } else if (Width == 3) {
-    require(Size == 16, s"load filter Size ($Size) should be 16")
-    canenq(0) := !(Cat(v.take(8)).andR)
-    enqidx(0) := firstValidIndex(v.take(8), false.B)
-    canenq(1) := !(Cat(v.drop(8).take(4)).andR)
-    enqidx(1) := firstValidIndex(v.drop(8).take(4), false.B) + 8.U
-    // four entries for prefetch
-    canenq(2) := !(Cat(v.drop(12)).andR)
-    enqidx(2) := firstValidIndex(v.drop(12), false.B) + 12.U
-  } else if (Width == 4) {
-    require(Size == 16, s"load filter Size ($Size) should be 16")
-    for (i <- 0 until Width) {
-      canenq(i) := !(Cat(VecInit(v.slice(i * 4, (i + 1) * 4))).andR)
-      enqidx(i) := firstValidIndex(v.slice(i * 4, (i + 1) * 4), false.B) + (i * 4).U
-    }
+  require(Width > 0 && Width <= 4, s"DTLB Filter Width ($Width) must equal or less than 4")
+  require(Size >= Width && ((Size & (Size-1)) == 0), s"DTLB Filter Size ($Size) must be bigger than Width ($Width) and a power of 2")
+  val step = Size / Width
+  for(i <- 0 until (Width - 1)) {
+    canenq(i) := !(Cat(VecInit(v.slice(i * step, (i + 1) * step))).andR)
+    enqidx(i) := firstValidIndex(v.slice(i * step, (i + 1) * step), false.B) + (i * step).U
   }
+  canenq(Width-1) := !(Cat(VecInit(v.slice((Width - 1) * step, Size))).andR)
+  enqidx(Width-1) := firstValidIndex(v.slice((Width - 1) * step, Size), false.B) + ((Width - 1) * step).U
 
   for (i <- 0 until Width) {
     enqvalid(i) := io.tlb.req(i).valid && !ptwResp_ReqMatchVec(i) && !entryIsMatchVec(i) && canenq(i)
@@ -349,7 +330,7 @@ class PTWNewFilter(Width: Int, Size: Int, FenceDelay: Int)(implicit p: Parameter
   val io = IO(new PTWFilterIO(Width, hasHint = true))
 
   val load_filter = VecInit(Seq.fill(1) {
-    val load_entry = Module(new PTWFilterEntry(Width = LdExuCnt + 1, Size = loadfiltersize, hasHint = true))
+    val load_entry = Module(new PTWFilterEntry(Width = LdExuCnt + PfNumInDtlbLD, Size = loadfiltersize, hasHint = true))
     load_entry.io
   })
 
@@ -359,15 +340,15 @@ class PTWNewFilter(Width: Int, Size: Int, FenceDelay: Int)(implicit p: Parameter
   })
 
   val prefetch_filter = VecInit(Seq.fill(1) {
-    val prefetch_entry = Module(new PTWFilterEntry(Width = 2, Size = prefetchfiltersize))
+    val prefetch_entry = Module(new PTWFilterEntry(Width = PfNumInDtlbPF, Size = prefetchfiltersize))
     prefetch_entry.io
   })
 
   val filter = load_filter ++ store_filter ++ prefetch_filter
 
-  load_filter.map(_.tlb.req := io.tlb.req.take(LdExuCnt + 1))
-  store_filter.map(_.tlb.req := io.tlb.req.drop(LdExuCnt + 1).take(StaCnt))
-  prefetch_filter.map(_.tlb.req := io.tlb.req.drop(LdExuCnt + 1 + StaCnt))
+  load_filter.map(_.tlb.req := io.tlb.req.take(LdExuCnt + PfNumInDtlbLD))
+  store_filter.map(_.tlb.req := io.tlb.req.drop(LdExuCnt + PfNumInDtlbLD).take(StaCnt))
+  prefetch_filter.map(_.tlb.req := io.tlb.req.drop(LdExuCnt + PfNumInDtlbLD + StaCnt))
 
   val flush = DelayN(io.sfence.valid || io.csr.satp.changed || (io.csr.priv.virt && io.csr.vsatp.changed), FenceDelay)
   val ptwResp = RegEnable(io.ptw.resp.bits, io.ptw.resp.fire)
@@ -395,11 +376,11 @@ class PTWNewFilter(Width: Int, Size: Int, FenceDelay: Int)(implicit p: Parameter
     io.tlb.resp.bits.getGpa(i) := false.B
   }
   io.tlb.resp.bits.vector(0) := load_filter(0).refill
-  io.tlb.resp.bits.vector(LdExuCnt + 1) := store_filter(0).refill
-  io.tlb.resp.bits.vector(LdExuCnt + 1 + StaCnt) := prefetch_filter(0).refill
+  io.tlb.resp.bits.vector(LdExuCnt + PfNumInDtlbLD) := store_filter(0).refill
+  io.tlb.resp.bits.vector(LdExuCnt + PfNumInDtlbLD + StaCnt) := prefetch_filter(0).refill
   io.tlb.resp.bits.getGpa(0) := load_filter(0).getGpa
-  io.tlb.resp.bits.getGpa(LdExuCnt + 1) := store_filter(0).getGpa
-  io.tlb.resp.bits.getGpa(LdExuCnt + 1 + StaCnt) := prefetch_filter(0).getGpa
+  io.tlb.resp.bits.getGpa(LdExuCnt + PfNumInDtlbLD) := store_filter(0).getGpa
+  io.tlb.resp.bits.getGpa(LdExuCnt + PfNumInDtlbLD + StaCnt) := prefetch_filter(0).getGpa
 
   val hintIO = io.hint.getOrElse(new TlbHintIO)
   val load_hintIO = load_filter(0).hint.getOrElse(new TlbHintIO)
@@ -414,11 +395,11 @@ class PTWNewFilter(Width: Int, Size: Int, FenceDelay: Int)(implicit p: Parameter
     io.tlb.resp.bits.data.memidx := load_filter(0).memidx
   }
   when (store_filter(0).refill) {
-    io.tlb.resp.bits.vector(LdExuCnt + 1) := true.B
+    io.tlb.resp.bits.vector(LdExuCnt + PfNumInDtlbLD) := true.B
     io.tlb.resp.bits.data.memidx := store_filter(0).memidx
   }
   when (prefetch_filter(0).refill) {
-    io.tlb.resp.bits.vector(LdExuCnt + 1 + StaCnt) := true.B
+    io.tlb.resp.bits.vector(LdExuCnt + PfNumInDtlbLD + StaCnt) := true.B
     io.tlb.resp.bits.data.memidx := 0.U.asTypeOf(new MemBlockidxBundle)
   }
 
diff --git a/src/main/scala/xiangshan/mem/Bundles.scala b/src/main/scala/xiangshan/mem/Bundles.scala
index 19bee84a186..8186d16a36f 100644
--- a/src/main/scala/xiangshan/mem/Bundles.scala
+++ b/src/main/scala/xiangshan/mem/Bundles.scala
@@ -123,6 +123,7 @@ object Bundles {
   class LsPrefetchTrainBundle(implicit p: Parameters) extends LsPipelineBundle {
     val meta_prefetch = UInt(L1PfSourceBits.W)
     val meta_access = Bool()
+    val is_from_hw_pf = Bool() // s0 source is from prefetch
 
     def fromLsPipelineBundle(input: LsPipelineBundle, latch: Boolean = false, enable: Bool = true.B) = {
       val inputReg = latch match {
@@ -132,6 +133,7 @@ object Bundles {
       connectSamePort(this, inputReg)
       this.meta_prefetch := DontCare
       this.meta_access := DontCare
+      this.is_from_hw_pf := DontCare
     }
 
     def toPrefetchReqBundle(): PrefetchReqBundle = {
diff --git a/src/main/scala/xiangshan/mem/MemBlock.scala b/src/main/scala/xiangshan/mem/MemBlock.scala
index 99cea05c7c0..d8ea8f14bfd 100644
--- a/src/main/scala/xiangshan/mem/MemBlock.scala
+++ b/src/main/scala/xiangshan/mem/MemBlock.scala
@@ -36,17 +36,17 @@ import xiangshan.backend.ctrlblock.{DebugLSIO, LsTopdownInfo}
 import xiangshan.backend.exu.MemExeUnit
 import xiangshan.backend.fu._
 import xiangshan.backend.fu.FuType._
-import xiangshan.backend.fu.NewCSR.{CsrTriggerBundle, TriggerUtil, PFEvent}
+import xiangshan.backend.fu.NewCSR.{CsrTriggerBundle, PFEvent, TriggerUtil}
 import xiangshan.backend.fu.util.{CSRConst, SdtrigExt}
 import xiangshan.backend.{BackendToTopBundle, TopToBackendBundle}
-import xiangshan.backend.rob.{RobDebugRollingIO, RobPtr, RobLsqIO}
+import xiangshan.backend.rob.{RobDebugRollingIO, RobLsqIO, RobPtr}
 import xiangshan.backend.datapath.NewPipelineConnect
 import xiangshan.backend.trace.{Itype, TraceCoreInterface}
 import xiangshan.backend.Bundles._
 import xiangshan.mem._
 import xiangshan.mem.mdp._
 import xiangshan.mem.Bundles._
-import xiangshan.mem.prefetch.{BasePrefecher, L1Prefetcher, SMSParams, SMSPrefetcher}
+import xiangshan.mem.prefetch.{BasePrefecher, L1Prefetcher, PrefetcherWrapper, SMSParams, SMSPrefetcher, TLBPlace}
 import xiangshan.cache._
 import xiangshan.cache.mmu._
 import coupledL2.PrefetchRecv
@@ -265,12 +265,11 @@ class MemBlockInlined()(implicit p: Parameters) extends LazyModule
   val ptw_to_l2_buffer = if (!coreParams.softPTW) LazyModule(new TLBuffer) else null
   val l1d_to_l2_buffer = if (coreParams.dcacheParametersOpt.nonEmpty) LazyModule(new TLBuffer) else null
   val dcache_port = TLNameNode("dcache_client") // to keep dcache-L2 port name
-  val l2_pf_sender_opt = coreParams.prefetcher.map(_ =>
-    BundleBridgeSource(() => new PrefetchRecv)
-  )
-  val l3_pf_sender_opt = if (p(SoCParamsKey).L3CacheParamsOpt.nonEmpty) coreParams.prefetcher.map(_ =>
-    BundleBridgeSource(() => new huancun.PrefetchRecv)
-  ) else None
+  // NOTE: we currently only use one output port to L2 and L3 prefetch sender respectively
+  val l2_pf_sender_opt = if (coreParams.prefetcher.nonEmpty) 
+    Some(BundleBridgeSource(() => new PrefetchRecv)) else None
+  val l3_pf_sender_opt = if (p(SoCParamsKey).L3CacheParamsOpt.nonEmpty && coreParams.prefetcher.nonEmpty)
+    Some(BundleBridgeSource(() => new huancun.PrefetchRecv)) else None
   val frontendBridge = LazyModule(new FrontendBridge)
   // interrupt sinks
   val clint_int_sink = IntSinkNode(IntSinkPortSimple(1, 2))
@@ -393,6 +392,7 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
   dontTouch(io.inner_hc_perfEvents)
   dontTouch(io.outer_hc_perfEvents)
 
+  val hartId = p(XSCoreParamsKey).HartId
   val redirect = RegNextWithEnable(io.redirect)
 
   private val dcache = outer.dcache.module
@@ -429,74 +429,6 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
   val loadMisalignBuffer = Module(new LoadMisalignBuffer)
   val storeMisalignBuffer = Module(new StoreMisalignBuffer)
 
-  val l1_pf_req = Wire(Decoupled(new L1PrefetchReq()))
-  dcache.io.sms_agt_evict_req.ready := false.B
-  val l1D_pf_enable = GatedRegNextN(io.ooo_to_mem.csrCtrl.pf_ctrl.l1D_pf_enable, 2, Some(false.B))
-  val prefetcherOpt: Option[BasePrefecher] = coreParams.prefetcher.map {
-    case _: SMSParams =>
-      val sms = Module(new SMSPrefetcher())
-      val enableSMS = Constantin.createRecord(s"enableSMS$hartId", initValue = true)
-      // constantinCtrl && master switch csrCtrl && single switch csrCtrl
-      sms.io.enable := enableSMS && l1D_pf_enable &&
-        GatedRegNextN(io.ooo_to_mem.csrCtrl.pf_ctrl.l2_pf_recv_enable, 2, Some(false.B))
-      sms.io_agt_en := GatedRegNextN(io.ooo_to_mem.csrCtrl.pf_ctrl.l1D_pf_enable_agt, 2, Some(false.B))
-      sms.io_pht_en := GatedRegNextN(io.ooo_to_mem.csrCtrl.pf_ctrl.l1D_pf_enable_pht, 2, Some(false.B))
-      sms.io_act_threshold := GatedRegNextN(io.ooo_to_mem.csrCtrl.pf_ctrl.l1D_pf_active_threshold, 2, Some(12.U))
-      sms.io_act_stride := GatedRegNextN(io.ooo_to_mem.csrCtrl.pf_ctrl.l1D_pf_active_stride, 2, Some(30.U))
-      sms.io_stride_en := false.B
-      sms.io_dcache_evict <> dcache.io.sms_agt_evict_req
-      val mbistSmsPl = MbistPipeline.PlaceMbistPipeline(1, "MbistPipeSms", hasMbist)
-      sms
-  }
-  prefetcherOpt.foreach{ pf => pf.io.l1_req.ready := false.B }
-  val hartId = p(XSCoreParamsKey).HartId
-  val l1PrefetcherOpt: Option[BasePrefecher] = coreParams.prefetcher.map {
-    case _ =>
-      val l1Prefetcher = Module(new L1Prefetcher())
-      val enableL1StreamPrefetcher = Constantin.createRecord(s"enableL1StreamPrefetcher$hartId", initValue = true)
-      // constantinCtrl && master switch csrCtrl && single switch csrCtrl
-      l1Prefetcher.io.enable := enableL1StreamPrefetcher && l1D_pf_enable &&
-        GatedRegNextN(io.ooo_to_mem.csrCtrl.pf_ctrl.l1D_pf_enable_stride, 2, Some(false.B))
-      l1Prefetcher.pf_ctrl <> dcache.io.pf_ctrl
-      l1Prefetcher.l2PfqBusy := io.l2PfqBusy
-
-      // stride will train on miss or prefetch hit
-      for (i <- 0 until LduCnt) {
-        val source = loadUnits(i).io.prefetch_train_l1
-        l1Prefetcher.stride_train(i).valid := source.valid && source.bits.isFirstIssue && (
-          source.bits.miss || isFromStride(source.bits.meta_prefetch)
-        )
-        l1Prefetcher.stride_train(i).bits := source.bits
-        val loadPc = RegNext(io.ooo_to_mem.issueLda(i).bits.uop.pc) // for s1
-        l1Prefetcher.stride_train(i).bits.uop.pc := Mux(
-          loadUnits(i).io.s2_ptr_chasing,
-          RegEnable(loadPc, loadUnits(i).io.s2_prefetch_spec),
-          RegEnable(RegEnable(loadPc, loadUnits(i).io.s1_prefetch_spec), loadUnits(i).io.s2_prefetch_spec)
-        )
-      }
-      for (i <- 0 until HyuCnt) {
-        val source = hybridUnits(i).io.prefetch_train_l1
-        l1Prefetcher.stride_train.drop(LduCnt)(i).valid := source.valid && source.bits.isFirstIssue && (
-          source.bits.miss || isFromStride(source.bits.meta_prefetch)
-        )
-        l1Prefetcher.stride_train.drop(LduCnt)(i).bits := source.bits
-        l1Prefetcher.stride_train.drop(LduCnt)(i).bits.uop.pc := Mux(
-          hybridUnits(i).io.ldu_io.s2_ptr_chasing,
-          RegNext(io.ooo_to_mem.hybridPc(i)),
-          RegNext(RegNext(io.ooo_to_mem.hybridPc(i)))
-        )
-      }
-      l1Prefetcher
-  }
-  // load prefetch to l1 Dcache
-  l1PrefetcherOpt match {
-    case Some(pf) => l1_pf_req <> Pipeline(in = pf.io.l1_req, depth = 1, pipe = false, name = Some("pf_queue_to_ldu_reg"))
-    case None =>
-      l1_pf_req.valid := false.B
-      l1_pf_req.bits := DontCare
-  }
-  val pf_train_on_hit = RegNextN(io.ooo_to_mem.csrCtrl.pf_ctrl.l1D_pf_train_on_hit, 2, Some(true.B))
-
   loadUnits.zipWithIndex.map(x => x._1.suggestName("LoadUnit_"+x._2))
   storeUnits.zipWithIndex.map(x => x._1.suggestName("StoreUnit_"+x._2))
   hybridUnits.zipWithIndex.map(x => x._1.suggestName("HybridUnit_"+x._2))
@@ -549,62 +481,6 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
   io.mem_to_ooo.otherFastWakeup.take(HyuCnt).zip(hybridUnits.map(_.io.ldu_io.fast_uop)).foreach{case(a,b)=> a:=b}
   val stOut = io.mem_to_ooo.writebackSta ++ io.mem_to_ooo.writebackHyuSta
 
-  // prefetch to l1 req
-  // Stream's confidence is always 1
-  // (LduCnt + HyuCnt) l1_pf_reqs ?
-  loadUnits.foreach(load_unit => {
-    load_unit.io.prefetch_req.valid <> l1_pf_req.valid
-    load_unit.io.prefetch_req.bits <> l1_pf_req.bits
-  })
-
-  hybridUnits.foreach(hybrid_unit => {
-    hybrid_unit.io.ldu_io.prefetch_req.valid <> l1_pf_req.valid
-    hybrid_unit.io.ldu_io.prefetch_req.bits <> l1_pf_req.bits
-  })
-
-  // NOTE: loadUnits(0) has higher bank conflict and miss queue arb priority than loadUnits(1) and loadUnits(2)
-  // when loadUnits(1)/loadUnits(2) stage 0 is busy, hw prefetch will never use that pipeline
-  val LowConfPorts = if (LduCnt == 2) Seq(1) else if (LduCnt == 3) Seq(1, 2) else Seq(0)
-  LowConfPorts.map{case i => loadUnits(i).io.prefetch_req.bits.confidence := 0.U}
-  hybridUnits.foreach(hybrid_unit => { hybrid_unit.io.ldu_io.prefetch_req.bits.confidence := 0.U })
-
-  val canAcceptHighConfPrefetch = loadUnits.map(_.io.canAcceptHighConfPrefetch) ++
-                                  hybridUnits.map(_.io.canAcceptLowConfPrefetch)
-  val canAcceptLowConfPrefetch = loadUnits.map(_.io.canAcceptLowConfPrefetch) ++
-                                 hybridUnits.map(_.io.canAcceptLowConfPrefetch)
-  l1_pf_req.ready := (0 until LduCnt + HyuCnt).map{
-    case i => {
-      if (LowConfPorts.contains(i)) {
-        loadUnits(i).io.canAcceptLowConfPrefetch
-      } else {
-        Mux(l1_pf_req.bits.confidence === 1.U, canAcceptHighConfPrefetch(i), canAcceptLowConfPrefetch(i))
-      }
-    }
-  }.reduce(_ || _)
-
-  // l1 pf fuzzer interface
-  val DebugEnableL1PFFuzzer = false
-  if (DebugEnableL1PFFuzzer) {
-    // l1 pf req fuzzer
-    val fuzzer = Module(new L1PrefetchFuzzer())
-    fuzzer.io.vaddr := DontCare
-    fuzzer.io.paddr := DontCare
-
-    // override load_unit prefetch_req
-    loadUnits.foreach(load_unit => {
-      load_unit.io.prefetch_req.valid <> fuzzer.io.req.valid
-      load_unit.io.prefetch_req.bits <> fuzzer.io.req.bits
-    })
-
-    // override hybrid_unit prefetch_req
-    hybridUnits.foreach(hybrid_unit => {
-      hybrid_unit.io.ldu_io.prefetch_req.valid <> fuzzer.io.req.valid
-      hybrid_unit.io.ldu_io.prefetch_req.bits <> fuzzer.io.req.bits
-    })
-
-    fuzzer.io.req.ready := l1_pf_req.ready
-  }
-
   // TODO: fast load wakeup
   val lsq     = Module(new LsqWrapper)
   val sbuffer = Module(new Sbuffer)
@@ -621,41 +497,6 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
   dcache.io.wfi.wfiReq := io.wfi.wfiReq
   lsq.io.wfi.wfiReq := io.wfi.wfiReq
 
-  // load/store prefetch to l2 cache
-  prefetcherOpt.foreach(sms_pf => {
-    l1PrefetcherOpt.foreach(l1_pf => {
-      val sms_pf_to_l2 = DelayNWithValid(sms_pf.io.l2_req, 2)
-      val l1_pf_to_l2 = DelayNWithValid(l1_pf.io.l2_req, 2)
-
-      outer.l2_pf_sender_opt.get.out.head._1.addr_valid := sms_pf_to_l2.valid || l1_pf_to_l2.valid
-      outer.l2_pf_sender_opt.get.out.head._1.addr := Mux(l1_pf_to_l2.valid, l1_pf_to_l2.bits.addr, sms_pf_to_l2.bits.addr)
-      outer.l2_pf_sender_opt.get.out.head._1.pf_source := Mux(l1_pf_to_l2.valid, l1_pf_to_l2.bits.source, sms_pf_to_l2.bits.source)
-      outer.l2_pf_sender_opt.get.out.head._1.l2_pf_en := RegNextN(io.ooo_to_mem.csrCtrl.pf_ctrl.l2_pf_enable, 2, Some(true.B))
-
-      val l2_trace = Wire(new LoadPfDbBundle)
-      l2_trace.paddr := outer.l2_pf_sender_opt.get.out.head._1.addr
-      val table = ChiselDB.createTable(s"L2PrefetchTrace$hartId", new LoadPfDbBundle, basicDB = false)
-      table.log(l2_trace, l1_pf_to_l2.valid, "StreamPrefetchTrace", clock, reset)
-      table.log(l2_trace, !l1_pf_to_l2.valid && sms_pf_to_l2.valid, "L2PrefetchTrace", clock, reset)
-
-      val l1_pf_to_l3 = ValidIODelay(l1_pf.io.l3_req, 4)
-      outer.l3_pf_sender_opt.foreach(_.out.head._1.addr_valid := l1_pf_to_l3.valid)
-      outer.l3_pf_sender_opt.foreach(_.out.head._1.addr := l1_pf_to_l3.bits)
-      outer.l3_pf_sender_opt.foreach(_.out.head._1.l2_pf_en := RegNextN(io.ooo_to_mem.csrCtrl.pf_ctrl.l2_pf_enable, 4, Some(true.B)))
-
-      val l3_trace = Wire(new LoadPfDbBundle)
-      l3_trace.paddr := outer.l3_pf_sender_opt.map(_.out.head._1.addr).getOrElse(0.U)
-      val l3_table = ChiselDB.createTable(s"L3PrefetchTrace$hartId", new LoadPfDbBundle, basicDB = false)
-      l3_table.log(l3_trace, l1_pf_to_l3.valid, "StreamPrefetchTrace", clock, reset)
-
-      XSPerfAccumulate("prefetch_fire_l2", outer.l2_pf_sender_opt.get.out.head._1.addr_valid)
-      XSPerfAccumulate("prefetch_fire_l3", outer.l3_pf_sender_opt.map(_.out.head._1.addr_valid).getOrElse(false.B))
-      XSPerfAccumulate("l1pf_fire_l2", l1_pf_to_l2.valid)
-      XSPerfAccumulate("sms_fire_l2", !l1_pf_to_l2.valid && sms_pf_to_l2.valid)
-      XSPerfAccumulate("sms_block_by_l1pf", l1_pf_to_l2.valid && sms_pf_to_l2.valid)
-    })
-  })
-
   // ptw
   val sfence = RegNext(RegNext(io.ooo_to_mem.sfence))
   val tlbcsr = RegNext(RegNext(io.ooo_to_mem.tlbCsr))
@@ -676,20 +517,45 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
     Seq()
   }
 
-  // dtlb
-  val dtlb_ld_tlb_ld = Module(new TLBNonBlock(LduCnt + HyuCnt + 1, 2, ldtlbParams))
-  val dtlb_st_tlb_st = Module(new TLBNonBlock(StaCnt, 1, sttlbParams))
-  val dtlb_prefetch_tlb_prefetch = Module(new TLBNonBlock(2, 2, pftlbParams))
-  val dtlb_ld = Seq(dtlb_ld_tlb_ld.io)
-  val dtlb_st = Seq(dtlb_st_tlb_st.io)
-  val dtlb_prefetch = Seq(dtlb_prefetch_tlb_prefetch.io)
-  /* tlb vec && constant variable */
-  val dtlb = dtlb_ld ++ dtlb_st ++ dtlb_prefetch
+  // dtlb parameters
+  var pf2tlbIndexMap: Seq[Int] = Seq()
+  var tmp_pf_num_in_dtlb_ld: Int = 0
+  var tmp_pf_num_in_dtlb_pf: Int = 1 // 1 for l2 prefetch
+  prefetcherSeq foreach { x =>
+    if(x.tlbPlace == TLBPlace.dtlb_ld){
+      pf2tlbIndexMap = pf2tlbIndexMap :+ (LduCnt + HyuCnt + tmp_pf_num_in_dtlb_ld)
+      tmp_pf_num_in_dtlb_ld += 1
+    }else if(x.tlbPlace == TLBPlace.dtlb_pf){
+      // 1 for l2 prefetch
+      pf2tlbIndexMap = pf2tlbIndexMap :+ (LduCnt + HyuCnt + PfNumInDtlbLD + StaCnt + tmp_pf_num_in_dtlb_pf)
+      tmp_pf_num_in_dtlb_pf += 1
+    }
+  }
   val (dtlb_ld_idx, dtlb_st_idx, dtlb_pf_idx) = (0, 1, 2)
-  val TlbSubSizeVec = Seq(LduCnt + HyuCnt + 1, StaCnt, 2) // (load + hyu + stream pf, store, sms+l2bop)
+  val TlbSubSizeVec = Seq(LduCnt + HyuCnt + PfNumInDtlbLD, StaCnt, PfNumInDtlbPF) // (load + hyu + stream pf, store, sms+l2bop)
   val DTlbSize = TlbSubSizeVec.sum
   val TlbStartVec = TlbSubSizeVec.scanLeft(0)(_ + _).dropRight(1)
   val TlbEndVec = TlbSubSizeVec.scanLeft(0)(_ + _).drop(1)
+  val L2toL1DTLBPortIndex = TlbStartVec(dtlb_pf_idx)
+  println(f"TLB Size:")
+  println(f"  size = $DTlbSize = ${TlbSubSizeVec}")
+  println(f"TLB Index Vec:")
+  println(f"  TlbStartVec = ${TlbStartVec}")
+  println(f"  TlbEndVec = ${TlbEndVec}")
+  println(f"TLB for Prefetcher:")
+  prefetcherSeq zip pf2tlbIndexMap foreach { case (pf, idx) =>
+    println(f"  TLB #$idx%-2d => ${pf.name}")
+  }
+  println(f"  TLB #$L2toL1DTLBPortIndex%-2d => L2Prefetcher")
+
+  // dtlb instantiation
+  val dtlb_ld_tlb_ld = Module(new TLBNonBlock(TlbSubSizeVec(dtlb_ld_idx), 2, ldtlbParams))
+  val dtlb_st_tlb_st = Module(new TLBNonBlock(TlbSubSizeVec(dtlb_st_idx), 1, sttlbParams))
+  val dtlb_prefetch_tlb_prefetch = Module(new TLBNonBlock(TlbSubSizeVec(dtlb_pf_idx), 2, pftlbParams))
+  val dtlb_ld = Seq(dtlb_ld_tlb_ld.io)
+  val dtlb_st = Seq(dtlb_st_tlb_st.io)
+  val dtlb_prefetch = Seq(dtlb_prefetch_tlb_prefetch.io)
+  val dtlb = dtlb_ld ++ dtlb_st ++ dtlb_prefetch
 
   val ptwio = Wire(new VectorTlbPtwIO(DTlbSize))
   val dtlb_reqs = dtlb.map(_.requestor).flatten
@@ -710,7 +576,7 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
   } else {
     // TODO: there will be bugs in TlbReplace when outReplace enable, since the order of Hyu is not right.
     if (ldtlbParams.outReplace) {
-      val replace_ld = Module(new TlbReplace(LduCnt + 1, ldtlbParams))
+      val replace_ld = Module(new TlbReplace(LduCnt + PfNumInDtlbLD, ldtlbParams))
       replace_ld.io.apply_sep(dtlb_ld.map(_.replace), ptwio.resp.bits.data.s1.entry.tag)
     }
     if (hytlbParams.outReplace) {
@@ -765,9 +631,9 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
     dtlb_st.foreach(_.ptw.resp.valid := ptw_resp_v && Cat(ptw_resp_next.vector.slice(TlbStartVec(dtlb_st_idx), TlbEndVec(dtlb_st_idx))).orR)
     dtlb_prefetch.foreach(_.ptw.resp.valid := ptw_resp_v && Cat(ptw_resp_next.vector.slice(TlbStartVec(dtlb_pf_idx), TlbEndVec(dtlb_pf_idx))).orR)
   }
-  dtlb_ld.foreach(_.ptw.resp.bits.getGpa := Cat(ptw_resp_next.getGpa.take(LduCnt + HyuCnt + 1)).orR)
-  dtlb_st.foreach(_.ptw.resp.bits.getGpa := Cat(ptw_resp_next.getGpa.slice(LduCnt + HyuCnt + 1, LduCnt + HyuCnt + 1 + StaCnt)).orR)
-  dtlb_prefetch.foreach(_.ptw.resp.bits.getGpa := Cat(ptw_resp_next.getGpa.drop(LduCnt + HyuCnt + 1 + StaCnt)).orR)
+  dtlb_ld.foreach(_.ptw.resp.bits.getGpa := Cat(ptw_resp_next.getGpa.slice(TlbStartVec(dtlb_ld_idx), TlbEndVec(dtlb_ld_idx))).orR)
+  dtlb_st.foreach(_.ptw.resp.bits.getGpa := Cat(ptw_resp_next.getGpa.slice(TlbStartVec(dtlb_st_idx), TlbEndVec(dtlb_st_idx))).orR)
+  dtlb_prefetch.foreach(_.ptw.resp.bits.getGpa := Cat(ptw_resp_next.getGpa.slice(TlbStartVec(dtlb_pf_idx), TlbEndVec(dtlb_pf_idx))).orR)
 
   val dtlbRepeater  = PTWNewFilter(ldtlbParams.fenceDelay, ptwio, ptw.io.tlb(1), sfence, tlbcsr, l2tlbParams.dfilterSize)
   val itlbRepeater3 = PTWRepeaterNB(passReady = false, itlbParams.fenceDelay, io.fetch_to_mem.itlb, ptw.io.tlb(0), sfence, tlbcsr)
@@ -789,6 +655,111 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
     require(p.req.bits.size.getWidth == d.bits.size.getWidth)
   }
 
+
+  /******************************************************************
+   * Prefetcher
+   *  - L1 site
+   *  - L2 site: provide tlb interface
+   *  - L1 prefetch request: set confidence
+   *  - L1 prefetcher: fuzzer interface
+   ******************************************************************/
+  
+  /** prefetcher site in L1 */
+  val l1_pf_req = Wire(Decoupled(new L1PrefetchReq()))
+  val prefetcher = Module(new PrefetcherWrapper)
+  prefetcher.io.pfCtrlFromTile.l2PfqBusy := io.l2PfqBusy
+  prefetcher.io.pfCtrlFromCSR := io.ooo_to_mem.csrCtrl.pf_ctrl
+  prefetcher.io.pfCtrlFromDCache <> dcache.io.pf_ctrl
+  prefetcher.io.fromDCache.sms_agt_evict_req <> dcache.io.sms_agt_evict_req
+  prefetcher.io.fromOOO.s1_loadPc := io.ooo_to_mem.issueLda.map(x => RegNext(x.bits.uop.pc)) ++ io.ooo_to_mem.hybridPc
+  prefetcher.io.fromOOO.s1_storePc := io.ooo_to_mem.storePc ++ io.ooo_to_mem.hybridPc
+  prefetcher.io.trainSource.s1_loadFireHint := loadUnits.map(_.io.s1_prefetch_spec) ++ hybridUnits.map(_.io.s1_prefetch_spec)
+  prefetcher.io.trainSource.s2_loadFireHint := loadUnits.map(_.io.s2_prefetch_spec) ++ hybridUnits.map(_.io.s2_prefetch_spec)
+  prefetcher.io.trainSource.s3_load := loadUnits.map(_.io.prefetch_train) ++ hybridUnits.map(_.io.prefetch_train)
+  prefetcher.io.trainSource.s3_ptrChasing := loadUnits.map(_.io.s3_ptr_chasing) ++ hybridUnits.map(_.io.ldu_io.s3_ptr_chasing)
+  prefetcher.io.trainSource.s1_storeFireHint := storeUnits.map(_.io.s1_prefetch_spec) ++ hybridUnits.map(_.io.s1_prefetch_spec)
+  prefetcher.io.trainSource.s2_storeFireHint := storeUnits.map(_.io.s2_prefetch_spec) ++ hybridUnits.map(_.io.s2_prefetch_spec)
+  prefetcher.io.trainSource.s3_store <> storeUnits.map(_.io.prefetch_train) ++ hybridUnits.map(_.io.prefetch_train)
+  (0 until prefetcherNum).foreach { i => //NOTE lyq: prefetcherNum minimum is 1 for simpler code generation, which is ugly
+    prefetcher.io.tlb_req(i).req.ready := false.B
+    prefetcher.io.tlb_req(i).resp.valid := false.B
+    prefetcher.io.tlb_req(i).resp.bits := DontCare
+    prefetcher.io.pmp_resp(i) := DontCare
+  }
+  pf2tlbIndexMap.zipWithIndex.foreach{ case (k, i) =>
+    dtlb_reqs(k) <> prefetcher.io.tlb_req(i)
+    prefetcher.io.pmp_resp(i) := pmp_check(k).resp
+  }
+  l1_pf_req <> prefetcher.io.l1_pf_to_l1
+  outer.l2_pf_sender_opt.foreach(_.out.head._1.addr_valid := prefetcher.io.l1_pf_to_l2.addr_valid)
+  outer.l2_pf_sender_opt.foreach(_.out.head._1.addr := prefetcher.io.l1_pf_to_l2.addr)
+  outer.l2_pf_sender_opt.foreach(_.out.head._1.pf_source :=  prefetcher.io.l1_pf_to_l2.pf_source)
+  outer.l2_pf_sender_opt.foreach(_.out.head._1.l2_pf_en := prefetcher.io.l1_pf_to_l2.l2_pf_en)
+  outer.l3_pf_sender_opt.foreach(_.out.head._1.addr_valid := prefetcher.io.l1_pf_to_l3.addr_valid)
+  outer.l3_pf_sender_opt.foreach(_.out.head._1.addr := prefetcher.io.l1_pf_to_l3.addr)
+  outer.l3_pf_sender_opt.foreach(_.out.head._1.l2_pf_en := prefetcher.io.l1_pf_to_l3.l2_pf_en)
+  XSPerfAccumulate("prefetch_fire_l1", l1_pf_req.fire)
+  XSPerfAccumulate("prefetch_fire_l2", outer.l2_pf_sender_opt.map(_.out.head._1.addr_valid).getOrElse(false.B))
+  XSPerfAccumulate("prefetch_fire_l3", outer.l3_pf_sender_opt.map(_.out.head._1.addr_valid).getOrElse(false.B))
+  
+  /** prefetcher site in L2 */
+  dtlb_reqs(L2toL1DTLBPortIndex) <> io.l2_tlb_req
+  dtlb_reqs(L2toL1DTLBPortIndex).resp.ready := true.B
+  io.l2_pmp_resp := pmp_check(L2toL1DTLBPortIndex).resp
+
+  /** prefetch to l1 req */
+  // Stream's confidence is always 1
+  // (LduCnt + HyuCnt) l1_pf_reqs ?
+  loadUnits.foreach(load_unit => {
+    load_unit.io.prefetch_req.valid <> l1_pf_req.valid
+    load_unit.io.prefetch_req.bits <> l1_pf_req.bits
+  })
+  hybridUnits.foreach(hybrid_unit => {
+    hybrid_unit.io.ldu_io.prefetch_req.valid <> l1_pf_req.valid
+    hybrid_unit.io.ldu_io.prefetch_req.bits <> l1_pf_req.bits
+  })
+  // NOTE: loadUnits(0) has higher bank conflict and miss queue arb priority than loadUnits(1) and loadUnits(2)
+  // when loadUnits(1)/loadUnits(2) stage 0 is busy, hw prefetch will never use that pipeline
+  val LowConfPorts = if (LduCnt == 2) Seq(1) else if (LduCnt == 3) Seq(1, 2) else Seq(0)
+  LowConfPorts.map{case i => loadUnits(i).io.prefetch_req.bits.confidence := 0.U}
+  hybridUnits.foreach(hybrid_unit => { hybrid_unit.io.ldu_io.prefetch_req.bits.confidence := 0.U })
+
+  val canAcceptHighConfPrefetch = loadUnits.map(_.io.canAcceptHighConfPrefetch) ++
+                                  hybridUnits.map(_.io.canAcceptLowConfPrefetch)
+  val canAcceptLowConfPrefetch = loadUnits.map(_.io.canAcceptLowConfPrefetch) ++
+                                 hybridUnits.map(_.io.canAcceptLowConfPrefetch)
+  l1_pf_req.ready := (0 until LduCnt + HyuCnt).map{
+    case i => {
+      if (LowConfPorts.contains(i)) {
+        loadUnits(i).io.canAcceptLowConfPrefetch
+      } else {
+        Mux(l1_pf_req.bits.confidence === 1.U, canAcceptHighConfPrefetch(i), canAcceptLowConfPrefetch(i))
+      }
+    }
+  }.reduce(_ || _)
+  /** l1 pf fuzzer interface */
+  val DebugEnableL1PFFuzzer = false
+  if (DebugEnableL1PFFuzzer) {
+    // l1 pf req fuzzer
+    val fuzzer = Module(new L1PrefetchFuzzer())
+    fuzzer.io.vaddr := DontCare
+    fuzzer.io.paddr := DontCare
+
+    // override load_unit prefetch_req
+    loadUnits.foreach(load_unit => {
+      load_unit.io.prefetch_req.valid <> fuzzer.io.req.valid
+      load_unit.io.prefetch_req.bits <> fuzzer.io.req.bits
+    })
+
+    // override hybrid_unit prefetch_req
+    hybridUnits.foreach(hybrid_unit => {
+      hybrid_unit.io.ldu_io.prefetch_req.valid <> fuzzer.io.req.valid
+      hybrid_unit.io.ldu_io.prefetch_req.bits <> fuzzer.io.req.bits
+    })
+
+    fuzzer.io.req.ready := l1_pf_req.ready
+  }
+
   for (i <- 0 until LduCnt) {
     io.debug_ls.debugLsInfo(i) := loadUnits(i).io.debug_ls
   }
@@ -943,28 +914,6 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
       loadUnits(i).io.stld_nuke_query(s) := stld_nuke_query(s)
     }
     loadUnits(i).io.lq_rep_full <> lsq.io.lq_rep_full
-    // load prefetch train
-    prefetcherOpt.foreach(pf => {
-      // sms will train on all miss load sources
-      val source = loadUnits(i).io.prefetch_train
-      pf.io.ld_in(i).valid := Mux(pf_train_on_hit,
-        source.valid,
-        source.valid && source.bits.isFirstIssue && source.bits.miss
-      )
-      pf.io.ld_in(i).bits := source.bits
-      val loadPc = RegNext(io.ooo_to_mem.issueLda(i).bits.uop.pc) // for s1
-      pf.io.ld_in(i).bits.uop.pc := Mux(
-        loadUnits(i).io.s2_ptr_chasing,
-        RegEnable(loadPc, loadUnits(i).io.s2_prefetch_spec),
-        RegEnable(RegEnable(loadPc, loadUnits(i).io.s1_prefetch_spec), loadUnits(i).io.s2_prefetch_spec)
-      )
-    })
-    l1PrefetcherOpt.foreach(pf => {
-      // stream will train on all load sources
-      val source = loadUnits(i).io.prefetch_train_l1
-      pf.io.ld_in(i).valid := source.valid && source.bits.isFirstIssue
-      pf.io.ld_in(i).bits := source.bits
-    })
 
     // load to load fast forward: load(i) prefers data(i)
     val l2l_fwd_out = loadUnits.map(_.io.l2l_fwd_out) ++ hybridUnits.map(_.io.ldu_io.l2l_fwd_out)
@@ -1082,34 +1031,6 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
     val stld_nuke_query = VecInit(storeUnits.map(_.io.stld_nuke_query) ++ hybridUnits.map(_.io.stu_io.stld_nuke_query))
     hybridUnits(i).io.ldu_io.stld_nuke_query := stld_nuke_query
     hybridUnits(i).io.ldu_io.lq_rep_full <> lsq.io.lq_rep_full
-    // load prefetch train
-    prefetcherOpt.foreach(pf => {
-      val source = hybridUnits(i).io.prefetch_train
-      pf.io.ld_in(LduCnt + i).valid := Mux(pf_train_on_hit,
-        source.valid,
-        source.valid && source.bits.isFirstIssue && source.bits.miss
-      )
-      pf.io.ld_in(LduCnt + i).bits := source.bits
-      pf.io.ld_in(LduCnt + i).bits.uop.pc := Mux(hybridUnits(i).io.ldu_io.s2_ptr_chasing, io.ooo_to_mem.hybridPc(i), RegNext(io.ooo_to_mem.hybridPc(i)))
-    })
-    l1PrefetcherOpt.foreach(pf => {
-      // stream will train on all load sources
-      val source = hybridUnits(i).io.prefetch_train_l1
-      pf.io.ld_in(LduCnt + i).valid := source.valid && source.bits.isFirstIssue &&
-                                       FuType.isLoad(source.bits.uop.fuType)
-      pf.io.ld_in(LduCnt + i).bits := source.bits
-      pf.io.st_in(StaCnt + i).valid := false.B
-      pf.io.st_in(StaCnt + i).bits := DontCare
-    })
-    prefetcherOpt.foreach(pf => {
-      val source = hybridUnits(i).io.prefetch_train
-      pf.io.st_in(StaCnt + i).valid := Mux(pf_train_on_hit,
-        source.valid,
-        source.valid && source.bits.isFirstIssue && source.bits.miss
-      ) && FuType.isStore(source.bits.uop.fuType)
-      pf.io.st_in(StaCnt + i).bits := source.bits
-      pf.io.st_in(StaCnt + i).bits.uop.pc := RegNext(io.ooo_to_mem.hybridPc(i))
-    })
 
     // load to load fast forward: load(i) prefers data(i)
     val l2l_fwd_out = loadUnits.map(_.io.l2l_fwd_out) ++ hybridUnits.map(_.io.ldu_io.l2l_fwd_out)
@@ -1203,32 +1124,6 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
   lsq.io.cmoOpReq <> dcache.io.cmoOpReq
   lsq.io.cmoOpResp <> dcache.io.cmoOpResp
 
-  // Prefetcher
-  val StreamDTLBPortIndex = TlbStartVec(dtlb_ld_idx) + LduCnt + HyuCnt
-  val PrefetcherDTLBPortIndex = TlbStartVec(dtlb_pf_idx)
-  val L2toL1DLBPortIndex = TlbStartVec(dtlb_pf_idx) + 1
-  prefetcherOpt match {
-  case Some(pf) =>
-    dtlb_reqs(PrefetcherDTLBPortIndex) <> pf.io.tlb_req
-    pf.io.pmp_resp := pmp_check(PrefetcherDTLBPortIndex).resp
-  case None =>
-    dtlb_reqs(PrefetcherDTLBPortIndex) := DontCare
-    dtlb_reqs(PrefetcherDTLBPortIndex).req.valid := false.B
-    dtlb_reqs(PrefetcherDTLBPortIndex).resp.ready := true.B
-  }
-  l1PrefetcherOpt match {
-    case Some(pf) =>
-      dtlb_reqs(StreamDTLBPortIndex) <> pf.io.tlb_req
-      pf.io.pmp_resp := pmp_check(StreamDTLBPortIndex).resp
-    case None =>
-        dtlb_reqs(StreamDTLBPortIndex) := DontCare
-        dtlb_reqs(StreamDTLBPortIndex).req.valid := false.B
-        dtlb_reqs(StreamDTLBPortIndex).resp.ready := true.B
-  }
-  dtlb_reqs(L2toL1DLBPortIndex) <> io.l2_tlb_req
-  dtlb_reqs(L2toL1DLBPortIndex).resp.ready := true.B
-  io.l2_pmp_resp := pmp_check(L2toL1DLBPortIndex).resp
-
   // StoreUnit
   for (i <- 0 until StdCnt) {
     stdExeUnits(i).io.flush <> redirect
@@ -1249,7 +1144,7 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
     stu.io.lsq_replenish <> lsq.io.sta.storeAddrInRe(i)
     // dtlb
     stu.io.tlb          <> dtlb_st.head.requestor(i)
-    stu.io.pmp          <> pmp_check(LduCnt + HyuCnt + 1 + i).resp
+    stu.io.pmp          <> pmp_check(TlbStartVec(dtlb_st_idx) + i).resp
 
     // -------------------------
     // Store Triggers
@@ -1306,25 +1201,6 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
     lsq.io.std.storeDataIn.map(_.bits.debug := 0.U.asTypeOf(new DebugBundle))
     lsq.io.std.storeDataIn.foreach(_.bits.isFromLoadUnit := DontCare)
 
-
-    // store prefetch train
-    l1PrefetcherOpt.foreach(pf => {
-      // stream will train on all load sources
-      pf.io.st_in(i).valid := false.B
-      pf.io.st_in(i).bits := DontCare
-    })
-
-    prefetcherOpt.foreach(pf => {
-      pf.io.st_in(i).valid := Mux(pf_train_on_hit,
-        stu.io.prefetch_train.valid,
-        stu.io.prefetch_train.valid && stu.io.prefetch_train.bits.isFirstIssue && (
-          stu.io.prefetch_train.bits.miss
-          )
-      )
-      pf.io.st_in(i).bits := stu.io.prefetch_train.bits
-      pf.io.st_in(i).bits.uop.pc := RegEnable(RegEnable(io.ooo_to_mem.storePc(i), stu.io.s1_prefetch_spec), stu.io.s2_prefetch_spec)
-    })
-
     // 1. sync issue info to store set LFST
     // 2. when store issue, broadcast issued sqPtr to wake up the following insts
     // io.stIn(i).valid := io.issue(exuParameters.LduCnt + i).valid
@@ -2028,8 +1904,7 @@ class MemBlockInlinedImp(outer: MemBlockInlined) extends LazyModuleImp(outer)
         ModuleNode(pmp)
       )
       ++ pmp_checkers.map(ModuleNode(_))
-      ++ (if (prefetcherOpt.isDefined) Seq(ModuleNode(prefetcherOpt.get)) else Nil)
-      ++ (if (l1PrefetcherOpt.isDefined) Seq(ModuleNode(l1PrefetcherOpt.get)) else Nil)
+      ++ Seq(ModuleNode(prefetcher))
     )
     val rightResetTree = ResetGenNode(
       Seq(
diff --git a/src/main/scala/xiangshan/mem/pipeline/HybridUnit.scala b/src/main/scala/xiangshan/mem/pipeline/HybridUnit.scala
index 8b71b842052..2b35c85b78f 100644
--- a/src/main/scala/xiangshan/mem/pipeline/HybridUnit.scala
+++ b/src/main/scala/xiangshan/mem/pipeline/HybridUnit.scala
@@ -105,7 +105,7 @@ class HybridUnit(implicit p: Parameters) extends XSModule
       val lq_rep_full  = Input(Bool())
 
       // misc
-      val s2_ptr_chasing = Output(Bool()) // provide right pc for hw prefetch
+      val s3_ptr_chasing = Output(Bool()) // provide right pc for hw prefetch
 
       // Load fast replay path
       val fast_rep_in  = Flipped(Decoupled(new LqWriteBundle))
@@ -138,11 +138,10 @@ class HybridUnit(implicit p: Parameters) extends XSModule
     }
 
     // speculative for gated control
-    val s0_prefetch_spec = Output(Bool())
     val s1_prefetch_spec = Output(Bool())
+    val s2_prefetch_spec = Output(Bool())
     // prefetch
-    val prefetch_train            = ValidIO(new LsPrefetchTrainBundle()) // provide prefetch info to sms
-    val prefetch_train_l1         = ValidIO(new LsPrefetchTrainBundle()) // provide prefetch info to stream & stride
+    val prefetch_train            = ValidIO(new LsPrefetchTrainBundle()) // provide prefetch info
     val canAcceptLowConfPrefetch  = Output(Bool())
     val canAcceptHighConfPrefetch = Output(Bool())
     val correctMissTrain          = Input(Bool())
@@ -1080,22 +1079,21 @@ class HybridUnit(implicit p: Parameters) extends XSModule
   io.ldu_io.fast_uop.bits := RegNext(s1_out.uop)
 
   //
-  io.ldu_io.s2_ptr_chasing                    := RegEnable(s1_try_ptr_chasing && !s1_cancel_ptr_chasing, false.B, s1_fire)
+  val s2_ptr_chasing = RegEnable(s1_try_ptr_chasing && !s1_cancel_ptr_chasing, false.B, s1_fire)
+  io.ldu_io.s3_ptr_chasing := RegEnable(s2_ptr_chasing, false.B, s2_fire)
 
-  // prefetch train
-  io.s0_prefetch_spec := s0_fire
+  // prefetch train --> s3
   io.s1_prefetch_spec := s1_fire
-  io.prefetch_train.valid              := s2_valid && !s2_actually_mmio && !s2_in.tlbMiss
-  io.prefetch_train.bits.fromLsPipelineBundle(s2_in)
-  io.prefetch_train.bits.miss          := Mux(s2_ld_flow, io.ldu_io.dcache.resp.bits.miss, io.stu_io.dcache.resp.bits.miss) // TODO: use trace with bank conflict?
-  io.prefetch_train.bits.meta_prefetch := Mux(s2_ld_flow, io.ldu_io.dcache.resp.bits.meta_prefetch, false.B)
-  io.prefetch_train.bits.meta_access   := Mux(s2_ld_flow, io.ldu_io.dcache.resp.bits.meta_access, false.B)
-
-  io.prefetch_train_l1.valid              := s2_valid && !s2_actually_mmio && s2_ld_flow
-  io.prefetch_train_l1.bits.fromLsPipelineBundle(s2_in)
-  io.prefetch_train_l1.bits.miss          := io.ldu_io.dcache.resp.bits.miss
-  io.prefetch_train_l1.bits.meta_prefetch := io.ldu_io.dcache.resp.bits.meta_prefetch
-  io.prefetch_train_l1.bits.meta_access   := io.ldu_io.dcache.resp.bits.meta_access
+  io.s2_prefetch_spec := s2_fire
+  val s2_prefetch_train_valid = s2_valid && !s2_actually_mmio && !s2_in.isHWPrefetch
+  io.prefetch_train.valid              := s2_prefetch_train_valid
+  io.prefetch_train.bits.fromLsPipelineBundle(s2_in, latch = true, enable = s2_prefetch_train_valid)
+  // TODO: use trace with bank conflict?
+  io.prefetch_train.bits.miss          := RegEnable(Mux(s2_ld_flow, io.ldu_io.dcache.resp.bits.miss, io.stu_io.dcache.resp.bits.miss), s2_prefetch_train_valid)
+  io.prefetch_train.bits.meta_prefetch := RegEnable(Mux(s2_ld_flow, io.ldu_io.dcache.resp.bits.meta_prefetch, false.B), s2_prefetch_train_valid)
+  io.prefetch_train.bits.meta_access   := RegEnable(Mux(s2_ld_flow, io.ldu_io.dcache.resp.bits.meta_access, false.B), s2_prefetch_train_valid)
+  io.prefetch_train.bits.is_from_hw_pf := RegNext(s2_in.isHWPrefetch)
+
   if (env.FPGAPlatform){
     io.ldu_io.dcache.s0_pc := DontCare
     io.ldu_io.dcache.s1_pc := DontCare
diff --git a/src/main/scala/xiangshan/mem/pipeline/LoadUnit.scala b/src/main/scala/xiangshan/mem/pipeline/LoadUnit.scala
index a775d278d7d..79f67aeefb0 100644
--- a/src/main/scala/xiangshan/mem/pipeline/LoadUnit.scala
+++ b/src/main/scala/xiangshan/mem/pipeline/LoadUnit.scala
@@ -154,8 +154,8 @@ class LoadUnit(implicit p: Parameters) extends XSModule
     val fromCsrTrigger = Input(new CsrTriggerBundle)
 
     // prefetch
-    val prefetch_train            = ValidIO(new LsPrefetchTrainBundle()) // provide prefetch info to sms
-    val prefetch_train_l1         = ValidIO(new LsPrefetchTrainBundle()) // provide prefetch info to stream & stride
+    val prefetch_train            = ValidIO(new LsPrefetchTrainBundle()) // provide prefetch info
+
     // speculative for gated control
     val s1_prefetch_spec = Output(Bool())
     val s2_prefetch_spec = Output(Bool())
@@ -192,7 +192,7 @@ class LoadUnit(implicit p: Parameters) extends XSModule
     val lq_rep_full  = Input(Bool())
 
     // misc
-    val s2_ptr_chasing = Output(Bool()) // provide right pc for hw prefetch
+    val s3_ptr_chasing = Output(Bool()) // provide right pc for hw prefetch
 
     // Load fast replay path
     val fast_rep_in  = Flipped(Decoupled(new LqWriteBundle))
@@ -1441,40 +1441,26 @@ class LoadUnit(implicit p: Parameters) extends XSModule
   io.fast_uop.valid := GatedValidRegNext(s1_fast_uop_valid) && (s2_valid && !s2_out.rep_info.need_rep && !s2_uncache && !(s2_prf && !s2_hw_prf)) && !s2_isvec && !s2_frm_mabuf
   io.fast_uop.bits := RegEnable(s1_out.uop, s1_fast_uop_valid)
 
-  //
-  io.s2_ptr_chasing                    := RegEnable(s1_try_ptr_chasing && !s1_cancel_ptr_chasing, false.B, s1_fire)
-
   // RegNext prefetch train for better timing
   // ** Now, prefetch train is valid at load s3 **
-  val s2_prefetch_train_valid = WireInit(false.B)
-  s2_prefetch_train_valid              := s2_valid && !s2_actually_uncache && (!s2_in.tlbMiss || s2_hw_prf)
-  io.prefetch_train.valid              := GatedValidRegNext(s2_prefetch_train_valid)
+  val s2_ptr_chasing = RegEnable(s1_try_ptr_chasing && !s1_cancel_ptr_chasing, false.B, s1_fire)
+  val s2_prefetch_train_valid = s2_valid && !s2_actually_uncache
+  io.prefetch_train.valid := GatedValidRegNext(s2_prefetch_train_valid)
   io.prefetch_train.bits.fromLsPipelineBundle(s2_in, latch = true, enable = s2_prefetch_train_valid)
-  io.prefetch_train.bits.miss          := RegEnable(io.dcache.resp.bits.miss, s2_prefetch_train_valid) // TODO: use trace with bank conflict?
+  io.prefetch_train.bits.miss := RegEnable(io.dcache.resp.bits.miss, s2_prefetch_train_valid) // TODO: use trace with bank conflict?
   io.prefetch_train.bits.meta_prefetch := RegEnable(io.dcache.resp.bits.meta_prefetch, s2_prefetch_train_valid)
-  io.prefetch_train.bits.meta_access   := RegEnable(io.dcache.resp.bits.meta_access, s2_prefetch_train_valid)
-  io.prefetch_train.bits.isFinalSplit      := false.B
+  io.prefetch_train.bits.meta_access := RegEnable(io.dcache.resp.bits.meta_access, s2_prefetch_train_valid)
+  io.prefetch_train.bits.is_from_hw_pf := RegNext(s2_hw_prf)
+  io.prefetch_train.bits.isFinalSplit := false.B
   io.prefetch_train.bits.misalignWith16Byte := false.B
   io.prefetch_train.bits.misalignNeedWakeUp := false.B
   io.prefetch_train.bits.updateAddrValid := false.B
   io.prefetch_train.bits.isMisalign := false.B
   io.prefetch_train.bits.hasException := false.B
   io.s1_prefetch_spec := s1_fire
-  io.s2_prefetch_spec := s2_prefetch_train_valid
-
-  val s2_prefetch_train_l1_valid = WireInit(false.B)
-  s2_prefetch_train_l1_valid              := s2_valid && !s2_actually_uncache
-  io.prefetch_train_l1.valid              := GatedValidRegNext(s2_prefetch_train_l1_valid)
-  io.prefetch_train_l1.bits.fromLsPipelineBundle(s2_in, latch = true, enable = s2_prefetch_train_l1_valid)
-  io.prefetch_train_l1.bits.miss          := RegEnable(io.dcache.resp.bits.miss, s2_prefetch_train_l1_valid)
-  io.prefetch_train_l1.bits.meta_prefetch := RegEnable(io.dcache.resp.bits.meta_prefetch, s2_prefetch_train_l1_valid)
-  io.prefetch_train_l1.bits.meta_access   := RegEnable(io.dcache.resp.bits.meta_access, s2_prefetch_train_l1_valid)
-  io.prefetch_train_l1.bits.isFinalSplit      := false.B
-  io.prefetch_train_l1.bits.misalignWith16Byte := false.B
-  io.prefetch_train_l1.bits.misalignNeedWakeUp := false.B
-  io.prefetch_train_l1.bits.updateAddrValid := false.B
-  io.prefetch_train_l1.bits.hasException := false.B
-  io.prefetch_train_l1.bits.isMisalign := false.B
+  io.s2_prefetch_spec := s2_fire
+  io.s3_ptr_chasing := RegEnable(s2_ptr_chasing, false.B, s2_fire)
+
   if (env.FPGAPlatform){
     io.dcache.s0_pc := DontCare
     io.dcache.s1_pc := DontCare
diff --git a/src/main/scala/xiangshan/mem/pipeline/StoreUnit.scala b/src/main/scala/xiangshan/mem/pipeline/StoreUnit.scala
index dedc2bac2e0..90accd9b4d1 100644
--- a/src/main/scala/xiangshan/mem/pipeline/StoreUnit.scala
+++ b/src/main/scala/xiangshan/mem/pipeline/StoreUnit.scala
@@ -526,29 +526,23 @@ class StoreUnit(implicit p: Parameters) extends XSModule
   // RegNext prefetch train for better timing
   // ** Now, prefetch train is valid at store s3 **
   val s2_prefetch_train_valid = WireInit(false.B)
-  s2_prefetch_train_valid := s2_valid && io.dcache.resp.fire && !s2_out.mmio && !s2_out.nc && !s2_in.tlbMiss && !s2_in.isHWPrefetch
-  if(EnableStorePrefetchSMS) {
-    io.s1_prefetch_spec := s1_fire
-    io.s2_prefetch_spec := s2_prefetch_train_valid
-    io.prefetch_train.valid := RegNext(s2_prefetch_train_valid)
-    io.prefetch_train.bits.fromLsPipelineBundle(s2_in, latch = true, enable = s2_prefetch_train_valid)
-  }else {
-    io.s1_prefetch_spec := false.B
-    io.s2_prefetch_spec := false.B
-    io.prefetch_train.valid := false.B
-    io.prefetch_train.bits.fromLsPipelineBundle(s2_in, latch = true, enable = false.B)
-  }
+  s2_prefetch_train_valid := s2_valid && io.dcache.resp.fire && !s2_out.mmio && !s2_out.nc && !s2_in.tlbMiss
+  io.prefetch_train.valid := RegNext(s2_prefetch_train_valid)
+  io.prefetch_train.bits.fromLsPipelineBundle(s2_in, latch = true, enable = s2_prefetch_train_valid)
   // override miss bit
   io.prefetch_train.bits.miss := RegEnable(io.dcache.resp.bits.miss, s2_prefetch_train_valid)
   // TODO: add prefetch and access bit
   io.prefetch_train.bits.meta_prefetch := false.B
   io.prefetch_train.bits.meta_access := false.B
+  io.prefetch_train.bits.is_from_hw_pf := RegNext(s2_in.isHWPrefetch)
   io.prefetch_train.bits.isFinalSplit := false.B
   io.prefetch_train.bits.misalignWith16Byte := false.B
   io.prefetch_train.bits.isMisalign := false.B
   io.prefetch_train.bits.misalignNeedWakeUp := false.B
   io.prefetch_train.bits.updateAddrValid := false.B
   io.prefetch_train.bits.hasException := false.B
+  io.s1_prefetch_spec := s1_fire
+  io.s2_prefetch_spec := s2_prefetch_train_valid
 
   // Pipeline
   // --------------------------------------------------------------------------------
diff --git a/src/main/scala/xiangshan/mem/prefetch/BasePrefecher.scala b/src/main/scala/xiangshan/mem/prefetch/BasePrefecher.scala
index bcc155848ac..6be68b7ce78 100644
--- a/src/main/scala/xiangshan/mem/prefetch/BasePrefecher.scala
+++ b/src/main/scala/xiangshan/mem/prefetch/BasePrefecher.scala
@@ -60,14 +60,16 @@ class L2PrefetchReq(implicit p: Parameters) extends XSBundle {
   val source = UInt(MemReqSource.reqSourceBits.W)
 }
 
+class L3PrefetchReq(implicit p: Parameters) extends L2PrefetchReq
+
 class PrefetcherIO()(implicit p: Parameters) extends XSBundle {
   val ld_in = Flipped(Vec(backendParams.LdExuCnt, ValidIO(new LsPrefetchTrainBundle())))
   val st_in = Flipped(Vec(backendParams.StaExuCnt, ValidIO(new LsPrefetchTrainBundle())))
   val tlb_req = new TlbRequestIO(nRespDups = 2)
   val pmp_resp = Flipped(new PMPRespBundle())
   val l1_req = DecoupledIO(new L1PrefetchReq())
-  val l2_req = ValidIO(new L2PrefetchReq())
-  val l3_req = ValidIO(UInt(PAddrBits.W)) // TODO: l3 pf source
+  val l2_req = DecoupledIO(new L2PrefetchReq())
+  val l3_req = DecoupledIO(new L3PrefetchReq())
   val enable = Input(Bool())
 }
 
@@ -79,11 +81,16 @@ class PrefetchReqBundle()(implicit p: Parameters) extends XSBundle {
   val pfHitStream = Bool()
 }
 
-trait PrefetcherParams
-
 abstract class BasePrefecher()(implicit p: Parameters) extends XSModule {
   val io = IO(new PrefetcherIO())
 
+  // By default, there are no tlb transformations, l2_req and l3_req
+  io.tlb_req.req.valid := false.B
+  io.tlb_req.req.bits := DontCare
+  io.tlb_req.req_kill := false.B
+  io.tlb_req.resp.ready := true.B
+  io.l2_req.valid := false.B
+  io.l2_req.bits  := DontCare
   io.l3_req.valid := false.B
   io.l3_req.bits  := DontCare
 }
diff --git a/src/main/scala/xiangshan/mem/prefetch/L1PrefetchComponent.scala b/src/main/scala/xiangshan/mem/prefetch/L1PrefetchComponent.scala
index 52108012214..d64d018c879 100644
--- a/src/main/scala/xiangshan/mem/prefetch/L1PrefetchComponent.scala
+++ b/src/main/scala/xiangshan/mem/prefetch/L1PrefetchComponent.scala
@@ -15,6 +15,11 @@ import xiangshan.mem.L1PrefetchSource
 import xiangshan.cache.HasDCacheParameters
 import xiangshan.cache.mmu._
 
+case class StreamStrideParams() extends PrefetcherParams{
+  override def name = "StreamStride"
+  override def tlbPlace = TLBPlace.dtlb_ld
+}
+
 trait HasL1PrefetchHelper extends HasCircularQueuePtrHelper with HasDCacheParameters {
   // region related
   val REGION_SIZE = 1024
@@ -374,7 +379,7 @@ class MutiLevelPrefetchFilter(implicit p: Parameters) extends XSModule with HasL
     val pmp_resp = Flipped(new PMPRespBundle())
     val l1_req = DecoupledIO(new L1PrefetchReq())
     val l2_pf_addr = ValidIO(new L2PrefetchReq())
-    val l3_pf_addr = ValidIO(UInt(PAddrBits.W)) // TODO: l3 pf source
+    val l3_pf_addr = ValidIO(new L3PrefetchReq())
     val confidence = Input(UInt(1.W))
     val l2PfqBusy = Input(Bool())
   })
@@ -423,7 +428,7 @@ class MutiLevelPrefetchFilter(implicit p: Parameters) extends XSModule with HasL
     val req = new L2PrefetchReq
     val debug_vaddr = UInt(VAddrBits.W)
   }, MLP_L2L3_SIZE))
-  val l3_pf_req_arb = Module(new RRArbiterInit(UInt(PAddrBits.W), MLP_L2L3_SIZE))
+  val l3_pf_req_arb = Module(new RRArbiterInit(new L3PrefetchReq, MLP_L2L3_SIZE))
 
   val l1_opt_replace_vec = VecInit(l1_array.zip(l1_valids).map{case (e, v) => e.may_be_replace(v)})
   val l2_opt_replace_vec = VecInit(l2_array.zip(l2_valids).map{case (e, v) => e.may_be_replace(v)})
@@ -812,11 +817,15 @@ class MutiLevelPrefetchFilter(implicit p: Parameters) extends XSModule with HasL
   for(i <- 0 until MLP_L2L3_SIZE) {
     val evict = s1_l2_alloc && (s1_l2_index === i.U)
     l3_pf_req_arb.io.in(i).valid := l2_array(i).can_send_pf(l2_valids(i)) && (l2_array(i).sink === SINK_L3) && !evict
-    l3_pf_req_arb.io.in(i).bits := l2_array(i).get_pf_addr()
+    l3_pf_req_arb.io.in(i).bits.addr := l2_array(i).get_pf_addr()
+    l3_pf_req_arb.io.in(i).bits.source := MuxLookup(l2_array(i).source.value, MemReqSource.Prefetch2L3Unknown.id.U)(Seq(
+      L1_HW_PREFETCH_STRIDE -> MemReqSource.Prefetch2L3Stride.id.U,
+      L1_HW_PREFETCH_STREAM -> MemReqSource.Prefetch2L3Stream.id.U
+    ))
   }
 
   when(l3_pf_req_arb.io.out.valid) {
-    l2_array(l3_pf_req_arb.io.chosen).sent_vec := l2_array(l3_pf_req_arb.io.chosen).sent_vec | get_candidate_oh(l3_pf_req_arb.io.out.bits)
+    l2_array(l3_pf_req_arb.io.chosen).sent_vec := l2_array(l3_pf_req_arb.io.chosen).sent_vec | get_candidate_oh(l3_pf_req_arb.io.out.bits.addr)
   }
 
   // reset meta to avoid muti-hit problem
@@ -917,7 +926,7 @@ class L1Prefetcher(implicit p: Parameters) extends BasePrefecher with HasStreamP
   io.l2_req.valid := pf_queue_filter.io.l2_pf_addr.valid && l2_in_pmem && enable && pf_ctrl.enable
   io.l2_req.bits := pf_queue_filter.io.l2_pf_addr.bits
 
-  val l3_in_pmem = PmemRanges.map(_.cover(pf_queue_filter.io.l3_pf_addr.bits)).reduce(_ || _)
+  val l3_in_pmem = PmemRanges.map(_.cover(pf_queue_filter.io.l3_pf_addr.bits.addr)).reduce(_ || _)
   io.l3_req.valid := pf_queue_filter.io.l3_pf_addr.valid && l3_in_pmem && enable && pf_ctrl.enable
   io.l3_req.bits := pf_queue_filter.io.l3_pf_addr.bits
 }
diff --git a/src/main/scala/xiangshan/mem/prefetch/PrefetcherWrapper.scala b/src/main/scala/xiangshan/mem/prefetch/PrefetcherWrapper.scala
new file mode 100644
index 00000000000..b8c52fb871c
--- /dev/null
+++ b/src/main/scala/xiangshan/mem/prefetch/PrefetcherWrapper.scala
@@ -0,0 +1,303 @@
+package xiangshan.mem.prefetch
+
+import chisel3._
+import chisel3.util._
+import org.chipsalliance.cde.config.Parameters
+import utility._
+import utils._
+import xiangshan._
+import xiangshan.backend.Bundles.DynInst
+import xiangshan.backend.fu.{FuType, PMPRespBundle}
+import xiangshan.cache.{DCacheLineReq, HasDCacheParameters, LoadPfDbBundle}
+import xiangshan.cache.mmu._
+import xiangshan.mem._
+import xiangshan.mem.Bundles.LsPrefetchTrainBundle
+import xiangshan.mem.trace._
+import freechips.rocketchip.diplomacy.BufferParams.default
+import utility.mbist.{MbistInterface, MbistPipeline}
+
+object TLBPlace extends Enumeration{
+  /* now only support dtlb_ld and dtlb_st */
+  // val none = Value("none")
+  // val itlb = Value("itlb")
+  val dtlb_ld = Value("dtlb_ld")
+  // val dtlb_st = Value("dtlb_st")
+  val dtlb_pf = Value("dtlb_pf")
+  // val l2tlb = Value("l2tlb")
+}
+
+trait PrefetcherParams {
+  def name: String
+  def tlbPlace = TLBPlace.dtlb_pf
+  def TRAIN_FILTER_SIZE = 6
+  def PREFETCH_FILTER_SIZE = 16
+}
+
+trait HasPrefetcherParams extends HasXSParameter{
+  def LD_TRAIN_WIDTH = backendParams.LdExuCnt
+  def ST_TRAIN_WIDTH = backendParams.StaExuCnt
+
+  // You can set the unified interface to N, and the invalid that you don't need can be set to 0
+  val L1_PF_REG_CNT = 1
+  val L2_PF_REG_CNT = 2
+  val L3_PF_REG_CNT = 4
+}
+
+abstract class PrefetchModule(implicit p: Parameters) extends XSModule
+  with HasPrefetcherParams with HasDCacheParameters
+
+abstract class PrefetchBundle(implicit p: Parameters) extends XSBundle
+  with HasPrefetcherParams with HasDCacheParameters
+
+
+class PrefetchCtrlFromTile(implicit p: Parameters) extends PrefetchBundle {
+  val l2PfqBusy = Bool()
+}
+
+class OOOToPrefetchIO(implicit p: Parameters) extends PrefetchBundle {
+  // for s1 and only FirstIssue
+  // Because of the pointer chase, the PC can only determine whose it is in S1
+  val s1_loadPc = Vec(LD_TRAIN_WIDTH, Output(UInt(VAddrBits.W)))
+  val s1_storePc = Vec(ST_TRAIN_WIDTH, Output(UInt(VAddrBits.W)))
+}
+
+class DCacheToPrefetchIO(implicit p: Parameters) extends PrefetchBundle {
+  val sms_agt_evict_req = DecoupledIO(new AGTEvictReq())
+}
+
+class TrainSourceIO(implicit p: Parameters) extends PrefetchBundle {
+  // load: ldu + hyu
+  val s1_loadFireHint = Output(Vec(LD_TRAIN_WIDTH, Bool()))
+  val s2_loadFireHint = Output(Vec(LD_TRAIN_WIDTH, Bool()))
+  val s3_load = Output(Vec(LD_TRAIN_WIDTH, ValidIO(new LsPrefetchTrainBundle())))
+  val s3_ptrChasing = Output(Vec(LD_TRAIN_WIDTH, Bool()))
+
+  // store: stu + hyu
+  val s1_storeFireHint = Output(Vec(ST_TRAIN_WIDTH, Bool()))
+  val s2_storeFireHint = Output(Vec(ST_TRAIN_WIDTH, Bool()))
+  val s3_store = Output(Vec(ST_TRAIN_WIDTH, ValidIO(new LsPrefetchTrainBundle())))
+}
+
+class PrefetcherWrapper(implicit p: Parameters) extends PrefetchModule {
+  println(s"PrefetcherWrapper:")
+  prefetcherSeq.zipWithIndex.foreach { case (pf, idx) =>
+    println(f"  Prefetcher #$idx%-2d => ${pf.name}")
+  }
+  val io = IO(new Bundle() {
+    // prefetch control
+    val pfCtrlFromTile = Input(new PrefetchCtrlFromTile)
+    val pfCtrlFromCSR = Flipped(new PrefetchCtrl)
+    val pfCtrlFromDCache = Flipped(new PrefetchControlBundle)
+    // replenish information
+    val fromDCache = Flipped(new DCacheToPrefetchIO)
+    val fromOOO = Flipped(new OOOToPrefetchIO)
+    // train
+    val trainSource = Flipped(new TrainSourceIO)
+    // tlb
+    val tlb_req = Vec(prefetcherNum, new TlbRequestIO(nRespDups = 2))
+    val pmp_resp = Vec(prefetcherNum, Flipped(new PMPRespBundle()))
+    // prefetch req sender
+    val l1_pf_to_l1 = DecoupledIO(new L1PrefetchReq())
+    val l1_pf_to_l2 = Output(new coupledL2.PrefetchRecv())
+    val l1_pf_to_l3 = Output(new huancun.PrefetchRecv())
+  })
+
+  def isLoadAccess(uop: DynInst): Bool = FuType.isLoad(uop.fuType) || FuType.isVLoad(uop.fuType)
+  def isStoreAccess(uop: DynInst): Bool = FuType.isStore(uop.fuType)
+
+  val hartId = p(XSCoreParamsKey).HartId
+  val l1D_pf_enable = RegNextN(io.pfCtrlFromCSR.l1D_pf_enable, 2, Some(true.B))
+  val pf_train_on_hit = RegNextN(io.pfCtrlFromCSR.l1D_pf_train_on_hit, 2, Some(true.B))
+  val s2_loadPcVec = (0 until LD_TRAIN_WIDTH).map{ i=>
+    RegEnable(io.fromOOO.s1_loadPc(i), io.trainSource.s1_loadFireHint(i))
+  }
+  val s3_loadPcVec = (0 until LD_TRAIN_WIDTH).map { i =>
+    RegEnable(s2_loadPcVec(i), io.trainSource.s2_loadFireHint(i))
+  }
+  val s2_storePcVec = (0 until ST_TRAIN_WIDTH).map{ i=>
+    RegEnable(io.fromOOO.s1_storePc(i), io.trainSource.s1_storeFireHint(i))
+  }
+  val s3_storePcVec = (0 until ST_TRAIN_WIDTH).map { i =>
+    RegEnable(s2_storePcVec(i), io.trainSource.s2_storeFireHint(i))
+  }
+  /* prefetch arbiter */
+  val l1_pf_arb = Module(new RRArbiterInit(new L1PrefetchReq, prefetcherNum))
+  val l2_pf_req = Wire(Decoupled(new L2PrefetchReq()))
+  val l2_pf_arb = Module(new RRArbiterInit(new L2PrefetchReq, prefetcherNum))
+  val l3_pf_req = Wire(Decoupled(new L3PrefetchReq()))
+  val l3_pf_arb = Module(new RRArbiterInit(new L3PrefetchReq, prefetcherNum))
+
+  // init
+  io.tlb_req.foreach{ x =>
+    x.req.valid := false.B
+    x.req_kill := false.B
+    x.req.bits := DontCare
+    x.resp.ready := true.B
+  }
+  l1_pf_arb.io.in.foreach{ x =>
+    x.valid := false.B
+    x.bits := DontCare
+  }
+  l2_pf_arb.io.in.foreach{ x =>
+    x.valid := false.B
+    x.bits := DontCare
+  }
+  l3_pf_arb.io.in.foreach{ x =>
+    x.valid := false.B
+    x.bits := DontCare
+  }
+
+
+  /** Prefetchor
+    * L1: Stride
+    * L2: Stride, SMS
+    * L3: Stride
+    */
+  val HasSMS = prefetcherSeq.exists(_.isInstanceOf[SMSParams])
+  val IdxSMS = prefetcherSeq.indexWhere(_.isInstanceOf[SMSParams])
+  val smsOpt: Option[SMSPrefetcher] = if(HasSMS) Some(Module(new SMSPrefetcher())) else None
+  smsOpt.foreach (pf => {
+    val enableSMS = Constantin.createRecord(s"enableSMS$hartId", initValue = true)
+    // constantinCtrl && master switch csrCtrl && single switch csrCtrl
+    pf.io.enable := enableSMS && l1D_pf_enable &&
+      GatedRegNextN(io.pfCtrlFromCSR.l2_pf_recv_enable, 2, Some(false.B))
+    pf.io_agt_en := GatedRegNextN(io.pfCtrlFromCSR.l1D_pf_enable_agt, 2, Some(false.B))
+    pf.io_pht_en := GatedRegNextN(io.pfCtrlFromCSR.l1D_pf_enable_pht, 2, Some(false.B))
+    pf.io_act_threshold := GatedRegNextN(io.pfCtrlFromCSR.l1D_pf_active_threshold, 2, Some(12.U))
+    pf.io_act_stride := GatedRegNextN(io.pfCtrlFromCSR.l1D_pf_active_stride, 2, Some(30.U))
+    pf.io_stride_en := false.B
+    pf.io_dcache_evict <> io.fromDCache.sms_agt_evict_req
+    val mbistSmsPl = MbistPipeline.PlaceMbistPipeline(1, "MbistPipeSms", hasMbist)
+
+    for (i <- 0 until LD_TRAIN_WIDTH) {
+      val source = io.trainSource.s3_load(i)
+      val primaryValid = source.valid && (!source.bits.tlbMiss || source.bits.is_from_hw_pf)
+      pf.io.ld_in(i).valid := Mux(
+        pf_train_on_hit,
+        primaryValid,
+        primaryValid && source.bits.isFirstIssue && source.bits.miss
+      ) // && isLoadAccess(source.bits.uop)
+      pf.io.ld_in(i).bits := source.bits
+      pf.io.ld_in(i).bits.uop.pc := Mux(
+        io.trainSource.s3_ptrChasing(i),
+        s2_loadPcVec(i),
+        s3_loadPcVec(i)
+      )
+    }
+
+    for (i <- 0 until ST_TRAIN_WIDTH) {
+      val source = io.trainSource.s3_store(i)
+      val primaryValid = source.valid && (!source.bits.tlbMiss || source.bits.is_from_hw_pf)
+      pf.io.st_in(i).valid := Mux(
+        pf_train_on_hit,
+        primaryValid,
+        primaryValid && source.bits.isFirstIssue && source.bits.miss
+      ) // && isStoreAccess(source.bits.uop)
+      pf.io.st_in(i).bits := source.bits
+      pf.io.st_in(i).bits.uop.pc := s3_storePcVec(i)
+    }
+
+    io.tlb_req(IdxSMS) <> pf.io.tlb_req
+    pf.io.pmp_resp := io.pmp_resp(IdxSMS)
+
+    l2_pf_arb.io.in(IdxSMS) <> pf.io.l2_req
+    pf.io.l1_req.ready := false.B
+    pf.io.l3_req.ready := false.B
+  })
+
+  val HasStreamStride = prefetcherSeq.exists(_.isInstanceOf[StreamStrideParams])
+  val IdxStreamStride = prefetcherSeq.indexWhere(_.isInstanceOf[StreamStrideParams])
+  val strideOpt: Option[L1Prefetcher] = if(HasStreamStride) Some(Module(new L1Prefetcher())) else None
+  strideOpt.foreach(pf => {
+    val enableL1StreamPrefetcher = Constantin.createRecord(s"enableL1StreamPrefetcher$hartId", initValue = true)
+    // constantinCtrl && master switch csrCtrl && single switch csrCtrl
+    pf.io.enable := enableL1StreamPrefetcher && l1D_pf_enable &&
+      GatedRegNextN(io.pfCtrlFromCSR.l1D_pf_enable_stride, 2, Some(false.B))
+
+    pf.pf_ctrl <> io.pfCtrlFromDCache
+    pf.l2PfqBusy := io.pfCtrlFromTile.l2PfqBusy
+
+    // stride will train on miss or prefetch hit
+    for(i <- 0 until LD_TRAIN_WIDTH){
+      val source = io.trainSource.s3_load(i)
+      pf.stride_train(i).valid := source.valid && source.bits.isFirstIssue && (
+        source.bits.miss || isFromStride(source.bits.meta_prefetch)
+      ) // && isLoadAccess(source.bits.uop)
+      pf.stride_train(i).bits := source.bits
+      pf.stride_train(i).bits.uop.pc := Mux(
+        io.trainSource.s3_ptrChasing(i),
+        s2_loadPcVec(i),
+        s3_loadPcVec(i)
+      )
+      pf.io.ld_in(i).valid := source.valid && source.bits.isFirstIssue && isLoadAccess(source.bits.uop)
+      pf.io.ld_in(i).bits := source.bits
+    }
+
+    for (i <- 0 until ST_TRAIN_WIDTH) {
+      pf.io.st_in(i).valid := false.B
+      pf.io.st_in(i).bits := DontCare
+    }
+
+    io.tlb_req(IdxStreamStride) <> pf.io.tlb_req
+    pf.io.pmp_resp := io.pmp_resp(IdxStreamStride)
+
+    l1_pf_arb.io.in(IdxStreamStride) <> pf.io.l1_req
+    l2_pf_arb.io.in(IdxStreamStride) <> pf.io.l2_req
+    l3_pf_arb.io.in(IdxStreamStride) <> pf.io.l3_req
+  })
+
+  /**
+   * load prefetch to l1 Dcache
+   * stride
+   */
+  io.l1_pf_to_l1 <> Pipeline(in = l1_pf_arb.io.out, depth = L1_PF_REG_CNT, pipe = false, name = Some("pf_to_ldu_reg"))
+
+  /** load/store prefetch to l2 cache
+   *  stride, sms
+   */
+  l2_pf_req <> Pipeline(in = l2_pf_arb.io.out, depth = L2_PF_REG_CNT, pipe = false, name = Some("pf_to_l2cache_reg"))
+  l2_pf_req.ready := true.B
+
+  // load/store prefetch to l3 cache
+  l3_pf_req <> Pipeline(in = l3_pf_arb.io.out, depth = L3_PF_REG_CNT, pipe = false, name = Some("pf_to_l3cache_reg"))
+  l3_pf_req.ready := true.B
+
+  io.fromDCache.sms_agt_evict_req.ready := false.B
+
+  io.l1_pf_to_l2.addr_valid := l2_pf_req.valid
+  io.l1_pf_to_l2.addr := l2_pf_req.bits.addr
+  io.l1_pf_to_l2.pf_source := l2_pf_req.bits.source
+  io.l1_pf_to_l2.l2_pf_en := RegNextN(io.pfCtrlFromCSR.l2_pf_enable, L2_PF_REG_CNT, Some(true.B))
+  io.l1_pf_to_l3.addr_valid := l3_pf_req.valid
+  io.l1_pf_to_l3.addr := l3_pf_req.bits.addr
+  io.l1_pf_to_l3.l2_pf_en := RegNextN(io.pfCtrlFromCSR.l2_pf_enable, L3_PF_REG_CNT, Some(true.B))
+
+  val l2_trace = Wire(new LoadPfDbBundle)
+  l2_trace.paddr := l2_pf_req.bits.addr
+  val l2_trace_table = ChiselDB.createTable(s"L2PrefetchTrace$hartId", new LoadPfDbBundle, basicDB = false)
+  when(l2_pf_req.bits.source === MemReqSource.Prefetch2L2Stream.id.U || l2_pf_req.bits.source === MemReqSource.Prefetch2L2Stride.id.U) {
+    l2_trace_table.log(l2_trace, l2_pf_req.valid, "L2StreamStride", clock, reset) 
+  }.elsewhen(l2_pf_req.bits.source === MemReqSource.Prefetch2L2SMS.id.U) {
+    l2_trace_table.log(l2_trace, l2_pf_req.valid, "L2SMS", clock, reset)
+  }.otherwise {
+    l2_trace_table.log(l2_trace, l2_pf_req.valid, "L2Unknown", clock, reset)
+  }
+
+  val l3_trace = Wire(new LoadPfDbBundle)
+  l3_trace.paddr := l3_pf_req.bits.addr
+  val l3_trace_table = ChiselDB.createTable(s"L3PrefetchTrace$hartId", new LoadPfDbBundle, basicDB = false)
+  when(l3_pf_req.bits.source === MemReqSource.Prefetch2L3Stream.id.U || l3_pf_req.bits.source === MemReqSource.Prefetch2L3Stride.id.U) {
+    l3_trace_table.log(l3_trace, l3_pf_req.valid, "L3StreamStride", clock, reset)
+  }.otherwise {
+    l3_trace_table.log(l3_trace, l3_pf_req.valid, "L3Unknown", clock, reset)
+  }
+
+  val arb_seq = Seq(l1_pf_arb, l2_pf_arb, l3_pf_arb)
+  arb_seq.zipWithIndex.foreach { case (arb, level) =>
+    prefetcherSeq.zipWithIndex.foreach { case (pf, idx) =>
+      XSPerfAccumulate(s"${pf.name}_fire_l${level+1}", arb.io.in(idx).fire)
+      XSPerfAccumulate(s"${pf.name}_block_l${level+1}", arb.io.in(idx).valid && !arb.io.in(idx).ready)
+    }
+  }
+
+}
diff --git a/src/main/scala/xiangshan/mem/prefetch/SMSPrefetcher.scala b/src/main/scala/xiangshan/mem/prefetch/SMSPrefetcher.scala
index 36af36ea983..7703f339eb8 100644
--- a/src/main/scala/xiangshan/mem/prefetch/SMSPrefetcher.scala
+++ b/src/main/scala/xiangshan/mem/prefetch/SMSPrefetcher.scala
@@ -55,11 +55,17 @@ case class SMSParams
   pht_lookup_queue_size: Int = 4,
   pf_filter_size: Int = 16,
   train_filter_size: Int = 8
-) extends PrefetcherParams
+) extends PrefetcherParams {
+  override def name = "SMS"
+}
 
 trait HasSMSModuleHelper extends HasCircularQueuePtrHelper with HasDCacheParameters
 { this: HasXSParameter =>
-  val smsParams = coreParams.prefetcher.get.asInstanceOf[SMSParams]
+  def smsParams = coreParams.prefetcher.find {
+    case p: SMSParams => true
+    case _ => false
+  }.get.asInstanceOf[SMSParams]
+
   val BLK_ADDR_WIDTH = VAddrBits - log2Up(dcacheParameters.blockBytes)
   val REGION_SIZE = smsParams.region_size
   val REGION_BLKS = smsParams.region_size / dcacheParameters.blockBytes
