diff --git a/src/main/scala/xiangshan/mem/lsqueue/LoadExceptionBuffer.scala b/src/main/scala/xiangshan/mem/lsqueue/LoadExceptionBuffer.scala
index 854f4560435..3b080d28ea3 100644
--- a/src/main/scala/xiangshan/mem/lsqueue/LoadExceptionBuffer.scala
+++ b/src/main/scala/xiangshan/mem/lsqueue/LoadExceptionBuffer.scala
@@ -64,12 +64,6 @@ class LqExceptionBuffer(implicit p: Parameters) extends XSModule with HasCircula
     s2_enqueue(w) := s2_valid(w) && s2_has_exception(w)
   }
 
-  when (req_valid && req.uop.robIdx.needFlush(io.redirect)) {
-    req_valid := s2_enqueue.asUInt.orR
-  } .elsewhen (s2_enqueue.asUInt.orR) {
-    req_valid := true.B
-  }
-
   def selectOldest[T <: LqWriteBundle](valid: Seq[Bool], bits: Seq[T]): (Seq[Bool], Seq[T]) = {
     assert(valid.length == bits.length)
     if (valid.length == 0 || valid.length == 1) {
@@ -81,8 +75,7 @@ class LqExceptionBuffer(implicit p: Parameters) extends XSModule with HasCircula
         res(i).bits := bits(i)
       }
       val oldest = Mux(valid(0) && valid(1),
-        Mux(isAfter(bits(0).uop.robIdx, bits(1).uop.robIdx) ||
-          (bits(0).uop.robIdx === bits(1).uop.robIdx && bits(0).uop.uopIdx > bits(1).uop.uopIdx), res(1), res(0)),
+        Mux(isAfter(bits(0).uop.lqIdx, bits(1).uop.lqIdx), res(1), res(0)),
         Mux(valid(0) && !valid(1), res(0), res(1)))
       (Seq(oldest.valid), Seq(oldest.bits))
     } else {
@@ -92,16 +85,11 @@ class LqExceptionBuffer(implicit p: Parameters) extends XSModule with HasCircula
     }
   }
 
-  val reqSel = selectOldest(s2_enqueue, s2_req)
+  val reqValid = req_valid && !req.uop.robIdx.needFlush(io.redirect)
+  val reqSel = selectOldest(s2_enqueue :+ reqValid, s2_req :+ req)
 
-  when (req_valid) {
-    req := Mux(
-      reqSel._1(0) && (isAfter(req.uop.robIdx, reqSel._2(0).uop.robIdx) || (req.uop.robIdx === reqSel._2(0).uop.robIdx && req.uop.uopIdx > reqSel._2(0).uop.uopIdx)),
-      reqSel._2(0),
-      req)
-  } .elsewhen (s2_enqueue.asUInt.orR) {
-    req := reqSel._2(0)
-  }
+  req_valid := reqSel._1(0)
+  req := reqSel._2(0)
 
   io.exceptionAddr.vaddr  := req.fullva
   io.exceptionAddr.vaNeedExt := req.vaNeedExt
diff --git a/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala b/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala
index ca4ff7e5959..985116977d8 100644
--- a/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala
+++ b/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala
@@ -105,12 +105,6 @@ class StoreExceptionBuffer(implicit p: Parameters) extends XSModule with HasCirc
     s2_enqueue(w) := s2_valid(w)
   }
 
-  when (req_valid && req.uop.robIdx.needFlush(io.redirect)) {
-    req_valid := s2_enqueue.asUInt.orR
-  }.elsewhen (s2_enqueue.asUInt.orR) {
-    req_valid := true.B
-  }
-
   def selectOldest[T <: LsPipelineBundle](valid: Seq[Bool], bits: Seq[T]): (Seq[Bool], Seq[T]) = {
     assert(valid.length == bits.length)
     if (valid.length == 0 || valid.length == 1) {
@@ -122,8 +116,7 @@ class StoreExceptionBuffer(implicit p: Parameters) extends XSModule with HasCirc
         res(i).bits := bits(i)
       }
       val oldest = Mux(valid(0) && valid(1),
-        Mux(isAfter(bits(0).uop.robIdx, bits(1).uop.robIdx) ||
-          (bits(0).uop.robIdx === bits(1).uop.robIdx && bits(0).uop.uopIdx > bits(1).uop.uopIdx), res(1), res(0)),
+        Mux(isAfter(bits(0).uop.sqIdx, bits(1).uop.sqIdx), res(1), res(0)),
         Mux(valid(0) && !valid(1), res(0), res(1)))
       (Seq(oldest.valid), Seq(oldest.bits))
     } else {
@@ -133,16 +126,11 @@ class StoreExceptionBuffer(implicit p: Parameters) extends XSModule with HasCirc
     }
   }
 
-  val reqSel = selectOldest(s2_enqueue, s2_req)
+  val reqValid = req_valid && !req.uop.robIdx.needFlush(io.redirect)
+  val reqSel = selectOldest(s2_enqueue :+ reqValid, s2_req :+ req)
 
-  when (req_valid) {
-    req := Mux(
-      reqSel._1(0) && (isAfter(req.uop.robIdx, reqSel._2(0).uop.robIdx) || (isNotBefore(req.uop.robIdx, reqSel._2(0).uop.robIdx) && req.uop.uopIdx > reqSel._2(0).uop.uopIdx)),
-      reqSel._2(0),
-      req)
-  } .elsewhen (s2_enqueue.asUInt.orR) {
-    req := reqSel._2(0)
-  }
+  req_valid := reqSel._1(0)
+  req := reqSel._2(0)
 
   io.exceptionAddr.vaddr     := req.fullva
   io.exceptionAddr.vaNeedExt := req.vaNeedExt
diff --git a/src/main/scala/xiangshan/mem/vector/VMergeBuffer.scala b/src/main/scala/xiangshan/mem/vector/VMergeBuffer.scala
index 060c3518376..dfe8a2c8ca0 100644
--- a/src/main/scala/xiangshan/mem/vector/VMergeBuffer.scala
+++ b/src/main/scala/xiangshan/mem/vector/VMergeBuffer.scala
@@ -108,6 +108,8 @@ abstract class BaseVMergeBuffer(isVStore: Boolean=false)(implicit p: Parameters)
     val hasExp                               = ExceptionNO.selectByFu(source.exceptionVec, fuCfg).asUInt.orR
     sink.robidx                             := source.uop.robIdx
     sink.uopidx                             := source.uop.uopIdx
+    sink.sqIdx                              := source.uop.sqIdx
+    sink.lqIdx                              := source.uop.lqIdx
     sink.feedback(VecFeedbacks.COMMIT)      := !hasExp
     sink.feedback(VecFeedbacks.FLUSH)       := hasExp
     sink.feedback(VecFeedbacks.LAST)        := true.B
diff --git a/src/main/scala/xiangshan/mem/vector/VecBundle.scala b/src/main/scala/xiangshan/mem/vector/VecBundle.scala
index a12afef06c0..ddf13fc2aec 100644
--- a/src/main/scala/xiangshan/mem/vector/VecBundle.scala
+++ b/src/main/scala/xiangshan/mem/vector/VecBundle.scala
@@ -205,6 +205,8 @@ class FeedbackToSplitIO(implicit p: Parameters) extends VLSUBundle{
 class FeedbackToLsqIO(implicit p: Parameters) extends VLSUBundle{
   val robidx = new RobPtr
   val uopidx = UopIdx()
+  val sqIdx = new SqPtr
+  val lqIdx = new LqPtr
   val vaddr = UInt(XLEN.W)
   val vaNeedExt = Bool()
   val gpaddr = UInt(GPAddrBits.W)
