diff --git a/src/main/scala/xiangshan/backend/fu/NewCSR/InterruptFilter.scala b/src/main/scala/xiangshan/backend/fu/NewCSR/InterruptFilter.scala
index c5b1ee62f97..d70a0f7c921 100644
--- a/src/main/scala/xiangshan/backend/fu/NewCSR/InterruptFilter.scala
+++ b/src/main/scala/xiangshan/backend/fu/NewCSR/InterruptFilter.scala
@@ -52,6 +52,8 @@ class InterruptFilter extends Module {
 
   val mtopiIsNotZero: Bool = (mip & mie & (~mideleg).asUInt) =/= 0.U
   val stopiIsNotZero: Bool = (hsip & hsie & (~hideleg).asUInt) =/= 0.U
+  val mtopiIsNotZeroReg: Bool = GatedValidRegNext(mtopiIsNotZero)
+  val stopiIsNotZeroReg: Bool = GatedValidRegNext(stopiIsNotZero)
 
   val NoSEIMask = (~(BigInt(1) << InterruptNO.SEI).U(64.W)).asUInt
   val mtopigather = mip & mie & (~mideleg).asUInt
@@ -257,8 +259,8 @@ class InterruptFilter extends Module {
   private val hsIidDefaultPrioLowSEI : Bool = hsIidIdxReg > seiPrioIdx
 
   // update mtopi
-  io.out.mtopi.IID := Mux(mtopiIsNotZero, mIidNum, 0.U)
-  io.out.mtopi.IPRIO := Mux(mtopiIsNotZero,
+  io.out.mtopi.IID := Mux(mtopiIsNotZeroReg, mIidNum, 0.U)
+  io.out.mtopi.IPRIO := Mux(mtopiIsNotZeroReg,
     Mux1H(Seq(
       (!mipriosRegTmp.isZero && !mipriosRegTmp.greaterThan255) -> mipriosRegTmp.prioNum,
       (mipriosRegTmp.greaterThan255 || mipriosRegTmp.isZero && mIidDefaultPrioLowMEI) -> 255.U,
@@ -268,8 +270,8 @@ class InterruptFilter extends Module {
   )
 
   // upadte stopi
-  io.out.stopi.IID := Mux(stopiIsNotZero, hsIidNum, 0.U)
-  io.out.stopi.IPRIO := Mux(stopiIsNotZero,
+  io.out.stopi.IID := Mux(stopiIsNotZeroReg, hsIidNum, 0.U)
+  io.out.stopi.IPRIO := Mux(stopiIsNotZeroReg,
     Mux1H(Seq(
       (!hsipriosRegTmp.isZero && !hsipriosRegTmp.greaterThan255) -> hsipriosRegTmp.prioNum,
       (hsipriosRegTmp.greaterThan255 || hsipriosRegTmp.isZero && hsIidDefaultPrioLowSEI) -> 255.U,
@@ -305,6 +307,23 @@ class InterruptFilter extends Module {
   val C3C4Enable   = Candidate3 & Candidate4
   val C3C5Enable   = Candidate3 & Candidate5
 
+  val CandidateNoValidReg: Bool = GatedValidRegNext(CandidateNoValid)
+  val Candidate123Reg = GatedValidRegNext(Candidate123)
+  val NoCandidate45Reg = GatedValidRegNext(!Candidate45)
+  val onlyC1EnableReg = GatedValidRegNext(onlyC1Enable)
+  val onlyC2EnableReg = GatedValidRegNext(onlyC2Enable)
+  val onlyC3EnableReg = GatedValidRegNext(onlyC3Enable)
+  val onlyC4EnableReg = GatedValidRegNext(onlyC4Enable)
+  val onlyC5EnableReg = GatedValidRegNext(onlyC5Enable)
+  val C1C4EnableReg = GatedValidRegNext(C1C4Enable)
+  val C1C5EnableReg = GatedValidRegNext(C1C5Enable)
+  val C2C4EnableReg = GatedValidRegNext(C2C4Enable)
+  val C3C4EnableReg = GatedValidRegNext(C3C4Enable)
+  val C3C5EnableReg = GatedValidRegNext(C3C5Enable)
+
+  val hvictlReg = RegNext(hvictl)
+  val vstopeiReg = RegNext(vstopei)
+
   val iidOnlyC1 = Wire(UInt(12.W))
   val iidOnlyC4 = Wire(UInt(12.W))
   val iidOnlyC5 = Wire(UInt(12.W))
@@ -326,7 +345,7 @@ class InterruptFilter extends Module {
 
   iidOnlyC1 := InterruptNO.SEI.U
   iidOnlyC4 := vsIidNum
-  iidOnlyC5 := hvictl.IID.asUInt
+  iidOnlyC5 := hvictlReg.IID.asUInt
 
   val iidC4Idx = Wire(UInt(12.W))
   iidC4Idx := vsIidIdxReg
@@ -338,16 +357,16 @@ class InterruptFilter extends Module {
   val iprioC4Tmp = Wire(UInt(8.W))
   val iprioC3C5Tmp = Wire(UInt(8.W))
 
-  val hvictlDPR = Mux(hvictl.DPR.asBool, 255.U, 0.U)
-  val C1GreaterThan255 = vstopei.IPRIO.asUInt(10, 8).orR
+  val hvictlDPR = Mux(hvictlReg.DPR.asBool, 255.U, 0.U)
+  val C1GreaterThan255 = vstopeiReg.IPRIO.asUInt(10, 8).orR
   val C4IsZero = !hvipriosRegTmp.prioNum.orR
-  val C2C5IsZero = !hvictl.IPRIO.asUInt.orR
+  val C2C5IsZero = !hvictlReg.IPRIO.asUInt.orR
   val C4HighVSEI = iidC4Idx < findIndex(InterruptNO.VSEI.U)
   val SEIHighC4 = findIndex(InterruptNO.SEI.U) < iidC4Idx
   val iprioC1GreaterThan255 = Mux(C1GreaterThan255, 255.U, iprioC1Tmp)
 
-  iprioC1 := vstopei.IPRIO.asUInt
-  iprioC2C5 := hvictl.IPRIO.asUInt
+  iprioC1 := vstopeiReg.IPRIO.asUInt
+  iprioC2C5 := hvictlReg.IPRIO.asUInt
   iprioC4 := hvipriosRegTmp.prioNum
 
   iprioC1Tmp := iprioC1(7, 0)
@@ -376,14 +395,14 @@ class InterruptFilter extends Module {
   }
 
   // C1,C5 enable
-  iidC1C5 := Mux(hvictl.DPR.asBool, iidOnlyC1, iidOnlyC5)
+  iidC1C5 := Mux(hvictlReg.DPR.asBool, iidOnlyC1, iidOnlyC5)
   when(C2C5IsZero) {
-    iprioC1C5 := Mux(hvictl.DPR.asBool, iprioC1GreaterThan255, 0.U)
+    iprioC1C5 := Mux(hvictlReg.DPR.asBool, iprioC1GreaterThan255, 0.U)
   }.elsewhen(iprioC1 < iprioC2C5) {
     iidC1C5 := iidOnlyC1
     iprioC1C5 := iprioC1Tmp
   }.elsewhen(iprioC1 === iprioC2C5) {
-    iprioC1C5 := Mux(hvictl.DPR.asBool, iprioC1Tmp, iprioC2C5)
+    iprioC1C5 := Mux(hvictlReg.DPR.asBool, iprioC1Tmp, iprioC2C5)
   }.otherwise {
     iidC1C5 := iidOnlyC5
     iprioC1C5 := iprioC3C5Tmp
@@ -408,37 +427,37 @@ class InterruptFilter extends Module {
   iidC3C4 := Mux(C4IsZero, Mux(C4HighVSEI, iidOnlyC4, iidOnlyC1), iidOnlyC4)
   iprioC3C4 := iprioC4Tmp
   // C3,C5 enable
-  iidC3C5 := Mux(C2C5IsZero, Mux(hvictl.DPR.asBool, iidOnlyC5, iidOnlyC1), iidOnlyC5)
+  iidC3C5 := Mux(C2C5IsZero, Mux(hvictlReg.DPR.asBool, iidOnlyC5, iidOnlyC1), iidOnlyC5)
   iprioC3C5 := iprioC3C5Tmp
 
   // update vstopi
-  io.out.vstopi.IID := Mux(CandidateNoValid,
+  io.out.vstopi.IID := Mux(CandidateNoValidReg,
     0.U,
     Mux1H(Seq(
-      (Candidate123 & !Candidate45) -> iidOnlyC1,
-      onlyC4Enable -> iidOnlyC4,
-      onlyC5Enable -> iidOnlyC5,
-      C1C4Enable -> iidC1C4,
-      C1C5Enable -> iidC1C5,
-      C2C4Enable -> iidC2C4,
-      C3C4Enable -> iidC3C4,
-      C3C5Enable -> iidC3C5,
+      (Candidate123Reg & NoCandidate45Reg) -> iidOnlyC1,
+      onlyC4EnableReg -> iidOnlyC4,
+      onlyC5EnableReg -> iidOnlyC5,
+      C1C4EnableReg -> iidC1C4,
+      C1C5EnableReg -> iidC1C5,
+      C2C4EnableReg -> iidC2C4,
+      C3C4EnableReg -> iidC3C4,
+      C3C5EnableReg -> iidC3C5,
     ))
   )
-  io.out.vstopi.IPRIO := Mux(CandidateNoValid,
+  io.out.vstopi.IPRIO := Mux(CandidateNoValidReg,
     0.U,
-    Mux(!hvictl.IPRIOM.asBool, 1.U,
+    Mux(!hvictlReg.IPRIOM.asBool, 1.U,
       Mux1H(Seq(
-        onlyC1Enable -> iprioOnlyC1,
-        onlyC2Enable -> iprioOnlyC2,
-        onlyC3Enable -> iprioOnlyC3,
-        onlyC4Enable -> iprioOnlyC4,
-        onlyC5Enable -> iprioOnlyC5,
-        C1C4Enable -> iprioC1C4,
-        C1C5Enable -> iprioC1C5,
-        C2C4Enable -> iprioC2C4,
-        C3C4Enable -> iprioC3C4,
-        C3C5Enable -> iprioC3C5,
+        onlyC1EnableReg -> iprioOnlyC1,
+        onlyC2EnableReg -> iprioOnlyC2,
+        onlyC3EnableReg -> iprioOnlyC3,
+        onlyC4EnableReg -> iprioOnlyC4,
+        onlyC5EnableReg -> iprioOnlyC5,
+        C1C4EnableReg -> iprioC1C4,
+        C1C5EnableReg -> iprioC1C5,
+        C2C4EnableReg -> iprioC2C4,
+        C3C4EnableReg -> iprioC3C4,
+        C3C5EnableReg -> iprioC3C5,
       ))
     )
   )
@@ -518,8 +537,8 @@ class InterruptFilter extends Module {
 
   // virtual interrupt with hvictl injection
   val vsIRModeCond = privState.isModeVS && vsstatusSIE || privState < PrivState.ModeVS
-  val SelectCandidate5 = onlyC5Enable || C3C5Enable ||
-                         C1C5Enable && (iprioC1 === iprioC2C5 && !hvictl.DPR.asBool || iprioC1 > iprioC2C5)
+  val SelectCandidate5 = onlyC5EnableReg || C3C5EnableReg ||
+                         C1C5EnableReg && (iprioC1 === iprioC2C5 && !hvictlReg.DPR.asBool || iprioC1 > iprioC2C5)
   // delay at least 6 cycles to maintain the atomic of sret/mret
   // 65bit indict current interrupt is NMI
   val intrVecReg = RegInit(0.U(8.W))
