diff --git a/src/main/scala/xiangshan/frontend/bpu/Bpu.scala b/src/main/scala/xiangshan/frontend/bpu/Bpu.scala
index 89ca30b2f52..a724a7ba489 100644
--- a/src/main/scala/xiangshan/frontend/bpu/Bpu.scala
+++ b/src/main/scala/xiangshan/frontend/bpu/Bpu.scala
@@ -80,7 +80,7 @@ class Bpu(implicit p: Parameters) extends BpuModule with HalfAlignHelper {
   abtb.io.enable        := ctrl.abtbEnable
   mbtb.io.enable        := ctrl.mbtbEnable
   tage.io.enable        := ctrl.tageEnable
-  sc.io.enable          := false.B
+  sc.io.enable          := ctrl.scEnable
   ittage.io.enable      := ctrl.ittageEnable
   ras.io.enable         := false.B
 
@@ -214,6 +214,8 @@ class Bpu(implicit p: Parameters) extends BpuModule with HalfAlignHelper {
   sc.io.foldedPathHist      := phr.io.s1_foldedPhr
   sc.io.trainFoldedPathHist := phr.io.trainFoldedPhr
   sc.io.train.valid         := train.valid
+  private val scTakenMask = sc.io.takenMask
+  dontTouch(scTakenMask)
 
   private val s2_ftqPtr = RegEnable(io.fromFtq.bpuPtr, s1_fire)
   private val s3_ftqPtr = RegEnable(s2_ftqPtr, s2_fire)
@@ -327,7 +329,7 @@ class Bpu(implicit p: Parameters) extends BpuModule with HalfAlignHelper {
   private val s3_ittageMeta = ittage.io.meta
 
   // sc meta
-  private val s3_scMeta = 0.U.asTypeOf(sc.io.meta)
+  private val s3_scMeta = sc.io.meta
 
   // ras meta
   private val s3_rasMeta = ras.io.specMeta
diff --git a/src/main/scala/xiangshan/frontend/bpu/sc/Bundles.scala b/src/main/scala/xiangshan/frontend/bpu/sc/Bundles.scala
index 58a19f0930e..9f087b2c7ec 100644
--- a/src/main/scala/xiangshan/frontend/bpu/sc/Bundles.scala
+++ b/src/main/scala/xiangshan/frontend/bpu/sc/Bundles.scala
@@ -24,28 +24,17 @@ import xiangshan.frontend.bpu.SignedSaturateCounter
 import xiangshan.frontend.bpu.WriteReqBundle
 
 class ScEntry(implicit p: Parameters) extends ScBundle {
-  val ctrs: SignedSaturateCounter = new SignedSaturateCounter(ctrWidth)
+  val ctr: SignedSaturateCounter = new SignedSaturateCounter(ctrWidth)
 }
 
 class ScThreshold(implicit p: Parameters) extends ScBundle {
-  val ctr: SaturateCounter = new SaturateCounter(thresholdCtrWidth)
-  val thres = UInt(thresholdThresWidth.W)
-  def satPos(ctr: UInt = this.ctr.value): Bool = ctr === ((1.U << thresholdCtrWidth) - 1.U)
-  def satNeg(ctr: UInt = this.ctr.value): Bool = ctr === 0.U
-  def neutralVal: UInt = (1 << (thresholdCtrWidth - 1)).U
-  def initVal:    UInt = 6.U
-  def minThres:   UInt = 6.U
-  def maxThres:   UInt = 31.U
+  val thres: SaturateCounter = new SaturateCounter(thresholdThresWidth)
+
+  def initVal: UInt = 6.U
+
   def update(cause: Bool): ScThreshold = {
-    val res    = Wire(new ScThreshold())
-    val newCtr = this.ctr.getUpdate(cause)
-    val newThres = Mux(
-      res.satPos(newCtr) && this.thres <= maxThres,
-      this.thres + 2.U,
-      Mux(res.satNeg(newCtr) && this.thres >= minThres, this.thres - 2.U, this.thres)
-    )
-    res.thres     := newThres
-    res.ctr.value := Mux(res.satPos(newCtr) || res.satNeg(newCtr), res.neutralVal, newCtr)
+    val res = Wire(new ScThreshold())
+    res.thres.value := this.thres.getUpdate(cause)
     res
   }
 }
@@ -53,23 +42,22 @@ class ScThreshold(implicit p: Parameters) extends ScBundle {
 object ScThreshold {
   def apply(implicit p: Parameters): ScThreshold = {
     val t = Wire(new ScThreshold())
-    t.ctr.value := t.neutralVal
-    t.thres     := t.initVal
+    t.thres.value := t.initVal
     t
   }
 }
 
 class PathTableSramWriteReq(val numSets: Int)(implicit p: Parameters) extends WriteReqBundle with HasScParameters {
-  val setIdx:    UInt         = UInt(log2Ceil(numSets).W)
-  val wayIdxVec: Vec[UInt]    = Vec(ResolveEntryBranchNumber, UInt(log2Ceil(NumWays).W))
-  val entryVec:  Vec[ScEntry] = Vec(ResolveEntryBranchNumber, new ScEntry())
+  val setIdx:   UInt         = UInt(log2Ceil(numSets).W)
+  val wayMask:  Vec[Bool]    = Vec(NumWays, Bool())
+  val entryVec: Vec[ScEntry] = Vec(NumWays, new ScEntry())
 }
 
 class PathTableTrain(val numSets: Int)(implicit p: Parameters) extends ScBundle {
-  val valid:     Bool         = Bool()
-  val setIdx:    UInt         = UInt(log2Ceil(numSets / NumBanks).W)
-  val wayIdxVec: Vec[UInt]    = Vec(ResolveEntryBranchNumber, UInt(log2Ceil(NumWays).W))
-  val entryVec:  Vec[ScEntry] = Vec(ResolveEntryBranchNumber, new ScEntry())
+  val valid:    Bool         = Bool()
+  val setIdx:   UInt         = UInt(log2Ceil(numSets / NumBanks).W)
+  val wayMask:  Vec[Bool]    = Vec(NumWays, Bool())
+  val entryVec: Vec[ScEntry] = Vec(NumWays, new ScEntry())
 }
 
 class ScMeta(implicit p: Parameters) extends ScBundle with HasScParameters {
diff --git a/src/main/scala/xiangshan/frontend/bpu/sc/Parameters.scala b/src/main/scala/xiangshan/frontend/bpu/sc/Parameters.scala
index 0945424246a..9cdea310e7c 100644
--- a/src/main/scala/xiangshan/frontend/bpu/sc/Parameters.scala
+++ b/src/main/scala/xiangshan/frontend/bpu/sc/Parameters.scala
@@ -32,7 +32,6 @@ case class ScParameters(
     ),
     ctrWidth:            Int = 6,
     weightCtrWidth:      Int = 6,
-    thresholdCtrWidth:   Int = 6,
     thresholdThresWidth: Int = 8,
     NumTables:           Int = 2,
     NumBanks:            Int = 2,
@@ -44,7 +43,6 @@ trait HasScParameters extends HasBpuParameters {
   def scParameters:        ScParameters     = bpuParameters.scParameters
   def ctrWidth:            Int              = scParameters.ctrWidth
   def weightCtrWidth:      Int              = scParameters.weightCtrWidth
-  def thresholdCtrWidth:   Int              = scParameters.thresholdCtrWidth
   def thresholdThresWidth: Int              = scParameters.thresholdThresWidth
   def TableInfos:          Seq[ScTableInfo] = scParameters.TableInfos
   def PathTableInfos:      Seq[ScTableInfo] = scParameters.PathTableInfos
diff --git a/src/main/scala/xiangshan/frontend/bpu/sc/Sc.scala b/src/main/scala/xiangshan/frontend/bpu/sc/Sc.scala
index 7fb620238ae..ce6aec304e6 100644
--- a/src/main/scala/xiangshan/frontend/bpu/sc/Sc.scala
+++ b/src/main/scala/xiangshan/frontend/bpu/sc/Sc.scala
@@ -82,7 +82,7 @@ class Sc(implicit p: Parameters) extends BasePredictor with HasScParameters with
   private val s1_resp: Seq[Vec[ScEntry]] = pathTable.map(_.io.resp)
 
   private val s1_pathPercsum: Vec[Vec[SInt]] = VecInit(s1_resp.map(entries =>
-    VecInit(entries.map(entry => getPercsum(entry.ctrs.value)))
+    VecInit(entries.map(entry => getPercsum(entry.ctr.value)))
   ))
 
   /*
@@ -111,7 +111,7 @@ class Sc(implicit p: Parameters) extends BasePredictor with HasScParameters with
 
   private val s2_scPred: Vec[Bool] = VecInit(s2_totalPercsum.map(_ > 0.S))
 
-  private val s2_thresholds    = scThreshold.map(entry => entry.thres)
+  private val s2_thresholds    = scThreshold.map(entry => entry.thres.value)
   private val updateThresholds = VecInit(s2_thresholds.map(t => (t << 3) +& 21.U))
 
   private val s2_useScPred = WireInit(VecInit.fill(NumWays)(false.B))
@@ -135,7 +135,7 @@ class Sc(implicit p: Parameters) extends BasePredictor with HasScParameters with
   /*
    *  train pipeline stage 1
    */
-  private val t1_trainValid = RegEnable(io.train.valid, io.enable)
+  private val t1_trainValid = RegNext(io.train.valid, init = false.B)
   private val t1_train      = RegEnable(io.train.bits, io.train.valid)
   private val t1_branches   = t1_train.branches
   private val t1_setIdx = PathTableInfos.map(info =>
@@ -146,53 +146,67 @@ class Sc(implicit p: Parameters) extends BasePredictor with HasScParameters with
       info.Size / NumWays
     )
   )
-  private val t1_meta      = t1_train.meta.sc
-  private val t1_oldCtrs   = VecInit(t1_meta.scResp.map(v => VecInit(v.map(r => r.asTypeOf(new ScEntry())))))
-  private val t1_takenMask = VecInit(t1_branches.map(b => b.valid && b.bits.taken))
-  private val t1_writeValidMask =
+  private val t1_meta    = t1_train.meta.sc
+  private val t1_oldCtrs = VecInit(t1_meta.scResp.map(v => VecInit(v.map(r => r.asTypeOf(new ScEntry())))))
+  private val t1_writeValidVec =
     VecInit(t1_branches.map(b => b.valid && b.bits.attribute.isConditional && t1_trainValid))
-  private val t1_writeValid     = t1_writeValidMask.reduce(_ || _)
-  private val t1_writeWayIdxVec = VecInit(t1_branches.map(b => b.bits.cfiPosition(log2Ceil(NumWays) - 1, 0)))
+  private val t1_writeValid        = t1_writeValidVec.reduce(_ || _)
+  private val t1_branchesTakenMask = VecInit(t1_branches.map(b => b.valid && b.bits.taken))
+  private val t1_branchesWayIdxVec = VecInit(t1_branches.map(b => b.bits.cfiPosition(log2Ceil(NumWays) - 1, 0)))
   require(
-    t1_writeWayIdxVec(0).getWidth == log2Ceil(NumWays),
-    s"t1_writeWayIdxVec entry width: ${t1_writeWayIdxVec(0).getWidth} should be the same as log2Ceil(NumWays): ${log2Ceil(NumWays)}"
+    t1_branchesWayIdxVec(0).getWidth == log2Ceil(NumWays),
+    s"t1_branchesWayIdxVec entry width: ${t1_branchesWayIdxVec(0).getWidth} should be the same as log2Ceil(NumWays): ${log2Ceil(NumWays)}"
   )
-  private val t1_newThresVec = VecInit(t1_writeWayIdxVec.zip(t1_takenMask).map { case (wayIdx, taken) =>
-    scThreshold(wayIdx).update(taken =/= t1_meta.scPred(wayIdx))
+
+  private val t1_mismatchMask = WireInit(VecInit.fill(NumWays)(false.B))
+  private val t1_writeThresVec = VecInit(scThreshold.indices.map { wayIdx =>
+    val updated = t1_branchesWayIdxVec.zip(t1_branchesTakenMask).foldLeft(scThreshold(wayIdx)) {
+      case (prevThres, (branchWayIdx, taken)) =>
+        val shouldUpdate = branchWayIdx === wayIdx.U
+        val mismatch     = taken =/= t1_meta.scPred(wayIdx)
+        t1_mismatchMask(wayIdx) := mismatch && shouldUpdate
+        val nextThres = prevThres.update(mismatch)
+        Mux(shouldUpdate, nextThres, prevThres)
+    }
+    WireInit(updated)
   })
+  dontTouch(t1_writeThresVec)
+
   private val t1_writeEntryVec = WireInit(
-    VecInit.fill(PathTableSize)(VecInit.fill(ResolveEntryBranchNumber)(0.U.asTypeOf(new ScEntry())))
+    VecInit.fill(PathTableSize)(VecInit.fill(NumWays)(0.U.asTypeOf(new ScEntry())))
   )
-
-  t1_oldCtrs.zip(t1_writeEntryVec).foreach {
-    case (oldEntry: Vec[ScEntry], writeEntries: Vec[ScEntry]) =>
-      t1_takenMask.zip(t1_writeWayIdxVec).zip(t1_writeValidMask).zipWithIndex.foreach {
-        case (((taken, wayIdx), valid), i) =>
-          when(valid) {
-            writeEntries(wayIdx).ctrs.value := oldEntry(wayIdx).ctrs.getUpdate(taken)
-          }.otherwise {
-            writeEntries(wayIdx).ctrs.value := 0.S
-          }
+  private val t1_writeWayMask = WireInit(VecInit.fill(PathTableSize)(VecInit.fill(NumWays)(false.B)))
+
+  t1_oldCtrs.zip(t1_writeEntryVec).zip(t1_writeWayMask).foreach {
+    case ((oldEntries: Vec[ScEntry], writeEntries: Vec[ScEntry]), writeWayMask: Vec[Bool]) =>
+      oldEntries.zip(writeEntries).zipWithIndex.foreach { case ((oldEntry, newEntry), wayIdx) =>
+        val newCtr = t1_branchesTakenMask.zip(t1_branchesWayIdxVec).zip(t1_writeValidVec).foldLeft(oldEntry.ctr) {
+          case (prevCtr, ((writeTaken, writeWayIdx), writeValidalid)) =>
+            val needUpdate = writeValidalid && writeWayIdx === wayIdx.U
+            val nextValue  = prevCtr.getUpdate(writeTaken)
+            val nextCtr    = WireInit(prevCtr)
+            nextCtr.value             := nextValue
+            writeWayMask(writeWayIdx) := needUpdate
+            Mux(needUpdate, nextCtr, prevCtr)
+        }
+        dontTouch(newCtr)
+        newEntry.ctr := WireInit(newCtr)
       }
   }
 
-  pathTable zip t1_setIdx zip t1_writeEntryVec foreach {
-    case ((table, idx), writeEntries) =>
-      table.io.update.valid     := t1_writeValid
-      table.io.update.setIdx    := idx
-      table.io.update.wayIdxVec := t1_writeWayIdxVec
-      table.io.update.entryVec  := writeEntries
+  dontTouch(t1_writeEntryVec)
+
+  pathTable zip t1_setIdx zip t1_writeEntryVec zip t1_writeWayMask foreach {
+    case (((table, idx), writeEntries), wayMask) =>
+      table.io.update.valid    := t1_writeValid
+      table.io.update.setIdx   := idx
+      table.io.update.wayMask  := wayMask
+      table.io.update.entryVec := writeEntries
   }
 
-  t1_writeValidMask.zip(t1_writeWayIdxVec).zip(t1_newThresVec).map {
-    case ((valid, wayIdx), newThres) =>
-      when(valid && t1_meta.scPred(wayIdx) =/= t1_takenMask(wayIdx)) {
-        scThreshold(wayIdx) := newThres
-      }
-      XSPerfAccumulate(
-        "sc_pred_right",
-        valid && (t1_meta.scPred(wayIdx) =/= t1_takenMask(wayIdx)) && t1_meta.useScPred(wayIdx)
-      )
+  when(t1_writeValid) {
+    scThreshold := t1_writeThresVec
   }
-  // TODO： add more performance counters
+  XSPerfAccumulate("sc_pred_wrong", t1_writeValid && t1_mismatchMask.reduce(_ || _))
+  // TODO: add more performance counters
 }
diff --git a/src/main/scala/xiangshan/frontend/bpu/sc/ScPathTable.scala b/src/main/scala/xiangshan/frontend/bpu/sc/ScPathTable.scala
index bba2294309b..d9c578e025e 100644
--- a/src/main/scala/xiangshan/frontend/bpu/sc/ScPathTable.scala
+++ b/src/main/scala/xiangshan/frontend/bpu/sc/ScPathTable.scala
@@ -76,21 +76,19 @@ class ScPathTable(val numSets: Int, val histLen: Int)(implicit p: Parameters)
   // update path table
   private val updateValid    = io.update.valid
   private val updateIdx      = io.update.setIdx
-  private val updateWayVec   = io.update.wayIdxVec
+  private val updateWayMask  = io.update.wayMask
   private val updateBankIdx  = updateIdx(log2Ceil(NumBanks) - 1, 0)
   private val updateBankMask = UIntToOH(updateBankIdx, NumBanks)
-  private val updateWayMask  = WireInit(VecInit.fill(NumWays)(false.B))
-  updateWayVec.foreach(wayIdx => updateWayMask(wayIdx) := 1.U)
 
   writeBuffer.zip(updateBankMask.asBools).foreach {
     case (buffer, bankEnable) =>
       val writeValid = updateValid && bankEnable
       buffer.io.write.head.valid       := writeValid
       buffer.io.write.head.bits.setIdx := updateIdx >> log2Ceil(NumBanks)
-      buffer.io.write.head.bits.wayIdxVec := Mux(
+      buffer.io.write.head.bits.wayMask := Mux(
         writeValid,
-        io.update.wayIdxVec,
-        VecInit.fill(ResolveEntryBranchNumber)(0.U.asTypeOf(UInt(log2Ceil(NumWays).W)))
+        updateWayMask,
+        VecInit.fill(NumWays)(false.B)
       )
       buffer.io.write.head.bits.entryVec := Mux(
         writeValid,
@@ -101,24 +99,10 @@ class ScPathTable(val numSets: Int, val histLen: Int)(implicit p: Parameters)
 
   sram.zip(writeBuffer).zipWithIndex.foreach {
     case ((bank, buffer), i) =>
-      // brand way extend to sram way
-      require(
-        ResolveEntryBranchNumber <= NumWays,
-        s"resolve branches: ${ResolveEntryBranchNumber} should be less than or equal to NumWays: ${NumWays}"
-      )
-      val wayMask     = WireInit(VecInit.fill(NumWays)(false.B))
-      val entryVec    = WireInit(VecInit.fill(NumWays)(0.U.asTypeOf(new ScEntry())))
-      val wayIdxVecIn = buffer.io.read.head.bits.wayIdxVec
-      val entryVecIn  = buffer.io.read.head.bits.entryVec
-      wayIdxVecIn.zip(entryVecIn).foreach {
-        case (wayIdx, entry) =>
-          wayMask(wayIdx)  := 1.U
-          entryVec(wayIdx) := entry
-      }
       bank.io.w.req.valid            := buffer.io.read.head.valid && !bank.io.r.req.valid
       bank.io.w.req.bits.setIdx      := buffer.io.read.head.bits.setIdx
-      bank.io.w.req.bits.waymask.get := wayMask.asUInt
-      bank.io.w.req.bits.data        := entryVec
+      bank.io.w.req.bits.waymask.get := buffer.io.read.head.bits.wayMask.asUInt
+      bank.io.w.req.bits.data        := buffer.io.read.head.bits.entryVec
       buffer.io.read.head.ready      := bank.io.w.req.ready && !bank.io.r.req.valid
   }
 }
