diff --git a/src/main/scala/xiangshan/frontend/BPU.scala b/src/main/scala/xiangshan/frontend/BPU.scala
index bef4a76f8e1..c10e73ef377 100644
--- a/src/main/scala/xiangshan/frontend/BPU.scala
+++ b/src/main/scala/xiangshan/frontend/BPU.scala
@@ -194,15 +194,13 @@ abstract class BasePredictor(implicit p: Parameters) extends XSModule
   io.s2_ready := true.B
   io.s3_ready := true.B
 
-  val (_, reset_vector) = DelayNWithValid(io.reset_vector, reset.asBool, 5, hasInit = false)
-
   val s0_pc_dup   = WireInit(io.in.bits.s0_pc) // fetchIdx(io.f0_pc)
   val s1_pc_dup   = s0_pc_dup.zip(io.s0_fire).map {case (s0_pc, s0_fire) => RegEnable(s0_pc, s0_fire)}
   val s2_pc_dup   = s1_pc_dup.zip(io.s1_fire).map {case (s1_pc, s1_fire) => SegmentedAddrNext(s1_pc, pcSegments, s1_fire, Some("s2_pc"))}
   val s3_pc_dup   = s2_pc_dup.zip(io.s2_fire).map {case (s2_pc, s2_fire) => SegmentedAddrNext(s2_pc, s2_fire, Some("s3_pc"))}
 
   when (RegNext(RegNext(reset.asBool) && !reset.asBool)) {
-    s1_pc_dup.map{case s1_pc => s1_pc := reset_vector}
+    s1_pc_dup.map{case s1_pc => s1_pc := io.reset_vector}
   }
 
   io.out.s1.pc := s1_pc_dup
@@ -278,9 +276,6 @@ class Predictor(implicit p: Parameters) extends XSModule with HasBPUConst with H
   predictors.io.ctrl := ctrl
   predictors.io.reset_vector := io.reset_vector
 
-
-  val (_, reset_vector) = DelayNWithValid(io.reset_vector, reset.asBool, 5, hasInit = false)
-
   val s0_stall_dup = dup_wire(Bool()) // For some reason s0 stalled, usually FTQ Full
   val s0_fire_dup, s1_fire_dup, s2_fire_dup, s3_fire_dup = dup_wire(Bool())
   val s1_valid_dup, s2_valid_dup, s3_valid_dup = dup_seq(RegInit(false.B))
@@ -290,7 +285,7 @@ class Predictor(implicit p: Parameters) extends XSModule with HasBPUConst with H
   val s0_pc_dup = dup(WireInit(0.U.asTypeOf(UInt(VAddrBits.W))))
   val s0_pc_reg_dup = s0_pc_dup.zip(s0_stall_dup).map{ case (s0_pc, s0_stall) => RegEnable(s0_pc, !s0_stall) }
   when (RegNext(RegNext(reset.asBool) && !reset.asBool)) {
-    s0_pc_reg_dup.map{case s0_pc => s0_pc := reset_vector}
+    s0_pc_reg_dup.map{case s0_pc => s0_pc := io.reset_vector}
   }
   val s1_pc = RegEnable(s0_pc_dup(0), s0_fire_dup(0))
   val s2_pc = RegEnable(s1_pc, s1_fire_dup(0))
diff --git a/src/main/scala/xiangshan/frontend/Frontend.scala b/src/main/scala/xiangshan/frontend/Frontend.scala
index 59b7bc9c566..fde3f94dee5 100644
--- a/src/main/scala/xiangshan/frontend/Frontend.scala
+++ b/src/main/scala/xiangshan/frontend/Frontend.scala
@@ -105,7 +105,7 @@ class FrontendInlinedImp (outer: FrontendInlined) extends LazyModuleImp(outer)
 
   // bpu ctrl
   bpu.io.ctrl := csrCtrl.bp_ctrl
-  bpu.io.reset_vector := RegEnable(io.reset_vector, reset.asBool)
+  bpu.io.reset_vector := io.reset_vector
 
 // pmp
   val PortNumber = ICacheParameters().PortNumber
