diff --git a/src/main/scala/xiangshan/backend/rename/BusyTable.scala b/src/main/scala/xiangshan/backend/rename/BusyTable.scala
index 040f3c8729b..13ab1b9c5d0 100644
--- a/src/main/scala/xiangshan/backend/rename/BusyTable.scala
+++ b/src/main/scala/xiangshan/backend/rename/BusyTable.scala
@@ -205,22 +205,22 @@ class VlBusyTable(numReadPorts: Int, numWritePorts: Int, numPhyPregs: Int, pregW
   val zeroTableUpdate = Wire(Vec(numPhyPregs, Bool()))
   val vlmaxTableUpdate = Wire(Vec(numPhyPregs, Bool()))
 
-  val wb0Mask = Mux(io.wbPregs(intSchdVlWbPort).valid, UIntToOH(io.wbPregs(intSchdVlWbPort).bits), 0.U)
-  val wb1Mask = Mux(io.wbPregs(vfSchdVlWbPort).valid, UIntToOH(io.wbPregs(vfSchdVlWbPort).bits), 0.U)
+  val intVlWb = Mux(io.wbPregs(intSchdVlWbPort).valid, UIntToOH(io.wbPregs(intSchdVlWbPort).bits), 0.U)
+  val vfVlWb = Mux(io.wbPregs(vfSchdVlWbPort).valid, UIntToOH(io.wbPregs(vfSchdVlWbPort).bits), 0.U)
 
   val zeroTable = VecInit((0 until numPhyPregs).zip(zeroTableUpdate).map{ case (idx, update) =>
-    RegEnable(update, 0.U(1.W), allocMask(idx) || ldCancelMask(idx) || wb0Mask(idx) || wb1Mask(idx))
+    RegEnable(update, 0.U(1.W), allocMask(idx) || ldCancelMask(idx) || intVlWb(idx) || vfVlWb(idx))
   }).asUInt
   val vlmaxTable = VecInit((0 until numPhyPregs).zip(vlmaxTableUpdate).map{ case (idx, update) =>
-    RegEnable(update, 0.U(1.W), allocMask(idx) || ldCancelMask(idx) || wb0Mask(idx) || wb1Mask(idx))
+    RegEnable(update, 0.U(1.W), allocMask(idx) || ldCancelMask(idx) || intVlWb(idx) || vfVlWb(idx))
   }).asUInt
 
 
   zeroTableUpdate.zipWithIndex.foreach{ case (update, idx) =>
-    when(wb0Mask(idx)) {
+    when(intVlWb(idx)) {
       // int schd vl write back, check whether the vl is zero
       update := !io_vl_Wb.vlWriteBackInfo.vlFromIntIsZero
-    }.elsewhen(wb1Mask(idx)) {
+    }.elsewhen(vfVlWb(idx)) {
       // vf schd vl write back, check whether the vl is zero
       update := !io_vl_Wb.vlWriteBackInfo.vlFromVfIsZero
     }.elsewhen(allocMask(idx) || ldCancelMask(idx)) {
@@ -231,10 +231,10 @@ class VlBusyTable(numReadPorts: Int, numWritePorts: Int, numPhyPregs: Int, pregW
   }
 
   vlmaxTableUpdate.zipWithIndex.foreach{ case (update, idx) =>
-    when(wb1Mask(idx)) {
+    when(intVlWb(idx)) {
       // int schd vl write back, check whether the vl is vlmax
       update := !io_vl_Wb.vlWriteBackInfo.vlFromIntIsVlmax
-    }.elsewhen(wb1Mask(idx)) {
+    }.elsewhen(vfVlWb(idx)) {
       // vf schd vl write back, check whether the vl is vlmax
       update := !io_vl_Wb.vlWriteBackInfo.vlFromVfIsVlmax
     }.elsewhen(allocMask(idx) || ldCancelMask(idx)) {
