diff --git a/src/main/scala/xiangshan/frontend/icache/Bundles.scala b/src/main/scala/xiangshan/frontend/icache/Bundles.scala
index d9a3008ad31..694e67c8772 100644
--- a/src/main/scala/xiangshan/frontend/icache/Bundles.scala
+++ b/src/main/scala/xiangshan/frontend/icache/Bundles.scala
@@ -291,20 +291,37 @@ class PmpCheckBundle(implicit p: Parameters) extends ICacheBundle {
 
 /* ***** Perf ***** */
 class ICachePerfInfo(implicit p: Parameters) extends ICacheBundle {
-  val only0Hit:     Bool      = Bool()
-  val only0Miss:    Bool      = Bool()
-  val hit0Hit1:     Bool      = Bool()
-  val hit0Miss1:    Bool      = Bool()
-  val miss0Hit1:    Bool      = Bool()
-  val miss0Miss1:   Bool      = Bool()
-  val hit0Except1:  Bool      = Bool()
-  val miss0Except1: Bool      = Bool()
-  val except0:      Bool      = Bool()
-  val bankHit:      Vec[Bool] = Vec(PortNumber, Bool())
-  val hit:          Bool      = Bool()
+  val hits:         Vec[Bool]          = Vec(PortNumber, Bool())
+  val isDoubleLine: Bool               = Bool()
+  val exception:    Vec[ExceptionType] = Vec(PortNumber, new ExceptionType)
+
+  // single line request, port 0 is hit/miss
+  def hit0NoReq1:  Bool = hits(0) && !isDoubleLine
+  def miss0NoReq1: Bool = !hits(0) && !isDoubleLine
+  // double line request, port 0/1 is hit/miss
+  def hit0Hit1:   Bool = hits(0) && hits(1) && isDoubleLine
+  def hit0Miss1:  Bool = hits(0) && !hits(1) && isDoubleLine
+  def miss0Hit1:  Bool = !hits(0) && hits(1) && isDoubleLine
+  def miss0Miss1: Bool = !hits(0) && !hits(1) && isDoubleLine
+  // has exception
+  def except0:      Bool = exception(0).hasException
+  def hit0Except1:  Bool = hits(0) && exception(1).hasException && isDoubleLine
+  def miss0Except1: Bool = !hits(0) && exception(1).hasException && isDoubleLine
+
+  def hit0: Bool = hits(0)
+  def hit1: Bool = hits(1) && isDoubleLine
+
+  def hit: Bool = hits(0) && (!isDoubleLine || hits(1))
+
+  def except: Bool = exception(0).hasException || (isDoubleLine && exception(1).hasException)
 }
 
 class ICacheTopdownInfo(implicit p: Parameters) extends ICacheBundle {
-  val icacheMiss: Bool = Output(Bool())
-  val itlbMiss:   Bool = Output(Bool())
+  val iCacheMissBubble: Bool = Output(Bool())
+  val itlbMissBubble:   Bool = Output(Bool())
+}
+
+class ICachePerfEventsInfo(implicit p: Parameters) extends ICacheBundle {
+  val miss:       Bool = Output(Bool())
+  val missBubble: Bool = Output(Bool())
 }
diff --git a/src/main/scala/xiangshan/frontend/icache/ICacheImp.scala b/src/main/scala/xiangshan/frontend/icache/ICacheImp.scala
index e3f4313cae6..48002a219f6 100644
--- a/src/main/scala/xiangshan/frontend/icache/ICacheImp.scala
+++ b/src/main/scala/xiangshan/frontend/icache/ICacheImp.scala
@@ -229,8 +229,8 @@ class ICacheImp(outer: ICache) extends LazyModuleImp(outer) with HasICacheParame
   XSPerfAccumulate("softPrefetch_block_ftq", softPrefetchValid && io.fromFtq.prefetchReq.valid)
 
   val perfEvents: Seq[(String, Bool)] = Seq(
-    ("icache_miss_cnt  ", false.B),
-    ("icache_miss_penalty", BoolStopWatch(start = false.B, stop = false.B || false.B, startHighPriority = true))
+    ("icache_miss_cnt", mainPipe.io.perfEvents.miss),
+    ("icache_miss_penalty", mainPipe.io.perfEvents.missBubble)
   )
   generatePerfEvent()
 }
diff --git a/src/main/scala/xiangshan/frontend/icache/ICacheMainPipe.scala b/src/main/scala/xiangshan/frontend/icache/ICacheMainPipe.scala
index 002455e4659..3afc25ab89c 100644
--- a/src/main/scala/xiangshan/frontend/icache/ICacheMainPipe.scala
+++ b/src/main/scala/xiangshan/frontend/icache/ICacheMainPipe.scala
@@ -62,8 +62,9 @@ class ICacheMainPipe(implicit p: Parameters) extends ICacheModule
     val errors: Vec[Valid[L1CacheErrorInfo]] = Output(Vec(PortNumber, ValidIO(new L1CacheErrorInfo)))
 
     /* *** Perf *** */
-    val perf:    ICachePerfInfo    = Output(new ICachePerfInfo)
-    val topdown: ICacheTopdownInfo = Output(new ICacheTopdownInfo)
+    val perf:       ICachePerfInfo       = Output(new ICachePerfInfo)
+    val topdown:    ICacheTopdownInfo    = Output(new ICacheTopdownInfo)
+    val perfEvents: ICachePerfEventsInfo = Output(new ICachePerfEventsInfo)
   }
 
   val io: ICacheMainPipeIO = IO(new ICacheMainPipeIO)
@@ -243,9 +244,9 @@ class ICacheMainPipe(implicit p: Parameters) extends ICacheModule
   )
   private val s1_mshrDatas = fromMiss.bits.data.asTypeOf(Vec(ICacheDataBanks, UInt((blockBits / ICacheDataBanks).W)))
 
-  private val s1_hits = (0 until PortNumber).map { i =>
+  private val s1_hits = VecInit((0 until PortNumber).map { i =>
     ValidHoldBypass(s1_mshrHits(i) || (RegNext(s0_fire) && s1_sramHits(i)), s1_fire || s1_flush)
-  }
+  })
 
   private val s1_bankMshrHit = getBankValid(s1_mshrHits, s1_offset)
 
@@ -490,31 +491,29 @@ class ICacheMainPipe(implicit p: Parameters) extends ICacheModule
     }
   }
 
-  /**
-    ******************************************************************************
-    * performance info. TODO: need to simplify the logic
-    ***********************************************************s*******************
-    */
-  io.perf.only0Hit     := s2_hits(0) && !s2_doubleline
-  io.perf.only0Miss    := !s2_hits(0) && !s2_doubleline
-  io.perf.hit0Hit1     := s2_hits(0) && s2_hits(1) && s2_doubleline
-  io.perf.hit0Miss1    := s2_hits(0) && !s2_hits(1) && s2_doubleline
-  io.perf.miss0Hit1    := !s2_hits(0) && s2_hits(1) && s2_doubleline
-  io.perf.miss0Miss1   := !s2_hits(0) && !s2_hits(1) && s2_doubleline
-  io.perf.hit0Except1  := s2_hits(0) && s2_exception(1).hasException && s2_doubleline
-  io.perf.miss0Except1 := !s2_hits(0) && s2_exception(1).hasException && s2_doubleline
-  io.perf.bankHit(0)   := s2_hits(0)
-  io.perf.bankHit(1)   := s2_hits(1) && s2_doubleline
-  io.perf.except0      := s2_exception(0).hasException
-  io.perf.hit          := s2_hits(0) && (!s2_doubleline || s2_hits(1))
-
-  /** <PERF> fetch bubble generated by icache miss */
+  /* *****************************************************************************
+   * perf
+   * ***************************************************************************** */
+  // when fired, tell ifu hit state of each cache line
+  // NOTE: we cannot use s2_hits, it will be reset when refilled from L2
+  private val s2_rawHits = RegEnable(s1_hits, 0.U.asTypeOf(s1_hits), s1_fire)
+  io.perf.hits         := s2_rawHits
+  io.perf.isDoubleLine := s2_doubleline
+  io.perf.exception    := s2_exceptionOut
+
+  // tell topdown when waiting for icache refill or itlb refill
+  io.topdown.iCacheMissBubble := s2_valid && !s2_fetchFinish
+  io.topdown.itlbMissBubble   := s0_valid && !fromWayLookup.ready
+
+  // tell icache top when waiting for refill
+  io.perfEvents.missBubble := s2_valid && !s2_fetchFinish
+  // and when fired, tell icache top if it is a miss
+  io.perfEvents.miss := s2_fire && (!s2_rawHits(0) || !s2_rawHits(1) && s2_doubleline)
+
+  // count fetch bubble, same as topdown, but an inner perf counter
   XSPerfAccumulate("icache_bubble_s2_miss", s2_valid && !s2_fetchFinish)
   XSPerfAccumulate("icache_bubble_s0_wayLookup", s0_valid && !fromWayLookup.ready)
 
-  io.topdown.icacheMiss := !s2_fetchFinish
-  io.topdown.itlbMiss   := s0_valid && !fromWayLookup.ready
-
   // class ICacheTouchDB(implicit p: Parameters) extends ICacheBundle{
   //   val blkPAddr  = UInt((PAddrBits - blockOffBits).W)
   //   val vSetIdx   = UInt(idxBits.W)
diff --git a/src/main/scala/xiangshan/frontend/icache/ICacheMissUnit.scala b/src/main/scala/xiangshan/frontend/icache/ICacheMissUnit.scala
index cd886001a1f..8464b9facbf 100644
--- a/src/main/scala/xiangshan/frontend/icache/ICacheMissUnit.scala
+++ b/src/main/scala/xiangshan/frontend/icache/ICacheMissUnit.scala
@@ -26,6 +26,7 @@ import org.chipsalliance.cde.config.Parameters
 import utility.ChiselDB
 import utility.Constantin
 import utility.XSPerfAccumulate
+import utility.XSPerfHistogram
 import xiangshan.XSCoreParamsKey
 
 class ICacheMissUnit(edge: TLEdgeOut)(implicit p: Parameters) extends ICacheModule with ICacheAddrHelper {
@@ -250,15 +251,29 @@ class ICacheMissUnit(edge: TLEdgeOut)(implicit p: Parameters) extends ICacheModu
   io.resp.bits.data     := respDataReg.asUInt
   io.resp.bits.corrupt  := corruptReg
 
-  /**
-    ******************************************************************************
-    * performance counter
-    ******************************************************************************
-    */
-  // Duplicate requests will be excluded.
+  /* *****************************************************************************
+   * perf
+   * ***************************************************************************** */
+  // Total requests, duplicate requests will be excluded.
   XSPerfAccumulate("enq_fetch_req", fetchDemux.io.in.fire)
   XSPerfAccumulate("enq_prefetch_req", prefetchDemux.io.in.fire)
 
+  // Mshr occupancy
+  XSPerfHistogram(
+    "fetchMshrReadyCnt",
+    PopCount(fetchDemux.io.out.map(_.ready)),
+    true.B,
+    0,
+    nFetchMshr
+  )
+  XSPerfHistogram(
+    "prefetchMshrReadyCnt",
+    PopCount(prefetchDemux.io.out.map(_.ready)),
+    true.B,
+    0,
+    nPrefetchMshr
+  )
+
   /**
     ******************************************************************************
     * ChiselDB: record ICache SRAM write log
diff --git a/src/main/scala/xiangshan/frontend/ifu/Ifu.scala b/src/main/scala/xiangshan/frontend/ifu/Ifu.scala
index cee19711fdf..beb40009383 100644
--- a/src/main/scala/xiangshan/frontend/ifu/Ifu.scala
+++ b/src/main/scala/xiangshan/frontend/ifu/Ifu.scala
@@ -131,8 +131,8 @@ class Ifu(implicit p: Parameters) extends IfuModule
   require(numOfStage > 1, "Ifu numOfStage must be greater than 1")
   private val topdownStages = RegInit(VecInit(Seq.fill(numOfStage)(0.U.asTypeOf(new FrontendTopDownBundle))))
   // bubble events in IFU, only happen in stage 1
-  private val icacheMissBubble = io.fromICache.topdown.icacheMiss
-  private val itlbMissBubble   = io.fromICache.topdown.itlbMiss
+  private val icacheMissBubble = io.fromICache.topdown.iCacheMissBubble
+  private val itlbMissBubble   = io.fromICache.topdown.itlbMissBubble
 
   // only driven by clock, not valid-ready
   topdownStages(0) := fromFtq.req.bits.topdown_info
@@ -1072,44 +1072,39 @@ class Ifu(implicit p: Parameters) extends IfuModule
 
   /* *** Perf *** */
   private val s3_perfInfo = RegEnable(s2_perfInfo, s2_fire)
-  private val s3_req0     = io.toIBuffer.fire
-  private val s3_req1     = io.toIBuffer.fire && s3_doubleline
-  private val s3_hit0     = io.toIBuffer.fire && s3_perfInfo.bankHit(0)
-  private val s3_hit1     = io.toIBuffer.fire && s3_doubleline & s3_perfInfo.bankHit(1)
-  private val s3_hit      = s3_perfInfo.hit
   val perfEvents: Seq[(String, Bool)] = Seq(
     ("frontendFlush                ", wbRedirect),
     ("ifu_req                      ", io.toIBuffer.fire),
     ("ifu_miss                     ", io.toIBuffer.fire && !s3_perfInfo.hit),
-    ("ifu_req_cacheline_0          ", s3_req0),
-    ("ifu_req_cacheline_1          ", s3_req1),
-    ("ifu_req_cacheline_0_hit      ", s3_hit1),
-    ("ifu_req_cacheline_1_hit      ", s3_hit1),
-    ("only_0_hit                   ", s3_perfInfo.only0Hit && io.toIBuffer.fire),
-    ("only_0_miss                  ", s3_perfInfo.only0Miss && io.toIBuffer.fire),
-    ("hit_0_hit_1                  ", s3_perfInfo.hit0Hit1 && io.toIBuffer.fire),
-    ("hit_0_miss_1                 ", s3_perfInfo.hit0Miss1 && io.toIBuffer.fire),
-    ("miss_0_hit_1                 ", s3_perfInfo.miss0Hit1 && io.toIBuffer.fire),
-    ("miss_0_miss_1                ", s3_perfInfo.miss0Miss1 && io.toIBuffer.fire)
+    ("ifu_req_cacheline_0          ", io.toIBuffer.fire),
+    ("ifu_req_cacheline_1          ", io.toIBuffer.fire && s3_perfInfo.isDoubleLine),
+    ("ifu_req_cacheline_0_hit      ", io.toIBuffer.fire && s3_perfInfo.hit0),
+    ("ifu_req_cacheline_1_hit      ", io.toIBuffer.fire && s3_perfInfo.hit1),
+    ("only_0_hit                   ", io.toIBuffer.fire && s3_perfInfo.hit0NoReq1),
+    ("only_0_miss                  ", io.toIBuffer.fire && s3_perfInfo.miss0NoReq1),
+    ("hit_0_hit_1                  ", io.toIBuffer.fire && s3_perfInfo.hit0Hit1),
+    ("hit_0_miss_1                 ", io.toIBuffer.fire && s3_perfInfo.hit0Miss1),
+    ("miss_0_hit_1                 ", io.toIBuffer.fire && s3_perfInfo.miss0Hit1),
+    ("miss_0_miss_1                ", io.toIBuffer.fire && s3_perfInfo.miss0Miss1)
   )
   generatePerfEvent()
 
-  XSPerfAccumulate("ifu_req", io.toIBuffer.fire)
-  XSPerfAccumulate("ifu_miss", io.toIBuffer.fire && !s3_hit)
-  XSPerfAccumulate("ifu_req_cacheline_0", s3_req0)
-  XSPerfAccumulate("ifu_req_cacheline_1", s3_req1)
-  XSPerfAccumulate("ifu_req_cacheline_0_hit", s3_hit0)
-  XSPerfAccumulate("ifu_req_cacheline_1_hit", s3_hit1)
   XSPerfAccumulate("frontendFlush", wbRedirect)
-  XSPerfAccumulate("only_0_hit", s3_perfInfo.only0Hit && io.toIBuffer.fire)
-  XSPerfAccumulate("only_0_miss", s3_perfInfo.only0Miss && io.toIBuffer.fire)
-  XSPerfAccumulate("hit_0_hit_1", s3_perfInfo.hit0Hit1 && io.toIBuffer.fire)
-  XSPerfAccumulate("hit_0_miss_1", s3_perfInfo.hit0Miss1 && io.toIBuffer.fire)
-  XSPerfAccumulate("miss_0_hit_1", s3_perfInfo.miss0Hit1 && io.toIBuffer.fire)
-  XSPerfAccumulate("miss_0_miss_1", s3_perfInfo.miss0Miss1 && io.toIBuffer.fire)
-  XSPerfAccumulate("hit_0_except_1", s3_perfInfo.hit0Except1 && io.toIBuffer.fire)
-  XSPerfAccumulate("miss_0_except_1", s3_perfInfo.miss0Except1 && io.toIBuffer.fire)
-  XSPerfAccumulate("except_0", s3_perfInfo.except0 && io.toIBuffer.fire)
+  XSPerfAccumulate("ifu_req", io.toIBuffer.fire)
+  XSPerfAccumulate("ifu_miss", io.toIBuffer.fire && !s3_perfInfo.hit)
+  XSPerfAccumulate("ifu_req_cacheline_0", io.toIBuffer.fire)
+  XSPerfAccumulate("ifu_req_cacheline_1", io.toIBuffer.fire && s3_perfInfo.isDoubleLine)
+  XSPerfAccumulate("ifu_req_cacheline_0_hit", io.toIBuffer.fire && s3_perfInfo.hit0)
+  XSPerfAccumulate("ifu_req_cacheline_1_hit", io.toIBuffer.fire && s3_perfInfo.hit1)
+  XSPerfAccumulate("only_0_hit", io.toIBuffer.fire && s3_perfInfo.hit0NoReq1)
+  XSPerfAccumulate("only_0_miss", io.toIBuffer.fire && s3_perfInfo.miss0NoReq1)
+  XSPerfAccumulate("hit_0_hit_1", io.toIBuffer.fire && s3_perfInfo.hit0Hit1)
+  XSPerfAccumulate("hit_0_miss_1", io.toIBuffer.fire && s3_perfInfo.hit0Miss1)
+  XSPerfAccumulate("miss_0_hit_1", io.toIBuffer.fire && s3_perfInfo.miss0Hit1)
+  XSPerfAccumulate("miss_0_miss_1", io.toIBuffer.fire && s3_perfInfo.miss0Miss1)
+  XSPerfAccumulate("except_0", io.toIBuffer.fire && s3_perfInfo.except0)
+  XSPerfAccumulate("hit_0_except_1", io.toIBuffer.fire && s3_perfInfo.hit0Except1)
+  XSPerfAccumulate("miss_0_except_1", io.toIBuffer.fire && s3_perfInfo.miss0Except1)
   XSPerfHistogram(
     "ifu2ibuffer_validCnt",
     PopCount(io.toIBuffer.bits.valid & io.toIBuffer.bits.enqEnable),
@@ -1129,11 +1124,8 @@ class Ifu(implicit p: Parameters) extends IfuModule
   private val fetchIBufferDumpData = Wire(new FetchToIBufferDB)
   fetchIBufferDumpData.startAddr  := s3_ftqReq.startAddr.toUInt
   fetchIBufferDumpData.instrCount := PopCount(io.toIBuffer.bits.enqEnable)
-  fetchIBufferDumpData.exception :=
-    (s3_perfInfo.except0 && io.toIBuffer.fire) ||
-      (s3_perfInfo.hit0Except1 && io.toIBuffer.fire) ||
-      (s3_perfInfo.miss0Except1 && io.toIBuffer.fire)
-  fetchIBufferDumpData.isCacheHit := s3_hit
+  fetchIBufferDumpData.exception  := io.toIBuffer.fire && s3_perfInfo.except
+  fetchIBufferDumpData.isCacheHit := io.toIBuffer.fire && s3_perfInfo.hit
 
   private val ifuWbToFtqDumpData = Wire(new IfuWbToFtqDB)
   ifuWbToFtqDumpData.startAddr         := wbFtqReq.startAddr.toUInt
