diff --git a/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtbInternalBank.scala b/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtbInternalBank.scala
index 27e4c277bf0..60cd677138c 100644
--- a/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtbInternalBank.scala
+++ b/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtbInternalBank.scala
@@ -195,6 +195,6 @@ class MainBtbInternalBank(
 
   XSPerfAccumulate(
     "counter_writebuffer_drop_write",
-    counterWriteBuffer.io.enq.ready && counterWriteBuffer.io.enq.valid
+    !counterWriteBuffer.io.enq.ready && counterWriteBuffer.io.enq.valid
   )
 }
