diff --git a/src/main/scala/xiangshan/cache/mmu/BitmapCheck.scala b/src/main/scala/xiangshan/cache/mmu/BitmapCheck.scala
index 6fffc3f7136..26691e31cc4 100644
--- a/src/main/scala/xiangshan/cache/mmu/BitmapCheck.scala
+++ b/src/main/scala/xiangshan/cache/mmu/BitmapCheck.scala
@@ -422,14 +422,23 @@ class BitmapCache(implicit p: Parameters) extends XSModule with HasPtwConst {
   // -----
   val addr_search = io.req.bits.tag
   val hitVecT = bitmapcache.map(_.hit(addr_search))
+  val hitIdxT = PriorityEncoder(hitVecT)
 
   // -----
   // -S1--
   // -----
+  val refillindex = bitmapReplace.way
+  dontTouch(refillindex)
+  val replacedata = RegEnable(bitmapcache(refillindex).data, io.refill.valid)
+  val replaceVec = RegNext(UIntToOH(refillindex))
+  val replaceval = RegNext(io.refill.valid)
+
   val index = RegEnable(addr_search(log2Up(XLEN)-1,0), io.req.fire)
   val order = RegEnable(io.req.bits.order, io.req.fire)
   val hitVec = RegEnable(VecInit(hitVecT), io.req.fire)
-  val CacheData = RegEnable(ParallelPriorityMux(hitVecT zip bitmapcache.map(_.data)), io.req.fire)
+  val hitIdx = RegEnable(hitIdxT, io.req.fire)
+  val CacheDatas = VecInit((0 until bitmapCachesize).map(i => {Mux(replaceVec(i).asBool && replaceval, replacedata, bitmapcache(i).data)}))
+  val CacheData = CacheDatas(hitIdx)
   val cfs = Wire(Vec(tlbcontiguous, Bool()))
 
   val cfsdata = CacheData.asTypeOf(Vec(XLEN/8, UInt(8.W)))(index(log2Up(XLEN)-1, log2Up(8)))
@@ -467,8 +476,6 @@ class BitmapCache(implicit p: Parameters) extends XSModule with HasPtwConst {
   val rf_data = io.refill.bits.data
   val rf_vd = io.refill.valid
   when (!flush && rf_vd) {
-    val refillindex = bitmapReplace.way
-    dontTouch(refillindex)
     bitmapcache(refillindex).refill(rf_addr,rf_data,true.B)
     bitmapReplace.access(refillindex)
   }
diff --git a/src/main/scala/xiangshan/cache/mmu/MMUConst.scala b/src/main/scala/xiangshan/cache/mmu/MMUConst.scala
index 8c4046794dc..41e4d69d10a 100644
--- a/src/main/scala/xiangshan/cache/mmu/MMUConst.scala
+++ b/src/main/scala/xiangshan/cache/mmu/MMUConst.scala
@@ -57,14 +57,14 @@ case class L2TLBParameters
   l2Associative: String = "fa",
   l2Replacer: Option[String] = Some("plru"),
   // l1
-  l1nSets: Int = 8,
+  l1nSets: Int = 4,
   l1nWays: Int = 2,
-  l1ReservedBits: Int = 10,
+  l1ReservedBits: Int = 2,
   l1Replacer: Option[String] = Some("setplru"),
   // l0
-  l0nSets: Int = 32,
+  l0nSets: Int = 64,
   l0nWays: Int = 4,
-  l0ReservedBits: Int = 3,
+  l0ReservedBits: Int = 4,
   l0Replacer: Option[String] = Some("setplru"),
   // sp
   spSize: Int = 16,
diff --git a/src/main/scala/xiangshan/cache/mmu/PageTableCache.scala b/src/main/scala/xiangshan/cache/mmu/PageTableCache.scala
index 404c0847503..cb1e005bf1b 100644
--- a/src/main/scala/xiangshan/cache/mmu/PageTableCache.scala
+++ b/src/main/scala/xiangshan/cache/mmu/PageTableCache.scala
@@ -268,7 +268,7 @@ class PtwCache()(implicit p: Parameters) extends XSModule with HasPtwConst with
     set = l2tlbParams.l1nSets,
     way = l2tlbParams.l1nWays,
     waySplit = 1,
-    dataSplit = 4,
+    dataSplit = 1,
     singlePort = sramSinglePort,
     readMCP2 = false,
     hasMbist = hasMbist,
@@ -307,7 +307,7 @@ class PtwCache()(implicit p: Parameters) extends XSModule with HasPtwConst with
     set = l2tlbParams.l0nSets,
     way = l2tlbParams.l0nWays,
     waySplit = 2,
-    dataSplit = 4,
+    dataSplit = 8,
     singlePort = sramSinglePort,
     readMCP2 = false,
     hasMbist = hasMbist,
@@ -632,11 +632,14 @@ class PtwCache()(implicit p: Parameters) extends XSModule with HasPtwConst with
     val wakeup_hitVec_delay = VecInit(wakeup_data_resp.zip(wakeup_vVec_delay.asBools).zip(wakeup_hVec_delay).map { case ((wayData, v), h) =>
       wayData.entries.hit(wakeup_delay_vpn, io.csr_dup(0).satp.asid, io.csr_dup(0).vsatp.asid, io.csr_dup(0).hgatp.vmid, s2xlate = wakeup_delay_h) && v && (wakeup_delay_h === h)})
 
-    val wakeup_setindex_delay = wakeup_req_delay.setIndex
-    val wakeup_way_info_delay = wakeup_req_delay.way_info
+    val wakeup_req_valid = OneCycleValid(wakeup_req_delay_valid, flush)
+    val wakeup_req = RegEnable(wakeup_req_delay, wakeup_req_delay_valid)
+    val wakeup_hitVec = RegEnable(wakeup_hitVec_delay, wakeup_req_delay_valid)
+    val wakeup_setindex = wakeup_req.setIndex
+    val wakeup_way_info = wakeup_req.way_info
 
-    when (wakeup_req_delay_valid && wakeup_hitVec_delay(OHToUInt(wakeup_way_info_delay))) {
-      l0BitmapReg(wakeup_setindex_delay)(OHToUInt(wakeup_way_info_delay))(wakeup_req_delay.pte_index) := wakeup_req_delay.check_success
+    when (wakeup_req_valid && wakeup_hitVec(OHToUInt(wakeup_way_info))) {
+      l0BitmapReg(wakeup_setindex)(OHToUInt(wakeup_way_info))(wakeup_req.pte_index) := wakeup_req.check_success
     }
   }
 
diff --git a/src/main/scala/xiangshan/cache/mmu/PageTableWalker.scala b/src/main/scala/xiangshan/cache/mmu/PageTableWalker.scala
index dd4546aa5e3..997ed646f25 100644
--- a/src/main/scala/xiangshan/cache/mmu/PageTableWalker.scala
+++ b/src/main/scala/xiangshan/cache/mmu/PageTableWalker.scala
@@ -109,14 +109,12 @@ class PTW()(implicit p: Parameters) extends XSModule with HasPtwConst with HasPe
   val mbmc = io.csr.mbmc
   val bitmap_enable = (if (HasBitmapCheck) true.B else false.B) && mbmc.BME === 1.U && mbmc.CMODE === 0.U
 
-  val satp = Wire(new TlbSatpBundle())
-  when (io.req.fire) {
-    satp := Mux(io.req.bits.req_info.s2xlate =/= noS2xlate, io.csr.vsatp, io.csr.satp)
-  } .otherwise {
-    satp := Mux(enableS2xlate, io.csr.vsatp, io.csr.satp)
-  }
+  // when req.fire, should use current_satp
+  val current_satp = Mux(io.req.bits.req_info.s2xlate =/= noS2xlate, io.csr.vsatp, io.csr.satp)
+  val satp = Mux(enableS2xlate, io.csr.vsatp, io.csr.satp)
   val s1Pbmte = Mux(req_s2xlate =/= noS2xlate, io.csr.hPBMTE, io.csr.mPBMTE)
 
+  val current_mode = current_satp.mode
   val mode = satp.mode
   val hgatp = io.csr.hgatp
   val flush = io.sfence.valid || io.csr.satp.changed || io.csr.vsatp.changed || io.csr.hgatp.changed || io.csr.priv.virt_changed
@@ -341,28 +339,28 @@ class PTW()(implicit p: Parameters) extends XSModule with HasPtwConst with HasPe
     val req = io.req.bits
     val gvpn_wire = Wire(UInt(ptePPNLen.W))
     if (EnableSv48) {
-      when (mode === Sv48) {
+      when (current_mode === Sv48) {
         level := Mux(req.l2Hit, 1.U, Mux(req.l3Hit.get, 2.U, 3.U))
         af_level := Mux(req.l2Hit, 1.U, Mux(req.l3Hit.get, 2.U, 3.U))
         gpf_level := Mux(req.l2Hit, 2.U, Mux(req.l3Hit.get, 3.U, 0.U))
-        ppn := Mux(req.l2Hit || req.l3Hit.get, io.req.bits.ppn, satp.ppn)
+        ppn := Mux(req.l2Hit || req.l3Hit.get, io.req.bits.ppn, current_satp.ppn)
         l3Hit := req.l3Hit.get
-        gvpn_wire := Mux(req.l2Hit || req.l3Hit.get, io.req.bits.ppn, satp.ppn)
+        gvpn_wire := Mux(req.l2Hit || req.l3Hit.get, io.req.bits.ppn, current_satp.ppn)
       } .otherwise {
         level := Mux(req.l2Hit, 1.U, 2.U)
         af_level := Mux(req.l2Hit, 1.U, 2.U)
         gpf_level := Mux(req.l2Hit, 2.U, 0.U)
-        ppn := Mux(req.l2Hit, io.req.bits.ppn, satp.ppn)
+        ppn := Mux(req.l2Hit, io.req.bits.ppn, current_satp.ppn)
         l3Hit := false.B
-        gvpn_wire := Mux(req.l2Hit, io.req.bits.ppn, satp.ppn)
+        gvpn_wire := Mux(req.l2Hit, io.req.bits.ppn, current_satp.ppn)
       }
     } else {
       level := Mux(req.l2Hit, 1.U, 2.U)
       af_level := Mux(req.l2Hit, 1.U, 2.U)
       gpf_level := Mux(req.l2Hit, 2.U, 0.U)
-      ppn := Mux(req.l2Hit, io.req.bits.ppn, satp.ppn)
+      ppn := Mux(req.l2Hit, io.req.bits.ppn, current_satp.ppn)
       l3Hit := false.B
-      gvpn_wire := Mux(req.l2Hit, io.req.bits.ppn, satp.ppn)
+      gvpn_wire := Mux(req.l2Hit, io.req.bits.ppn, current_satp.ppn)
     }
     vpn := io.req.bits.req_info.vpn
     l2Hit := req.l2Hit
@@ -744,15 +742,11 @@ class LLPTW(implicit p: Parameters) extends XSModule with HasPtwConst with HasPe
     mem_arb.io.in(i).valid := is_mems(i) && !io.mem.req_mask(i)
   }
 
-  // when an entry dup with another entry whose mem_arb.io.out.fire,
-  // should block hptw req and change state(i) to state_mem_waiting
-  val block_hptw_req = WireInit(VecInit(Seq.fill(l2tlbParams.llptwsize)(false.B)))
-
   // process hptw requests in serial
   val hyper_arb1 = Module(new RRArbiterInit(new LLPTWEntry(), l2tlbParams.llptwsize))
   for (i <- 0 until l2tlbParams.llptwsize) {
     hyper_arb1.io.in(i).bits := entries(i)
-    hyper_arb1.io.in(i).valid := is_hptw_req(i) && !(Cat(is_hptw_resp).orR) && !(Cat(is_last_hptw_resp).orR) && !block_hptw_req(i)
+    hyper_arb1.io.in(i).valid := is_hptw_req(i) && !(Cat(is_hptw_resp).orR) && !(Cat(is_last_hptw_resp).orR)
   }
   val hyper_arb2 = Module(new RRArbiterInit(new LLPTWEntry(), l2tlbParams.llptwsize))
   for(i <- 0 until l2tlbParams.llptwsize) {
@@ -834,7 +828,7 @@ class LLPTW(implicit p: Parameters) extends XSModule with HasPtwConst with HasPe
     to_hptw_req -> state_hptw_req
   ))
   val enq_state = Mux(from_pre(io.in.bits.req_info.source) && enq_state_normal =/= state_addr_check, state_idle, enq_state_normal)
-  when (io.in.fire  && (if (HasBitmapCheck) !io.in.bits.bitmapCheck.get.jmp_bitmap_check else true.B)) {
+  when (io.in.fire && (if (HasBitmapCheck) !io.in.bits.bitmapCheck.get.jmp_bitmap_check else true.B)) {
     // if prefetch req does not need mem access, just give it up.
     // so there will be at most 1 + FilterSize entries that needs re-access page cache
     // so 2 + FilterSize is enough to avoid dead-lock
@@ -887,11 +881,20 @@ class LLPTW(implicit p: Parameters) extends XSModule with HasPtwConst with HasPe
   val hptw_need_addr_check = RegNext(hasHptwResp && io.hptw.resp.fire && !flush) && state(hptw_resp_ptr_reg) === state_addr_check
 
   val ptes = io.mem.resp.bits.value.asTypeOf(Vec(blockBits / XLEN, new PteBundle()))
-  val gpaddr = MakeGPAddr(entries(hptw_resp_ptr_reg).ppn, getVpnn(entries(hptw_resp_ptr_reg).req_info.vpn, 0))
-  val hptw_resp = entries(hptw_resp_ptr_reg).hptw_resp
-  val hpaddr = Cat(hptw_resp.genPPNS2(get_pn(gpaddr)), get_off(gpaddr))
+  val gpaddr = MakeGPAddr(entries(io.hptw.resp.bits.id).ppn, getVpnn(entries(io.hptw.resp.bits.id).req_info.vpn, 0))
+  val hptw_resp = io.hptw.resp.bits.h_resp
+  val hpaddr = RegEnable(Cat(hptw_resp.genPPNS2(get_pn(gpaddr)), get_off(gpaddr)), io.hptw.resp.fire)
   val addr = RegEnable(MakeAddr(io.in.bits.ppn(ppnLen - 1, 0), getVpnn(io.in.bits.req_info.vpn, 0)), io.in.fire)
 
+  when (hyper_arb1.io.out.fire) {
+    for (i <- state.indices) {
+      when (state(i) === state_hptw_req && entries(i).ppn === hyper_arb1.io.out.bits.ppn && entries(i).req_info.s2xlate === allStage && hyper_arb1.io.chosen === i.U) {
+        state(i) := state_hptw_resp
+        entries(i).wait_id := hyper_arb1.io.chosen
+      }
+    }
+  }
+
   io.pmp(0).req.valid := need_addr_check
   io.pmp(0).req.bits.addr := addr
   io.pmp(0).req.bits.cmd := TlbCmd.read
@@ -924,10 +927,10 @@ class LLPTW(implicit p: Parameters) extends XSModule with HasPtwConst with HasPe
         state(i) := state_mem_waiting
         entries(i).hptw_resp := entries(mem_arb.io.chosen).hptw_resp
         entries(i).wait_id := mem_arb.io.chosen
-        block_hptw_req(i) := true.B
       }
     }
   }
+
   when (io.mem.resp.fire) {
     state.indices.map{i =>
       when (state(i) === state_mem_waiting && io.mem.resp.bits.id === entries(i).wait_id) {
@@ -959,15 +962,6 @@ class LLPTW(implicit p: Parameters) extends XSModule with HasPtwConst with HasPe
     }
   }
 
-  when (hyper_arb1.io.out.fire) {
-    for (i <- state.indices) {
-      when (state(i) === state_hptw_req && entries(i).ppn === hyper_arb1.io.out.bits.ppn && entries(i).req_info.s2xlate === allStage && hyper_arb1.io.chosen === i.U) {
-        state(i) := state_hptw_resp
-        entries(i).wait_id := hyper_arb1.io.chosen
-      }
-    }
-  }
-
   when (hyper_arb2.io.out.fire) {
     for (i <- state.indices) {
       when (state(i) === state_last_hptw_req && entries(i).ppn === hyper_arb2.io.out.bits.ppn && entries(i).req_info.s2xlate === allStage && hyper_arb2.io.chosen === i.U) {
@@ -986,7 +980,9 @@ class LLPTW(implicit p: Parameters) extends XSModule with HasPtwConst with HasPe
         }
       }
     }
+  }
 
+  if (HasBitmapCheck) {
     when (io.bitmap.get.resp.fire) {
       for (i <- state.indices) {
         when (is_bitmap_resp(i) && io.bitmap.get.resp.bits.id === entries(i).wait_id) {
@@ -1025,6 +1021,7 @@ class LLPTW(implicit p: Parameters) extends XSModule with HasPtwConst with HasPe
       }
     }
   }
+
   when (io.out.fire) {
     assert(state(mem_ptr) === state_mem_out)
     state(mem_ptr) := state_idle
