diff --git a/src/main/scala/xiangshan/backend/fu/util/Trigger.scala b/src/main/scala/xiangshan/backend/fu/util/Trigger.scala
index b51c996df8c..3f27b8d05f6 100644
--- a/src/main/scala/xiangshan/backend/fu/util/Trigger.scala
+++ b/src/main/scala/xiangshan/backend/fu/util/Trigger.scala
@@ -289,23 +289,20 @@ trait SdtrigExt extends HasXSParameter{
     val overallLess = Wire(Vec(pcVecWidth, Bool()))
 
     overallEqual.zipWithIndex.map{case (o, i) => o :=
-      (!carry(i) && lowPCEqual(i) && highPCEqual) ||
+      !carry(i) && highPCEqual  && lowPCEqual(i) ||
         carry(i) && highPC1Equal && lowPCEqual(i)
     }
 
     // [left : greater 1. nonCarry 2. carry
     overallGreater.zipWithIndex.map { case (o, i) => o :=
-      !carry(i) && (highPCGreater || highPCEqual && lowPCGreater(i)) ||
-       carry(i) && (highPCGreater || highPCEqual || highPC1Equal && lowPCGreater(i))
+      !carry(i) && (highPCGreater || highPCEqual && (lowPCEqual(i) || lowPCGreater(i))) ||
+       carry(i) && (highPC1Greater || highPC1Equal && (lowPCEqual(i) || lowPCGreater(i)))
     }
 
-    // right) : less 1. nonCarryRight 2. carryRight
-    val carryRight = Wire(Vec(pcVecWidth, Bool()))
-    carryRight.zipWithIndex.map{case (c, i) => c := pcVec(i)(highPos) =/= pcVec(pcVecWidth - 1)(highPos)}
-
+    // right) : less 1. nonCarry 2. carry
     overallLess.zipWithIndex.map { case (o, i) => o :=
-      !carryRight(i) && (highPCLess || highPCEqual && lowPCLess(i)) ||
-       carryRight(i) && (highPC1Less || highPC1Equal || highPCEqual && lowPCLess(i))
+      !carry(i) && (highPCLess || highPCEqual && lowPCLess(i)) ||
+       carry(i) && (highPC1Less || highPC1Equal && lowPCLess(i))
     }
 
     val ret = Wire(Vec(pcVecWidth, Bool()))
diff --git a/src/test/scala/xiangshan/frontend/FrontTrigger.scala b/src/test/scala/xiangshan/frontend/FrontTrigger.scala
new file mode 100644
index 00000000000..17b4e26205d
--- /dev/null
+++ b/src/test/scala/xiangshan/frontend/FrontTrigger.scala
@@ -0,0 +1,47 @@
+package xiangshan.frontend
+
+import chisel3._
+import chiseltest._
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.must.Matchers
+import top.DefaultConfig
+import utility.{LogUtilsOptions, LogUtilsOptionsKey}
+import xiangshan.{DebugOptionsKey, XSCoreParameters, XSCoreParamsKey}
+
+
+class FrontendTriggerTest extends AnyFlatSpec with ChiselScalatestTester with Matchers {
+  val defaultConfig = (new DefaultConfig).alterPartial {
+    case XSCoreParamsKey => XSCoreParameters()
+  }.alter((site, here, up) => {
+    case LogUtilsOptionsKey => LogUtilsOptions(
+      here(DebugOptionsKey).EnableDebug,
+      here(DebugOptionsKey).EnablePerfDebug,
+      here(DebugOptionsKey).FPGAPlatform
+    )
+  })
+
+  println("test start")
+
+  behavior of "FrontendTrigger"
+  it should "run" in {
+    test(new FrontendTrigger()(defaultConfig)).withAnnotations(Seq(VerilatorBackendAnnotation)) {
+      m: FrontendTrigger =>
+        m.io.frontendTrigger.debugMode.poke(false.B)
+        m.io.frontendTrigger.tEnableVec.map(_.poke(false.B))
+        m.io.frontendTrigger.tEnableVec(1).poke(true.B)
+        m.io.frontendTrigger.triggerCanRaiseBpExp.poke(true.B)
+        m.io.frontendTrigger.tUpdate.valid.poke(true.B)
+        m.io.frontendTrigger.tUpdate.bits.addr.poke(1.U)
+        m.io.frontendTrigger.tUpdate.bits.tdata.matchType.poke(3.U)
+        m.io.frontendTrigger.tUpdate.bits.tdata.tdata2.poke("h1234_567f".U)
+        m.io.pc.zipWithIndex.map { case (pc, i) =>
+          pc.poke((0x1234_5676 + 2 * i).U)
+        }
+        for (_ <- 0 until 4) { m.clock.step() }
+        m.io.triggered.zipWithIndex.map { case (action, i) =>
+          println(s"out.triggerAction${i}：" + action.peek().litValue.toString(16) + "\n")
+        }
+    }
+    println("test done")
+  }
+}
