diff --git a/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala b/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala
index cd475832cf9..5a578953808 100644
--- a/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala
+++ b/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala
@@ -817,7 +817,11 @@ class StoreQueue(implicit p: Parameters) extends XSModule
     }
     is(s_wb) {
       when (io.mmioStout.fire || io.vecmmioStout.fire) {
-        uncacheState := s_wait
+        when (uncacheUop.exceptionVec(storeAccessFault)) {
+          uncacheState := s_idle
+        }.otherwise {
+          uncacheState := s_wait
+        }
       }
     }
     is(s_wait) {
