diff --git a/src/main/scala/xiangshan/frontend/bpu/tage/Tage.scala b/src/main/scala/xiangshan/frontend/bpu/tage/Tage.scala
index a899d2c1df4..dba65b9b63a 100644
--- a/src/main/scala/xiangshan/frontend/bpu/tage/Tage.scala
+++ b/src/main/scala/xiangshan/frontend/bpu/tage/Tage.scala
@@ -439,8 +439,8 @@ class Tage(implicit p: Parameters) extends BasePredictor with HasTageParameters
   t2_allocateEntry.tag   := Mux1H(t2_allocateTableMaskOH, t2_tempTag) ^ t2_allocateBranch.bits.cfiPosition
   t2_allocateEntry.takenCtr.value := Mux(
     t2_allocateBranch.bits.taken,
-    (1 << (TakenCtrWidth - 1)).U,    // weak taken
-    (1 << (TakenCtrWidth - 1) - 1).U // weak not taken
+    (1 << (TakenCtrWidth - 1)).U,      // weak taken
+    ((1 << (TakenCtrWidth - 1)) - 1).U // weak not taken
   )
 
   tables.zipWithIndex.foreach { case (table, tableIdx) =>
