diff --git a/src/main/scala/xiangshan/frontend/icache/ICacheMainPipe.scala b/src/main/scala/xiangshan/frontend/icache/ICacheMainPipe.scala
index 458b52efd3f..b65be2acf2b 100644
--- a/src/main/scala/xiangshan/frontend/icache/ICacheMainPipe.scala
+++ b/src/main/scala/xiangshan/frontend/icache/ICacheMainPipe.scala
@@ -255,12 +255,17 @@ class ICacheMainPipe(implicit p: Parameters) extends ICacheModule with HasICache
   private val s1_req_offset  = s1_req_vaddr(0)(log2Ceil(blockBytes) - 1, 0)
 
   // do metaArray ECC check
-  private val s1_meta_corrupt = VecInit((s1_req_ptags zip s1_meta_codes zip s1_waymasks).map {
-    case ((meta, code), waymask) =>
-      val hit_num = PopCount(waymask)
+  private val s1_meta_corrupt = VecInit((0 until PortNumber).map { i =>
+    val hit_num = PopCount(s1_waymasks(i))
+    // if !s1_doubleline, s1_waymasks(1) is invalid and should not be checked
+    val needThisLine = if (i == 0) true.B else s1_doubleline
+    needThisLine && (
       // NOTE: if not hit, encodeMetaECC(meta) =/= code can also be true, but we don't care about it
-      (encodeMetaECC(meta) =/= code && hit_num === 1.U) || // hit one way, but parity code does not match, ECC failure
-      hit_num > 1.U                                        // hit multi-way, must be an ECC failure
+      // hit one way, but parity code does not match, ECC failure
+      (encodeMetaECC(s1_req_ptags(i)) =/= s1_meta_codes(i) && hit_num === 1.U) ||
+        // hit multi-way, must be an ECC failure
+        hit_num > 1.U
+    )
   })
   // force clear meta_corrupt when parity check is disabled
   when(!ecc_enable) {
@@ -388,6 +393,18 @@ class ICacheMainPipe(implicit p: Parameters) extends ICacheModule with HasICache
   }
   // meta error is checked in s1 stage
   private val s2_meta_corrupt = RegEnable(s1_meta_corrupt, 0.U.asTypeOf(s1_meta_corrupt), s1_fire)
+
+  /* NOTE: if !s2_doubline:
+   * - s2_meta_corrupt(1) should be false.B (as waymask(1) is invalid, and meta ecc is not checked)
+   * - s2_data_corrupt(1) should also be false.B, as getLineSel() should not select line(1)
+   * so we don't need to check s2_doubleline in the following io.errors and toMetaFlush ports
+   * we add a sanity check to make sure the above assumption holds
+   */
+  assert(
+    !(!s2_doubleline && (s2_meta_corrupt(1) || s2_data_corrupt(1))),
+    "meta or data corrupt detected on line 1 but s2_doubleline is false.B"
+  )
+
   // send errors to top
   // TODO: support RERI spec standard interface
   (0 until PortNumber).foreach { i =>
