diff --git a/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtb.scala b/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtb.scala
index b0eb8cc6fc5..2b1a0ac64ba 100644
--- a/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtb.scala
+++ b/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtb.scala
@@ -142,10 +142,11 @@ class MainBtb(implicit p: Parameters) extends BasePredictor with HasMainBtbParam
   private val t1_writeAlignBankMask = t1_rotator.rotate(VecInit(UIntToOH(t1_writeAlignBankIdx).asBools))
 
   alignBanks.zipWithIndex.foreach { case (b, i) =>
-    b.io.write.req.valid         := t1_fire && t1_writeAlignBankMask(i)
-    b.io.write.req.bits.startPc  := t1_startPcVec(i)
-    b.io.write.req.bits.branches := t1_train.branches
-    b.io.write.req.bits.meta     := t1_meta.entries(i)
+    b.io.write.req.valid          := t1_fire
+    b.io.write.req.bits.needWrite := t1_writeAlignBankMask(i)
+    b.io.write.req.bits.startPc   := t1_startPcVec(i)
+    b.io.write.req.bits.branches  := t1_train.branches
+    b.io.write.req.bits.meta      := t1_meta.entries(i)
     // see comments in MainBtbAlignBank.scala
     b.io.write.req.bits.mispredictInfo := t1_mispredictInfo
   }
diff --git a/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtbAlignBank.scala b/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtbAlignBank.scala
index fabd3fdfda3..a970e0f82db 100644
--- a/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtbAlignBank.scala
+++ b/src/main/scala/xiangshan/frontend/bpu/mbtb/MainBtbAlignBank.scala
@@ -50,6 +50,7 @@ class MainBtbAlignBank(
 
     class Write extends Bundle {
       class Req extends Bundle {
+        val needWrite: Bool = Bool()
         // similar to Read.Req.startPc, calculated in MainBtb top
         val startPc:  PrunedAddr             = new PrunedAddr(VAddrBits)
         val branches: Vec[Valid[BranchInfo]] = Vec(ResolveEntryBranchNumber, Valid(new BranchInfo))
@@ -196,6 +197,7 @@ class MainBtbAlignBank(
    * send write req to internal banks (srams)
    */
   private val t1_fire             = w.req.valid
+  private val t1_needWrite        = w.req.bits.needWrite
   private val t1_startPc          = w.req.bits.startPc
   private val t1_branches         = w.req.bits.branches
   private val t1_meta             = w.req.bits.meta
@@ -211,7 +213,7 @@ class MainBtbAlignBank(
   private val t1_hit     = t1_hitMask.orR
 
   // Write entry only when there's a mispredict, and if:
-  private val t1_entryNeedWrite = t1_mispredictInfo.valid && (
+  private val t1_entryNeedWrite = t1_needWrite && t1_mispredictInfo.valid && (
     // 1. not hit, always write a new entry, use mbtb replacer's victim way.
     !t1_hit ||
       // 2. hit, do write only if:
