diff --git a/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala b/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala
index 9541c561bf9..cd475832cf9 100644
--- a/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala
+++ b/src/main/scala/xiangshan/mem/lsqueue/StoreQueue.scala
@@ -1045,13 +1045,19 @@ class StoreQueue(implicit p: Parameters) extends XSModule
   // If the last flow with an exception is the LastFlow of this instruction, the flag is not set.
   // compare robidx to select the last flow
   require(EnsbufferWidth == 2, "The vector store exception handle process only support EnsbufferWidth == 2 yet.")
-  val robidxEQ = uop(rdataPtrExt(0).value).robIdx === uop(rdataPtrExt(1).value).robIdx
+  val robidxEQ = dataBuffer.io.enq(0).valid && dataBuffer.io.enq(1).valid &&
+    uop(rdataPtrExt(0).value).robIdx === uop(rdataPtrExt(1).value).robIdx
+  val robidxNE = dataBuffer.io.enq(0).valid && dataBuffer.io.enq(1).valid && (
+    uop(rdataPtrExt(0).value).robIdx =/= uop(rdataPtrExt(1).value).robIdx
+  )
+  val onlyCommit0 = dataBuffer.io.enq(0).valid && !dataBuffer.io.enq(1).valid
 
   val vecCommitLastFlow = 
     // robidx equal => check if 1 is last flow
     robidxEQ && vecCommitHasExceptionLastFlow(1) || 
     // robidx not equal => 0 must be the last flow, just check if 1 is last flow when 1 has exception
-    !robidxEQ && vecCommitHasExceptionValid(1) && vecCommitHasExceptionLastFlow(1)
+    robidxNE && (vecCommitHasExceptionValid(1) && vecCommitHasExceptionLastFlow(1) || !vecCommitHasExceptionValid(1)) ||
+    onlyCommit0 && vecCommitHasExceptionLastFlow(0)
   
 
   val vecExceptionFlagCancel  = (0 until EnsbufferWidth).map{ i =>
