diff --git a/src/main/scala/xiangshan/frontend/ibuffer/Bundles.scala b/src/main/scala/xiangshan/frontend/ibuffer/Bundles.scala
index e3dbf365589..94624864553 100644
--- a/src/main/scala/xiangshan/frontend/ibuffer/Bundles.scala
+++ b/src/main/scala/xiangshan/frontend/ibuffer/Bundles.scala
@@ -127,6 +127,7 @@ class IBufEntry(implicit p: Parameters) extends IBufferBundle {
     result.foldpc           := foldpc
     result.pd               := pd
     result.predTaken        := predTaken
+    result.identifiedCfi    := identifiedCfi
     result.ftqPtr           := ftqPtr
     result.exceptionType    := exceptionType
     result.backendException := backendException
@@ -147,6 +148,7 @@ class IBufOutEntry(implicit p: Parameters) extends IBufferBundle {
   val foldpc:           UInt          = UInt(MemPredPCWidth.W)
   val pd:               PreDecodeInfo = new PreDecodeInfo
   val predTaken:        Bool          = Bool()
+  val identifiedCfi:    Bool          = Bool()
   val ftqPtr:           FtqPtr        = new FtqPtr
   val exceptionType:    UInt          = IBufferExceptionType()
   val backendException: Bool          = Bool()
@@ -170,7 +172,7 @@ class IBufOutEntry(implicit p: Parameters) extends IBufferBundle {
     cf.trigger                                       := triggered
     cf.pd                                            := pd
     cf.pred_taken                                    := predTaken
-    cf.identifiedCfi                                 := instrEndOffset
+    cf.identifiedCfi                                 := identifiedCfi
     cf.crossPageIPFFix                               := IBufferExceptionType.isCrossPage(exceptionType)
     cf.storeSetHit                                   := DontCare
     cf.waitForRobIdx                                 := DontCare
