diff --git a/src/main/scala/xiangshan/backend/fu/NewCSR/CSREvents/TrapEntryDEvent.scala b/src/main/scala/xiangshan/backend/fu/NewCSR/CSREvents/TrapEntryDEvent.scala
index e8dee70914a..f46aecdb389 100644
--- a/src/main/scala/xiangshan/backend/fu/NewCSR/CSREvents/TrapEntryDEvent.scala
+++ b/src/main/scala/xiangshan/backend/fu/NewCSR/CSREvents/TrapEntryDEvent.scala
@@ -50,6 +50,7 @@ class TrapEntryDEventModule(implicit val p: Parameters) extends Module with CSRE
   private val hasDebugEbreakException      = in.hasDebugEbreakException
   private val hasSingleStep                = in.hasSingleStep
   private val criticalErrorStateEnterDebug = in.criticalErrorStateEnterDebug
+  private val isFetchMalAddr               = in.isFetchMalAddr
 
   private val hasExceptionInDmode = debugMode && hasTrap
   val causeIntr = DcsrCause.Haltreq.asUInt
@@ -88,7 +89,7 @@ class TrapEntryDEventModule(implicit val p: Parameters) extends Module with CSRE
   out.dcsr.bits.V             := current.privState.V.asUInt
   out.dcsr.bits.PRV           := current.privState.PRVM.asUInt
   out.dcsr.bits.CAUSE         := Mux(hasDebugIntr, causeIntr, causeExp)
-  out.dpc.bits.epc            := trapPC(63, 1)
+  out.dpc.bits.epc            := Mux(isFetchMalAddr, in.fetchMalTval(63, 1), trapPC(63, 1))
 
   out.targetPc.bits.pc        := debugPc
   out.targetPc.bits.raiseIPF  := false.B
