diff --git a/src/main/scala/xiangshan/backend/dispatch/Dispatch.scala b/src/main/scala/xiangshan/backend/dispatch/Dispatch.scala
index f9a6ca9a3e0..573b0f66d5e 100644
--- a/src/main/scala/xiangshan/backend/dispatch/Dispatch.scala
+++ b/src/main/scala/xiangshan/backend/dispatch/Dispatch.scala
@@ -200,23 +200,26 @@ class Dispatch(implicit p: Parameters) extends XSModule with HasPerfEvents {
   val s_holdRobidx :: s_updateRobidx :: Nil = Enum(2)
   val singleStepState = RegInit(s_updateRobidx)
   
-  val robidxStepNext  = WireInit(0.U.asTypeOf(io.fromRename(0).bits.robIdx))
+  val robidxStepHold  = WireInit(0.U.asTypeOf(io.fromRename(0).bits.robIdx))
   val robidxStepReg   = RegInit(0.U.asTypeOf(io.fromRename(0).bits.robIdx))
   val robidxCanCommitStepping = WireInit(0.U.asTypeOf(io.fromRename(0).bits.robIdx))
+  robidxStepReg := robidxCanCommitStepping
   
   when(!io.singleStep) {
     singleStepState := s_updateRobidx
   }.elsewhen(io.singleStep && io.fromRename(0).fire && io.enqRob.req(0).valid) {
     singleStepState := s_holdRobidx
-    robidxStepNext := io.fromRename(0).bits.robIdx
+    robidxStepHold := io.fromRename(0).bits.robIdx
   }
   
   when(singleStepState === s_updateRobidx) {
-    robidxStepReg := robidxStepNext
-    robidxCanCommitStepping := robidxStepNext
+    robidxCanCommitStepping := robidxStepHold
   }.elsewhen(singleStepState === s_holdRobidx) {
-    robidxStepReg := robidxStepReg 
-    robidxCanCommitStepping := robidxStepReg
+    when(io.redirect.valid){
+      robidxCanCommitStepping.flag := !robidxStepReg.flag
+    }.otherwise {
+      robidxCanCommitStepping := robidxStepReg
+    }
   }
 
   val updatedUop = Wire(Vec(RenameWidth, new DynInst))
@@ -247,7 +250,7 @@ class Dispatch(implicit p: Parameters) extends XSModule with HasPerfEvents {
     } else {
       updatedUop(i).loadWaitBit := isLs(i) && !isStore(i) && io.fromRename(i).bits.loadWaitBit
     }
-    // // update singleStep, singleStep exception only enable in next machine instruction.
+    // update singleStep, singleStep exception only enable in next machine instruction.
     updatedUop(i).singleStep := io.singleStep && (io.fromRename(i).bits.robIdx =/= robidxCanCommitStepping)
     when (io.fromRename(i).fire) {
       XSDebug(TriggerAction.isDmode(updatedUop(i).trigger) || updatedUop(i).exceptionVec(breakPoint), s"Debug Mode: inst ${i} has frontend trigger exception\n")
