diff --git a/src/main/scala/xiangshan/backend/exu/ExeUnit.scala b/src/main/scala/xiangshan/backend/exu/ExeUnit.scala
index 04052010439..a7d436a7773 100644
--- a/src/main/scala/xiangshan/backend/exu/ExeUnit.scala
+++ b/src/main/scala/xiangshan/backend/exu/ExeUnit.scala
@@ -400,7 +400,8 @@ class ExeUnitImp(
   io.vxrm.foreach(exuio => funcUnits.foreach(fu => fu.io.vxrm.foreach(fuio => fuio <> exuio)))
   io.vlIsZero.foreach(exuio => funcUnits.foreach(fu => fu.io.vlIsZero.foreach(fuio => exuio := fuio)))
   io.vlIsVlmax.foreach(exuio => funcUnits.foreach(fu => fu.io.vlIsVlmax.foreach(fuio => exuio := fuio)))
-  io.instrAddrTransType.foreach(exuio => funcUnits.foreach(fu => fu.io.instrAddrTransType.foreach(fuio => fuio := exuio)))
+  // RegNext for better timing and it should be fine
+  io.instrAddrTransType.foreach(exuio => funcUnits.foreach(fu => fu.io.instrAddrTransType.foreach(fuio => fuio := RegNext(exuio))))
 
   // debug info
   io.out.bits.debug     := 0.U.asTypeOf(io.out.bits.debug)
diff --git a/src/main/scala/xiangshan/backend/fu/wrapper/BranchUnit.scala b/src/main/scala/xiangshan/backend/fu/wrapper/BranchUnit.scala
index d11c67e1070..7614b57a97f 100644
--- a/src/main/scala/xiangshan/backend/fu/wrapper/BranchUnit.scala
+++ b/src/main/scala/xiangshan/backend/fu/wrapper/BranchUnit.scala
@@ -3,7 +3,7 @@ package xiangshan.backend.fu.wrapper
 import org.chipsalliance.cde.config.Parameters
 import chisel3._
 import chisel3.util.log2Up
-import utility.SignExt
+import utility.{SignExt, ZeroExt}
 import xiangshan.backend.decode.ImmUnion
 import xiangshan.backend.fu.{BranchModule, FuConfig, FuncUnit}
 import xiangshan.backend.datapath.DataConfig.VAddrData
@@ -11,19 +11,18 @@ import xiangshan.{RedirectLevel, SelImm, XSModule}
 
 class AddrAddModule(implicit p: Parameters) extends XSModule {
   val io = IO(new Bundle {
-    val pc = Input(UInt(VAddrBits.W))
+    val pcExtend = Input(UInt((VAddrBits + 1).W))
     val taken = Input(Bool())
     val isRVC = Input(Bool())
     val imm = Input(UInt(32.W)) // branch inst only support 12 bits immediate num
     val target = Output(UInt(XLEN.W))
     val nextPcOffset = Input(UInt((log2Up(PredictWidth) + 1).W))
   })
-  val pcExtend = SignExt(io.pc, VAddrBits + 1)
   val immMinWidth = FuConfig.BrhCfg.immType.map(x => SelImm.getImmUnion(x).len).max
   print(s"[Branch]: immMinWidth = $immMinWidth\n")
   io.target := SignExt(Mux(io.taken,
-  pcExtend + SignExt(io.imm(immMinWidth + 2, 0), VAddrBits + 1),
-  pcExtend + (io.nextPcOffset << instOffsetBits).asUInt
+    io.pcExtend + SignExt(io.imm(immMinWidth + 2, 0), VAddrBits + 1),
+    io.pcExtend + (io.nextPcOffset << instOffsetBits).asUInt
   ), XLEN)
 }
 
@@ -35,7 +34,11 @@ class BranchUnit(cfg: FuConfig)(implicit p: Parameters) extends FuncUnit(cfg) {
   dataModule.io.func := io.in.bits.ctrl.fuOpType
   dataModule.io.pred_taken := io.in.bits.ctrl.predictInfo.get.taken
 
-  addModule.io.pc := io.in.bits.data.pc.get // pc
+  val pcExtend = Mux(io.instrAddrTransType.get.shouldBeSext,
+    SignExt(io.in.bits.data.pc.get, VAddrBits + 1),
+    ZeroExt(io.in.bits.data.pc.get, VAddrBits + 1)
+  )
+  addModule.io.pcExtend := pcExtend
   addModule.io.imm := io.in.bits.data.imm // imm
   addModule.io.taken := dataModule.io.taken
   addModule.io.isRVC := io.in.bits.ctrl.preDecode.get.isRVC
diff --git a/src/main/scala/xiangshan/backend/fu/wrapper/JumpUnit.scala b/src/main/scala/xiangshan/backend/fu/wrapper/JumpUnit.scala
index c0c92731b90..24e3bdebe04 100644
--- a/src/main/scala/xiangshan/backend/fu/wrapper/JumpUnit.scala
+++ b/src/main/scala/xiangshan/backend/fu/wrapper/JumpUnit.scala
@@ -2,7 +2,7 @@ package xiangshan.backend.fu.wrapper
 
 import org.chipsalliance.cde.config.Parameters
 import chisel3._
-import fudian.SignExt
+import utility.{SignExt, ZeroExt}
 import xiangshan.RedirectLevel
 import xiangshan.backend.fu.{FuConfig, FuncUnit, JumpDataModule, PipedFuncUnit}
 import xiangshan.backend.datapath.DataConfig.VAddrData
@@ -15,7 +15,10 @@ class JumpUnit(cfg: FuConfig)(implicit p: Parameters) extends PipedFuncUnit(cfg)
 
   // associated with AddrData's position of JmpCfg.srcData
   private val src = io.in.bits.data.src(0)
-  private val pc = SignExt(io.in.bits.data.pc.get, cfg.destDataBits)
+  private val pc = Mux(io.instrAddrTransType.get.shouldBeSext,
+    SignExt(io.in.bits.data.pc.get, cfg.destDataBits),
+    ZeroExt(io.in.bits.data.pc.get, cfg.destDataBits)
+  )
   private val imm = io.in.bits.data.imm
   private val func = io.in.bits.ctrl.fuOpType
   private val isRVC = io.in.bits.ctrl.preDecode.get.isRVC
