diff --git a/src/main/scala/xiangshan/backend/issue/EnqEntry.scala b/src/main/scala/xiangshan/backend/issue/EnqEntry.scala
index f6ff0f948d2..c35999c4f35 100644
--- a/src/main/scala/xiangshan/backend/issue/EnqEntry.scala
+++ b/src/main/scala/xiangshan/backend/issue/EnqEntry.scala
@@ -125,11 +125,7 @@ class EnqEntry(isComp: Boolean)(implicit p: Parameters, params: IssueBlockParams
       val enqDelay1WakeupRCIdx = Mux1H(enqDelay1WakeupSrcExuWriteRC.map(_._1), enqDelay1WakeupSrcExuWriteRC.map(_._2.bits.rcDest.get))
       val enqDelay1ReplaceRC   = enqDelay1WakeupSrcExuWriteRC.map(x => x._2.bits.rfWen && x._2.bits.rcDest.get === entryReg.status.srcStatus(i).regCacheIdx.get).fold(false.B)(_ || _)
 
-      enqDelayUseRegCache.get(i)            := MuxCase(entryReg.status.srcStatus(i).useRegCache.get, Seq(
-                                                  enqDelayOut1.srcCancelByLoad(i)  -> false.B,
-                                                  enqDelay1WakeupRC                -> true.B,
-                                                  enqDelay1ReplaceRC               -> false.B,
-                                               ))
+      enqDelayUseRegCache.get(i)            := entryReg.status.srcStatus(i).useRegCache.get && !(enqDelayOut1.srcCancelByLoad(i) || enqDelay1ReplaceRC) || enqDelay1WakeupRC
       enqDelayRegCacheIdx.get(i)            := Mux(enqDelay1WakeupRC, enqDelay1WakeupRCIdx, entryReg.status.srcStatus(i).regCacheIdx.get)
     }
   }
@@ -154,7 +150,7 @@ class EnqEntry(isComp: Boolean)(implicit p: Parameters, params: IssueBlockParams
     }
   }
 
-  EntryRegCommonConnect(common, hasWakeupIQ, validReg, entryUpdate, entryReg, currentStatus, io.commonIn, true)
+  EntryRegCommonConnect(common, hasWakeupIQ, validReg, entryUpdate, entryReg, currentStatus, io.commonIn, true, isComp)
 
   //output
   CommonOutConnect(io.commonOut, common, hasWakeupIQ, validReg, entryUpdate, entryReg, currentStatus, io.commonIn, true, isComp)
diff --git a/src/main/scala/xiangshan/backend/issue/EntryBundles.scala b/src/main/scala/xiangshan/backend/issue/EntryBundles.scala
index a71bc55340d..ab21e0063d2 100644
--- a/src/main/scala/xiangshan/backend/issue/EntryBundles.scala
+++ b/src/main/scala/xiangshan/backend/issue/EntryBundles.scala
@@ -160,12 +160,12 @@ object EntryBundles extends HasCircularQueuePtrHelper {
     val canIssue              = Bool()
     val enqReady              = Bool()
     val deqSuccess            = Bool()
-    val srcWakeup             = Vec(params.numRegSrc, Bool())
     val srcWakeupByWB         = Vec(params.numRegSrc, Bool())
     val vlWakeupByIntWb       = Bool()
     val vlWakeupByVfWb        = Bool()
     val srcCancelVec          = Vec(params.numRegSrc, Bool())
     val srcLoadCancelVec      = Vec(params.numRegSrc, Bool())
+    val srcLoadTransCancelVec = Vec(params.numRegSrc, Bool())
     val srcLoadDependencyNext = Vec(params.numRegSrc, Vec(LoadPipelineWidth, UInt(LoadDependencyWidth.W)))
   }
 
@@ -174,7 +174,6 @@ object EntryBundles extends HasCircularQueuePtrHelper {
     common.flushed            := status.robIdx.needFlush(commonIn.flush)
     common.deqSuccess         := (if (params.isVecMemIQ) status.issued else true.B) &&
       commonIn.issueResp.valid && RespType.succeed(commonIn.issueResp.bits.resp) && !common.srcLoadCancelVec.asUInt.orR
-    common.srcWakeup          := common.srcWakeupByWB.zip(hasIQWakeupGet.srcWakeupByIQ).map { case (x, y) => x || y.asUInt.orR }
     common.srcWakeupByWB      := commonIn.wakeUpFromWB.map{ bundle => 
                                     val psrcSrcTypeVec = status.srcStatus.map(_.psrc) zip status.srcStatus.map(_.srcType)
                                     if (params.numRegSrc == 5) {
@@ -188,10 +187,10 @@ object EntryBundles extends HasCircularQueuePtrHelper {
     common.canIssue           := validReg && status.canIssue
     common.enqReady           := !validReg || commonIn.transSel
     common.clear              := common.flushed || common.deqSuccess || commonIn.transSel
-    common.srcCancelVec.zip(common.srcLoadCancelVec).zip(hasIQWakeupGet.srcWakeupByIQWithoutCancel).zipWithIndex.foreach { case (((srcCancel, srcLoadCancel), wakeUpByIQVec), srcIdx) =>
-      val ldTransCancel = if(params.hasIQWakeUp) Mux1H(wakeUpByIQVec, hasIQWakeupGet.wakeupLoadDependencyByIQVec.map(dep => LoadShouldCancel(Some(dep), commonIn.ldCancel))) else false.B
-      srcLoadCancel := LoadShouldCancel(Some(status.srcStatus(srcIdx).srcLoadDependency), commonIn.ldCancel)
-      srcCancel := srcLoadCancel || ldTransCancel
+    common.srcCancelVec.zip(hasIQWakeupGet.srcWakeupByIQWithoutCancel).zipWithIndex.foreach { case ((srcCancel, wakeUpByIQVec), srcIdx) =>
+      common.srcLoadTransCancelVec(srcIdx) := (if(params.hasIQWakeUp) Mux1H(wakeUpByIQVec, hasIQWakeupGet.wakeupLoadDependencyByIQVec.map(dep => LoadShouldCancel(Some(dep), commonIn.ldCancel))) else false.B)
+      common.srcLoadCancelVec(srcIdx) := LoadShouldCancel(Some(status.srcStatus(srcIdx).srcLoadDependency), commonIn.ldCancel)
+      srcCancel := common.srcLoadTransCancelVec(srcIdx) || common.srcLoadCancelVec(srcIdx)
     }
     common.srcLoadDependencyNext.zip(status.srcStatus.map(_.srcLoadDependency)).foreach { case (ldsNext, lds) =>
       ldsNext.zip(lds).foreach{ case (ldNext, ld) => ldNext := ld << 1 }
@@ -279,18 +278,22 @@ object EntryBundles extends HasCircularQueuePtrHelper {
     }.reduce(_ || _)
   }
 
-  def EntryRegCommonConnect(common: CommonWireBundle, hasIQWakeup: Option[CommonIQWakeupBundle], validReg: Bool, entryUpdate: EntryBundle, entryReg: EntryBundle, status: Status, commonIn: CommonInBundle, isEnq: Boolean)(implicit p: Parameters, params: IssueBlockParams) = {
+  def EntryRegCommonConnect(common: CommonWireBundle, hasIQWakeup: Option[CommonIQWakeupBundle], validReg: Bool, entryUpdate: EntryBundle, entryReg: EntryBundle, status: Status, commonIn: CommonInBundle, isEnq: Boolean, isComp: Boolean)(implicit p: Parameters, params: IssueBlockParams) = {
     val hasIQWakeupGet                                 = hasIQWakeup.getOrElse(0.U.asTypeOf(new CommonIQWakeupBundle))
-    val cancelByLd                                     = common.srcCancelVec.asUInt.orR
-    val cancelWhenWakeup                               = VecInit(hasIQWakeupGet.srcWakeupByIQButCancel.map(_.asUInt.orR)).asUInt.orR
+    val cancelBypassVec                                = Wire(Vec(params.numRegSrc, Bool()))
+    val srcCancelByLoad                                = common.srcLoadCancelVec.asUInt.orR
     val respIssueFail                                  = commonIn.issueResp.valid && RespType.isBlocked(commonIn.issueResp.bits.resp)
     entryUpdate.status.robIdx                         := status.robIdx
     entryUpdate.status.fuType                         := IQFuType.readFuType(status.fuType, params.getFuCfgs.map(_.fuType))
     entryUpdate.status.srcStatus.zip(status.srcStatus).zipWithIndex.foreach { case ((srcStatusNext, srcStatus), srcIdx) =>
-      val cancel = common.srcCancelVec(srcIdx)
-      val wakeupByIQ = hasIQWakeupGet.srcWakeupByIQ(srcIdx).asUInt.orR
+      val srcLoadCancel = common.srcLoadCancelVec(srcIdx)
+      val loadTransCancel = common.srcLoadTransCancelVec(srcIdx)
+      val wakeupByWB = common.srcWakeupByWB(srcIdx)
+      val wakeupByIQ = hasIQWakeupGet.srcWakeupByIQ(srcIdx).asUInt.orR && !loadTransCancel
       val wakeupByIQOH = hasIQWakeupGet.srcWakeupByIQ(srcIdx)
-      val wakeup = common.srcWakeup(srcIdx)
+      val wakeupByMemIQ = wakeupByIQOH.zip(commonIn.wakeUpFromIQ).filter(_._2.bits.params.isMemExeUnit).map(_._1).fold(false.B)(_ || _)
+      cancelBypassVec(srcIdx) := (if (isComp) Mux(hasIQWakeupGet.srcWakeupByIQWithoutCancel(srcIdx).asUInt.orR, loadTransCancel, srcLoadCancel)
+                                  else srcLoadCancel)
 
       val ignoreOldVd = Wire(Bool())
       val vlWakeUpByIntWb = common.vlWakeupByIntWb
@@ -323,15 +326,15 @@ object EntryBundles extends HasCircularQueuePtrHelper {
 
       srcStatusNext.psrc                              := srcStatus.psrc
       srcStatusNext.srcType                           := Mux(ignoreOldVd, SrcType.no, srcStatus.srcType)
-      srcStatusNext.srcState                          := Mux(cancel, false.B, wakeup | srcStatus.srcState | ignoreOldVd)
+      srcStatusNext.srcState                          := srcStatus.srcState & !srcLoadCancel | wakeupByWB | wakeupByIQ | ignoreOldVd
       srcStatusNext.dataSources.value                 := (if (params.inVfSchd && params.readVfRf && params.hasIQWakeUp) {
                                                             // Vf / Mem -> Vf
-                                                            val isWakeupByMemIQ = wakeupByIQOH.zip(commonIn.wakeUpFromIQ).filter(_._2.bits.params.isMemExeUnit).map(_._1).fold(false.B)(_ || _)
                                                             MuxCase(srcStatus.dataSources.value, Seq(
-                                                              (wakeupByIQ && isWakeupByMemIQ)    -> DataSource.bypass2,
-                                                              (wakeupByIQ && !isWakeupByMemIQ)   -> DataSource.bypass,
-                                                              srcStatus.dataSources.readBypass   -> DataSource.bypass2,
-                                                              srcStatus.dataSources.readBypass2  -> DataSource.reg,
+                                                              ignoreOldVd                       -> DataSource.imm,
+                                                              (wakeupByIQ && wakeupByMemIQ)     -> DataSource.bypass2,
+                                                              (wakeupByIQ && !wakeupByMemIQ)    -> DataSource.bypass,
+                                                              srcStatus.dataSources.readBypass  -> DataSource.bypass2,
+                                                              srcStatus.dataSources.readBypass2 -> DataSource.reg,
                                                             ))
                                                           }
                                                           else if (params.inMemSchd && params.readVfRf && params.hasIQWakeUp) {
@@ -345,6 +348,7 @@ object EntryBundles extends HasCircularQueuePtrHelper {
                                                           }
                                                           else {
                                                             MuxCase(srcStatus.dataSources.value, Seq(
+                                                              ignoreOldVd                        -> DataSource.imm,
                                                               wakeupByIQ                         -> DataSource.bypass,
                                                               srcStatus.dataSources.readBypass   -> DataSource.reg,
                                                             ))
@@ -364,20 +368,15 @@ object EntryBundles extends HasCircularQueuePtrHelper {
         val wakeupRCIdx = Mux1H(wakeupSrcExuWriteRC.map(_._1), wakeupSrcExuWriteRC.map(_._2.bits.rcDest.get))
         val replaceRC   = wakeupSrcExuWriteRC.map(x => x._2.bits.rfWen && x._2.bits.rcDest.get === srcStatus.regCacheIdx.get).fold(false.B)(_ || _)
 
-        srcStatusNext.useRegCache.get                 := MuxCase(srcStatus.useRegCache.get, Seq(
-                                                            cancel    -> false.B,
-                                                            wakeupRC  -> true.B,
-                                                            replaceRC -> false.B,
-                                                         ))
+        srcStatusNext.useRegCache.get                 := srcStatus.useRegCache.get && !(srcLoadCancel || replaceRC) || wakeupRC
         srcStatusNext.regCacheIdx.get                 := Mux(wakeupRC, wakeupRCIdx, srcStatus.regCacheIdx.get)
       }
     }
     entryUpdate.status.blocked                        := false.B
     entryUpdate.status.issued                         := MuxCase(status.issued, Seq(
-      (cancelByLd || cancelWhenWakeup || respIssueFail) -> false.B,
-      commonIn.deqSel                                   -> true.B,
-      !status.srcReady                                  -> false.B,
-    ))
+                                                          (commonIn.deqSel && !cancelBypassVec.asUInt.orR)  -> true.B,
+                                                          (srcCancelByLoad || respIssueFail)                -> false.B,
+                                                         ))
     entryUpdate.status.firstIssue                     := commonIn.deqSel || status.firstIssue
     entryUpdate.status.issueTimer                     := Mux(commonIn.deqSel, 0.U, Mux(status.issued, Mux(status.issueTimer === "b11".U, status.issueTimer, status.issueTimer + 1.U), "b11".U))
     entryUpdate.status.deqPortIdx                     := Mux(commonIn.deqSel, commonIn.deqPortIdxWrite, Mux(status.issued, status.deqPortIdx, 0.U))
@@ -438,17 +437,9 @@ object EntryBundles extends HasCircularQueuePtrHelper {
       }
     }
 
-    val srcLoadDependencyForCancel                     = Wire(chiselTypeOf(common.srcLoadDependencyNext))
     val srcLoadDependencyOut                           = Wire(chiselTypeOf(common.srcLoadDependencyNext))
     if(params.hasIQWakeUp) {
-      val wakeupSrcLoadDependency                      = hasIQWakeupGet.srcWakeupByIQWithoutCancel.map(x => Mux1H(x, hasIQWakeupGet.wakeupLoadDependencyByIQVec))
       val wakeupSrcLoadDependencyNext                  = hasIQWakeupGet.srcWakeupByIQWithoutCancel.map(x => Mux1H(x, hasIQWakeupGet.shiftedWakeupLoadDependencyByIQVec))
-      srcLoadDependencyForCancel.zipWithIndex.foreach { case (ldOut, srcIdx) =>
-        ldOut                                         := (if (isComp) Mux(hasIQWakeupGet.srcWakeupByIQWithoutCancel(srcIdx).asUInt.orR,
-                                                                      wakeupSrcLoadDependency(srcIdx),
-                                                                      status.srcStatus(srcIdx).srcLoadDependency)
-                                                          else status.srcStatus(srcIdx).srcLoadDependency)
-      }
       srcLoadDependencyOut.zipWithIndex.foreach { case (ldOut, srcIdx) =>
         ldOut                                         := (if (isComp) Mux(hasIQWakeupGet.srcWakeupByIQWithoutCancel(srcIdx).asUInt.orR,
                                                                       wakeupSrcLoadDependencyNext(srcIdx),
@@ -456,10 +447,12 @@ object EntryBundles extends HasCircularQueuePtrHelper {
                                                           else common.srcLoadDependencyNext(srcIdx))
       }
     } else {
-      srcLoadDependencyForCancel                      := status.srcStatus.map(_.srcLoadDependency)
       srcLoadDependencyOut                            := common.srcLoadDependencyNext
     }
-    commonOut.cancelBypass                            := srcLoadDependencyForCancel.map(x => LoadShouldCancel(Some(x), commonIn.ldCancel)).reduce(_ | _)
+    commonOut.cancelBypass                            := VecInit(hasIQWakeupGet.srcWakeupByIQWithoutCancel.zipWithIndex.map{ case (wakeupVec, srcIdx) =>
+                                                            if (isComp) Mux(wakeupVec.asUInt.orR, common.srcLoadTransCancelVec(srcIdx), common.srcLoadCancelVec(srcIdx))
+                                                            else common.srcLoadCancelVec(srcIdx)
+                                                         }).asUInt.orR
     commonOut.entry.bits.status.srcStatus.map(_.srcLoadDependency).zipWithIndex.foreach { case (ldOut, srcIdx) =>
       ldOut                                           := srcLoadDependencyOut(srcIdx)
     }
diff --git a/src/main/scala/xiangshan/backend/issue/OthersEntry.scala b/src/main/scala/xiangshan/backend/issue/OthersEntry.scala
index b0f0276c0ba..24dada2f588 100644
--- a/src/main/scala/xiangshan/backend/issue/OthersEntry.scala
+++ b/src/main/scala/xiangshan/backend/issue/OthersEntry.scala
@@ -53,7 +53,7 @@ class OthersEntry(isComp: Boolean)(implicit p: Parameters, params: IssueBlockPar
     entryRegNext := entryUpdate
   }
 
-  EntryRegCommonConnect(common, hasWakeupIQ, validReg, entryUpdate, entryReg, entryReg.status, io.commonIn, false)
+  EntryRegCommonConnect(common, hasWakeupIQ, validReg, entryUpdate, entryReg, entryReg.status, io.commonIn, false, isComp)
 
   //output
   CommonOutConnect(io.commonOut, common, hasWakeupIQ, validReg, entryUpdate, entryReg, entryReg.status, io.commonIn, false, isComp)
diff --git a/src/main/scala/xiangshan/backend/regcache/RegCacheTagModule.scala b/src/main/scala/xiangshan/backend/regcache/RegCacheTagModule.scala
index e4ddf86a8e7..1f0196df91c 100644
--- a/src/main/scala/xiangshan/backend/regcache/RegCacheTagModule.scala
+++ b/src/main/scala/xiangshan/backend/regcache/RegCacheTagModule.scala
@@ -93,7 +93,6 @@ class RegCacheTagModule
     }
     .elsewhen (io.cancelVec(i)) {
       v(i)   := false.B
-      loadDependency(i) := 0.U.asTypeOf(loadDependency(i))
     }
     .elsewhen (loadDependency(i).map(x => x.orR).reduce(_ || _)) {
       loadDependency(i) := loadDependency(i).map(l => l << 1)
diff --git a/src/main/scala/xiangshan/backend/rename/BusyTable.scala b/src/main/scala/xiangshan/backend/rename/BusyTable.scala
index 5306aa095a1..2de99d4b070 100644
--- a/src/main/scala/xiangshan/backend/rename/BusyTable.scala
+++ b/src/main/scala/xiangshan/backend/rename/BusyTable.scala
@@ -123,11 +123,10 @@ class BusyTable(numReadPorts: Int, numWritePorts: Int, numPhyPregs: Int, pregWB:
   val ldCancelMask = loadDependency.map(x => LoadShouldCancel(Some(x), loadCancel))
 
   loadDependency.zipWithIndex.foreach{ case (ldDp, idx) =>
-    when(allocMask(idx) || wbMask(idx) || ldCancelMask(idx)) {
-      ldDp := 0.U.asTypeOf(ldDp)
-    }.elsewhen(wakeUpMask(idx)) {
+    when(wakeUpMask(idx)) {
       ldDp := (if (wakeUpIn.nonEmpty) Mux1H(wakeupOHVec(idx), shiftLoadDependency) else 0.U.asTypeOf(ldDp))
-    }.elsewhen(ldDp.map(x => x.orR).reduce(_ | _)) {
+    }
+    .elsewhen(ldDp.map(x => x.orR).reduce(_ | _)) {
       ldDp := VecInit(ldDp.map(x => x << 1))
     }
   }
@@ -139,22 +138,24 @@ class BusyTable(numReadPorts: Int, numWritePorts: Int, numPhyPregs: Int, pregWB:
     rename alloc => wakeUp / cancel => ... => wakeUp / cancel => wakeUp
   or
     rename alloc => wbMask  //TODO we still need wbMask because wakeUp signal is partial now
-  the bypass state lasts for a maximum of one cycle, cancel(=> busy) or else(=> regFile)
+  in wakeUpMask, we filter ogCancel and loadTransCancel at the same cycle
    */
   val table = VecInit((0 until numPhyPregs).zip(tableUpdate).map{ case (idx, update) =>
     RegEnable(update, 0.U(1.W), allocMask(idx) || ldCancelMask(idx) || wakeUpMask(idx) || wbMask(idx))
   }).asUInt
 
   tableUpdate.zipWithIndex.foreach{ case (update, idx) =>
-    when(allocMask(idx) || ldCancelMask(idx)) {
+    when(wakeUpMask(idx) || wbMask(idx)) {
+      update := false.B                                   //ready
+    }
+    .elsewhen(allocMask(idx) || ldCancelMask(idx)) {
       update := true.B                                    //busy
       if (idx == 0 && pregWB.isInstanceOf[IntWB]) {
           // Int RegFile 0 is always ready
           update := false.B
       }
-    }.elsewhen(wakeUpMask(idx) || wbMask(idx)) {
-      update := false.B                                   //ready
-    }.otherwise {
+    }
+    .otherwise {
       update := table(idx)
     }
   }
