diff --git a/src/main/scala/xiangshan/frontend/BPU.scala b/src/main/scala/xiangshan/frontend/BPU.scala
index 080569e4b16..cfea2612403 100644
--- a/src/main/scala/xiangshan/frontend/BPU.scala
+++ b/src/main/scala/xiangshan/frontend/BPU.scala
@@ -954,7 +954,7 @@ class Predictor(implicit p: Parameters) extends XSModule with HasBPUConst with H
     val misPredictMask:      UInt      = io.ftq_to_bpu.update.bits.mispred_mask.asUInt
     val takenMask: UInt =
       io.ftq_to_bpu.update.bits.br_taken_mask.asUInt |
-        io.ftq_to_bpu.update.bits.ftb_entry.always_taken.asUInt // Always taken branch is recorded in history
+        io.ftq_to_bpu.update.bits.ftb_entry.strong_bias.asUInt // Always taken branch is recorded in history
     val takenIdx:      UInt = (PriorityEncoder(takenMask) + 1.U((log2Ceil(numBr) + 1).W)).asUInt
     val misPredictIdx: UInt = (PriorityEncoder(misPredictMask) + 1.U((log2Ceil(numBr) + 1).W)).asUInt
     val shouldShiftMask: UInt = Mux(takenMask.orR, LowerMask(takenIdx).asUInt, ((1 << numBr) - 1).asUInt) &
diff --git a/src/main/scala/xiangshan/frontend/FTB.scala b/src/main/scala/xiangshan/frontend/FTB.scala
index bfbe108dff5..75d5b26c92b 100644
--- a/src/main/scala/xiangshan/frontend/FTB.scala
+++ b/src/main/scala/xiangshan/frontend/FTB.scala
@@ -187,7 +187,9 @@ class FTBEntry(implicit p: Parameters) extends FTBEntry_part with FTBParams with
 
   val last_may_be_rvi_call = Bool()
 
-  val always_taken = Vec(numBr, Bool())
+  // Mark the conditional branch for the first jump and the jalr instruction that appears for the first time,
+  // and train the tag/ittage without using its results when strong_bias is true.
+  val strong_bias = Vec(numBr, Bool())
 
   def getSlotForBr(idx: Int): FtbSlot = {
     require(idx <= numBr - 1)
@@ -369,7 +371,7 @@ class FTBEntry(implicit p: Parameters) extends FTBEntry_part with FTBParams with
     val isJalrDiff           = this.isJalr === that.isJalr
     val lastMayBeRviCallDiff = this.last_may_be_rvi_call === that.last_may_be_rvi_call
     val alwaysTakenDiff: IndexedSeq[Bool] =
-      this.always_taken.zip(that.always_taken).map {
+      this.strong_bias.zip(that.strong_bias).map {
         case (x, y) => x === y
       }
     VecInit(
@@ -797,12 +799,12 @@ class FTB(implicit p: Parameters) extends BasePredictor with FTBParams with BPUU
       out_fp & in_fp & s2_hit & s2_ftb_entry <-
         io.out.s2.full_pred zip io.in.bits.resp_in(0).s2.full_pred zip s2_hit_dup zip s2_ftb_entry_dup
     )
-      out_fp.br_taken_mask(i) := in_fp.br_taken_mask(i) || s2_hit && s2_ftb_entry.always_taken(i)
+      out_fp.br_taken_mask(i) := in_fp.br_taken_mask(i) || s2_hit && s2_ftb_entry.strong_bias(i)
     for (
       out_fp & in_fp & s3_hit & s3_ftb_entry <-
         io.out.s3.full_pred zip io.in.bits.resp_in(0).s3.full_pred zip s3_hit_dup zip s3_ftb_entry_dup
     )
-      out_fp.br_taken_mask(i) := in_fp.br_taken_mask(i) || s3_hit && s3_ftb_entry.always_taken(i)
+      out_fp.br_taken_mask(i) := in_fp.br_taken_mask(i) || s3_hit && s3_ftb_entry.strong_bias(i)
   }
 
   // Update logic
diff --git a/src/main/scala/xiangshan/frontend/FauFTB.scala b/src/main/scala/xiangshan/frontend/FauFTB.scala
index d0e3dd8bbc4..7d4c86a4397 100644
--- a/src/main/scala/xiangshan/frontend/FauFTB.scala
+++ b/src/main/scala/xiangshan/frontend/FauFTB.scala
@@ -104,7 +104,7 @@ class FauFTB(implicit p: Parameters) extends BasePredictor with FauFTBParams {
     fp.multiHit := false.B
     fp.fromFtbEntry(e, s1_pc_dup(0))
     for (i <- 0 until numBr) {
-      fp.br_taken_mask(i) := c(i)(1) || e.always_taken(i)
+      fp.br_taken_mask(i) := c(i)(1) || e.strong_bias(i)
     }
   }
   val s1_hit_full_pred   = Mux1H(s1_hit_oh, s1_possible_full_preds)
@@ -149,7 +149,7 @@ class FauFTB(implicit p: Parameters) extends BasePredictor with FauFTBParams {
   val u_s0_hit    = u_s0_hit_oh.orR
   val u_s0_br_update_valids =
     VecInit((0 until numBr).map(w =>
-      u.bits.ftb_entry.brValids(w) && u.valid && !u.bits.ftb_entry.always_taken(w) &&
+      u.bits.ftb_entry.brValids(w) && u.valid && !u.bits.ftb_entry.strong_bias(w) &&
         !(PriorityEncoder(u.bits.br_taken_mask) < w.U)
     ))
 
diff --git a/src/main/scala/xiangshan/frontend/ITTAGE.scala b/src/main/scala/xiangshan/frontend/ITTAGE.scala
index a38e1a9a55f..8010c97901d 100644
--- a/src/main/scala/xiangshan/frontend/ITTAGE.scala
+++ b/src/main/scala/xiangshan/frontend/ITTAGE.scala
@@ -414,7 +414,9 @@ class ITTage(implicit p: Parameters) extends BaseITTage {
   val update  = io.update.bits
   val updateValid =
     update.is_jalr && !update.is_ret && u_valid && update.ftb_entry.jmpValid &&
-      update.jmp_taken && update.cfi_idx.valid && update.cfi_idx.bits === update.ftb_entry.tailSlot.offset
+      update.jmp_taken && update.cfi_idx.valid && update.cfi_idx.bits === update.ftb_entry.tailSlot.offset && !update.ftb_entry.strong_bias(
+        numBr - 1
+      )
 
   // meta is splited by composer
   val updateMeta = update.meta.asTypeOf(new ITTageMeta)
diff --git a/src/main/scala/xiangshan/frontend/NewFtq.scala b/src/main/scala/xiangshan/frontend/NewFtq.scala
index c41c6cc5ad0..b567d3cddec 100644
--- a/src/main/scala/xiangshan/frontend/NewFtq.scala
+++ b/src/main/scala/xiangshan/frontend/NewFtq.scala
@@ -246,12 +246,12 @@ class FTBEntryGen(implicit p: Parameters) extends XSModule with HasBackendRedire
     val mispred_mask      = Output(Vec(numBr + 1, Bool()))
 
     // for perf counters
-    val is_init_entry            = Output(Bool())
-    val is_old_entry             = Output(Bool())
-    val is_new_br                = Output(Bool())
-    val is_jalr_target_modified  = Output(Bool())
-    val is_always_taken_modified = Output(Bool())
-    val is_br_full               = Output(Bool())
+    val is_init_entry           = Output(Bool())
+    val is_old_entry            = Output(Bool())
+    val is_new_br               = Output(Bool())
+    val is_jalr_target_modified = Output(Bool())
+    val is_strong_bias_modified = Output(Bool())
+    val is_br_full              = Output(Bool())
   })
 
   // no mispredictions detected at predecode
@@ -284,7 +284,7 @@ class FTBEntryGen(implicit p: Parameters) extends XSModule with HasBackendRedire
     init_br_slot.valid  := true.B
     init_br_slot.offset := io.cfiIndex.bits
     init_br_slot.setLowerStatByTarget(io.start_addr, io.target, numBr == 1)
-    init_entry.always_taken(0) := true.B // set to always taken on init
+    init_entry.strong_bias(0) := true.B // set to strong bias on init
   }
 
   // case jmp
@@ -292,6 +292,7 @@ class FTBEntryGen(implicit p: Parameters) extends XSModule with HasBackendRedire
     init_entry.tailSlot.offset := pd.jmpOffset
     init_entry.tailSlot.valid  := new_jmp_is_jal || new_jmp_is_jalr
     init_entry.tailSlot.setLowerStatByTarget(io.start_addr, Mux(cfi_is_jalr, io.target, pd.jalTarget), isShare = false)
+    init_entry.strong_bias(numBr - 1) := new_jmp_is_jalr // set strong bias for the jalr on init
   }
 
   val jmpPft = getLower(io.start_addr) +& pd.jmpOffset +& Mux(pd.rvcMask(pd.jmpOffset), 1.U, 2.U)
@@ -329,9 +330,9 @@ class FTBEntryGen(implicit p: Parameters) extends XSModule with HasBackendRedire
       slot.valid  := true.B
       slot.offset := new_br_offset
       slot.setLowerStatByTarget(io.start_addr, io.target, i == numBr - 1)
-      old_entry_modified.always_taken(i) := true.B
+      old_entry_modified.strong_bias(i) := true.B
     }.elsewhen(new_br_offset > oe.allSlotsForBr(i).offset) {
-      old_entry_modified.always_taken(i) := false.B
+      old_entry_modified.strong_bias(i) := false.B
       // all other fields remain unchanged
     }.otherwise {
       // case i == 0, remain unchanged
@@ -339,7 +340,7 @@ class FTBEntryGen(implicit p: Parameters) extends XSModule with HasBackendRedire
         val noNeedToMoveFromFormerSlot = (i == numBr - 1).B && !oe.brSlots.last.valid
         when(!noNeedToMoveFromFormerSlot) {
           slot.fromAnotherSlot(oe.allSlotsForBr(i - 1))
-          old_entry_modified.always_taken(i) := oe.always_taken(i)
+          old_entry_modified.strong_bias(i) := oe.strong_bias(i)
         }
       }
     }
@@ -371,20 +372,28 @@ class FTBEntryGen(implicit p: Parameters) extends XSModule with HasBackendRedire
   val jalr_target_modified = cfi_is_jalr && (old_target =/= io.target) && old_tail_is_jmp // TODO: pass full jalr target
   when(jalr_target_modified) {
     old_entry_jmp_target_modified.setByJmpTarget(io.start_addr, io.target)
-    old_entry_jmp_target_modified.always_taken := 0.U.asTypeOf(Vec(numBr, Bool()))
+    old_entry_jmp_target_modified.strong_bias := 0.U.asTypeOf(Vec(numBr, Bool()))
   }
 
-  val old_entry_always_taken    = WireInit(oe)
-  val always_taken_modified_vec = Wire(Vec(numBr, Bool())) // whether modified or not
+  val old_entry_strong_bias    = WireInit(oe)
+  val strong_bias_modified_vec = Wire(Vec(numBr, Bool())) // whether modified or not
   for (i <- 0 until numBr) {
-    old_entry_always_taken.always_taken(i) :=
-      oe.always_taken(i) && io.cfiIndex.valid && oe.brValids(i) && io.cfiIndex.bits === oe.brOffset(i)
-    always_taken_modified_vec(i) := oe.always_taken(i) && !old_entry_always_taken.always_taken(i)
+    when(br_recorded_vec(0)) {
+      old_entry_strong_bias.strong_bias(0) :=
+        oe.strong_bias(0) && io.cfiIndex.valid && oe.brValids(0) && io.cfiIndex.bits === oe.brOffset(0)
+    }.elsewhen(br_recorded_vec(numBr - 1)) {
+      old_entry_strong_bias.strong_bias(0) := false.B
+      old_entry_strong_bias.strong_bias(numBr - 1) :=
+        oe.strong_bias(numBr - 1) && io.cfiIndex.valid && oe.brValids(numBr - 1) && io.cfiIndex.bits === oe.brOffset(
+          numBr - 1
+        )
+    }
+    strong_bias_modified_vec(i) := oe.strong_bias(i) && oe.brValids(i) && !old_entry_strong_bias.strong_bias(i)
   }
-  val always_taken_modified = always_taken_modified_vec.reduce(_ || _)
+  val strong_bias_modified = strong_bias_modified_vec.reduce(_ || _)
 
   val derived_from_old_entry =
-    Mux(is_new_br, old_entry_modified, Mux(jalr_target_modified, old_entry_jmp_target_modified, old_entry_always_taken))
+    Mux(is_new_br, old_entry_modified, Mux(jalr_target_modified, old_entry_jmp_target_modified, old_entry_strong_bias))
 
   io.new_entry := Mux(!hit, init_entry, derived_from_old_entry)
 
@@ -399,12 +408,12 @@ class FTBEntryGen(implicit p: Parameters) extends XSModule with HasBackendRedire
   io.mispred_mask.last := io.new_entry.jmpValid && io.mispredict_vec(pd.jmpOffset)
 
   // for perf counters
-  io.is_init_entry            := !hit
-  io.is_old_entry             := hit && !is_new_br && !jalr_target_modified && !always_taken_modified
-  io.is_new_br                := hit && is_new_br
-  io.is_jalr_target_modified  := hit && jalr_target_modified
-  io.is_always_taken_modified := hit && always_taken_modified
-  io.is_br_full               := hit && is_new_br && may_have_to_replace
+  io.is_init_entry           := !hit
+  io.is_old_entry            := hit && !is_new_br && !jalr_target_modified && !strong_bias_modified
+  io.is_new_br               := hit && is_new_br
+  io.is_jalr_target_modified := hit && jalr_target_modified
+  io.is_strong_bias_modified := hit && strong_bias_modified
+  io.is_br_full              := hit && is_new_br && may_have_to_replace
 }
 
 class FtqPcMemWrapper(numOtherReads: Int)(implicit p: Parameters) extends XSModule with HasBackendRedirectInfo {
@@ -1628,12 +1637,12 @@ class Ftq(implicit p: Parameters) extends XSModule with HasCircularQueuePtrHelpe
   val ftb_old_entry = u(ftbEntryGen.is_old_entry)
 
   val ftb_modified_entry =
-    u(ftbEntryGen.is_new_br || ftbEntryGen.is_jalr_target_modified || ftbEntryGen.is_always_taken_modified)
+    u(ftbEntryGen.is_new_br || ftbEntryGen.is_jalr_target_modified || ftbEntryGen.is_strong_bias_modified)
   val ftb_modified_entry_new_br               = u(ftbEntryGen.is_new_br)
   val ftb_modified_entry_ifu_redirected       = u(ifuRedirected(do_commit_ptr.value))
   val ftb_modified_entry_jalr_target_modified = u(ftbEntryGen.is_jalr_target_modified)
   val ftb_modified_entry_br_full              = ftb_modified_entry && ftbEntryGen.is_br_full
-  val ftb_modified_entry_always_taken         = ftb_modified_entry && ftbEntryGen.is_always_taken_modified
+  val ftb_modified_entry_strong_bias          = ftb_modified_entry && ftbEntryGen.is_strong_bias_modified
 
   def getFtbEntryLen(pc: UInt, entry: FTBEntry) = (entry.getFallThrough(pc) - pc) >> instOffsetBits
   val gen_ftb_entry_len = getFtbEntryLen(update.pc, ftbEntryGen.new_entry)
@@ -1645,32 +1654,32 @@ class Ftq(implicit p: Parameters) extends XSModule with HasCircularQueuePtrHelpe
   XSPerfHistogram("ftq_has_entry", validEntries, true.B, 0, FtqSize + 1, 1)
 
   val perfCountsMap = Map(
-    "BpInstr"                         -> PopCount(mbpInstrs),
-    "BpBInstr"                        -> PopCount(mbpBRights | mbpBWrongs),
-    "BpRight"                         -> PopCount(mbpRights),
-    "BpWrong"                         -> PopCount(mbpWrongs),
-    "BpBRight"                        -> PopCount(mbpBRights),
-    "BpBWrong"                        -> PopCount(mbpBWrongs),
-    "BpJRight"                        -> PopCount(mbpJRights),
-    "BpJWrong"                        -> PopCount(mbpJWrongs),
-    "BpIRight"                        -> PopCount(mbpIRights),
-    "BpIWrong"                        -> PopCount(mbpIWrongs),
-    "BpCRight"                        -> PopCount(mbpCRights),
-    "BpCWrong"                        -> PopCount(mbpCWrongs),
-    "BpRRight"                        -> PopCount(mbpRRights),
-    "BpRWrong"                        -> PopCount(mbpRWrongs),
-    "ftb_false_hit"                   -> PopCount(ftb_false_hit),
-    "ftb_hit"                         -> PopCount(ftb_hit),
-    "ftb_new_entry"                   -> PopCount(ftb_new_entry),
-    "ftb_new_entry_only_br"           -> PopCount(ftb_new_entry_only_br),
-    "ftb_new_entry_only_jmp"          -> PopCount(ftb_new_entry_only_jmp),
-    "ftb_new_entry_has_br_and_jmp"    -> PopCount(ftb_new_entry_has_br_and_jmp),
-    "ftb_old_entry"                   -> PopCount(ftb_old_entry),
-    "ftb_modified_entry"              -> PopCount(ftb_modified_entry),
-    "ftb_modified_entry_new_br"       -> PopCount(ftb_modified_entry_new_br),
-    "ftb_jalr_target_modified"        -> PopCount(ftb_modified_entry_jalr_target_modified),
-    "ftb_modified_entry_br_full"      -> PopCount(ftb_modified_entry_br_full),
-    "ftb_modified_entry_always_taken" -> PopCount(ftb_modified_entry_always_taken)
+    "BpInstr"                        -> PopCount(mbpInstrs),
+    "BpBInstr"                       -> PopCount(mbpBRights | mbpBWrongs),
+    "BpRight"                        -> PopCount(mbpRights),
+    "BpWrong"                        -> PopCount(mbpWrongs),
+    "BpBRight"                       -> PopCount(mbpBRights),
+    "BpBWrong"                       -> PopCount(mbpBWrongs),
+    "BpJRight"                       -> PopCount(mbpJRights),
+    "BpJWrong"                       -> PopCount(mbpJWrongs),
+    "BpIRight"                       -> PopCount(mbpIRights),
+    "BpIWrong"                       -> PopCount(mbpIWrongs),
+    "BpCRight"                       -> PopCount(mbpCRights),
+    "BpCWrong"                       -> PopCount(mbpCWrongs),
+    "BpRRight"                       -> PopCount(mbpRRights),
+    "BpRWrong"                       -> PopCount(mbpRWrongs),
+    "ftb_false_hit"                  -> PopCount(ftb_false_hit),
+    "ftb_hit"                        -> PopCount(ftb_hit),
+    "ftb_new_entry"                  -> PopCount(ftb_new_entry),
+    "ftb_new_entry_only_br"          -> PopCount(ftb_new_entry_only_br),
+    "ftb_new_entry_only_jmp"         -> PopCount(ftb_new_entry_only_jmp),
+    "ftb_new_entry_has_br_and_jmp"   -> PopCount(ftb_new_entry_has_br_and_jmp),
+    "ftb_old_entry"                  -> PopCount(ftb_old_entry),
+    "ftb_modified_entry"             -> PopCount(ftb_modified_entry),
+    "ftb_modified_entry_new_br"      -> PopCount(ftb_modified_entry_new_br),
+    "ftb_jalr_target_modified"       -> PopCount(ftb_modified_entry_jalr_target_modified),
+    "ftb_modified_entry_br_full"     -> PopCount(ftb_modified_entry_br_full),
+    "ftb_modified_entry_strong_bias" -> PopCount(ftb_modified_entry_strong_bias)
   ) ++ mispred_stage_map ++ br_mispred_stage_map ++ jalr_mispred_stage_map ++
     correct_stage_map ++ br_correct_stage_map ++ jalr_correct_stage_map
 
diff --git a/src/main/scala/xiangshan/frontend/Tage.scala b/src/main/scala/xiangshan/frontend/Tage.scala
index 0861c85f00e..4d6f8a146bb 100644
--- a/src/main/scala/xiangshan/frontend/Tage.scala
+++ b/src/main/scala/xiangshan/frontend/Tage.scala
@@ -681,7 +681,7 @@ class Tage(implicit p: Parameters) extends BaseTage {
   val u_valid = io.update.valid
   val update  = io.update.bits
   val updateValids = VecInit((0 until TageBanks).map(w =>
-    update.ftb_entry.brValids(w) && u_valid && !update.ftb_entry.always_taken(w) &&
+    update.ftb_entry.brValids(w) && u_valid && !update.ftb_entry.strong_bias(w) &&
       !(PriorityEncoder(update.br_taken_mask) < w.U)
   ))
